From c1f85bf8817c84fc9a8d5345098b36629c81848a Mon Sep 17 00:00:00 2001
From: Christian Reisswig <reisswig@scriwalker.(none)>
Date: Sun, 3 Mar 2013 23:00:22 -0800
Subject: [PATCH 02/11] GRHydro: Make recon_method steerable at recovery.

---
 param.ccl |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/param.ccl b/param.ccl
index 44f6b22..9dad537 100644
--- a/param.ccl
+++ b/param.ccl
@@ -134,7 +134,7 @@ CCTK_INT GRHydro_MaxNumSandRVars "The maximum number of save and restore variabl
   0:16		:: "A small range, depending on testing or not"
 } 16
 
-keyword recon_method "Which reconstruction method to use"
+keyword recon_method "Which reconstruction method to use" STEERABLE=RECOVER
 {
   "tvd"	 :: "Slope limited TVD"
   "ppm"  :: "PPM reconstruction"
-- 
1.7.10.4

