Index: schedule.ccl
===================================================================
--- schedule.ccl	(revision 72)
+++ schedule.ccl	(working copy)
@@ -55,6 +55,12 @@
   OPTIONS: global
 } "Initialise refluxing variables"
 
+SCHEDULE Refluxing_Init2 AT POST_RECOVER_VARIABLES
+{
+  LANG: C
+  OPTIONS: global
+} "Initialise refluxing variables after recoverign grid structure from checkpoint"
+
 if (delayed_refluxing && delayed_refluxing_sources) {
   SCHEDULE Refluxing_DelayedCorrectionReduction AT evol AFTER MoL_Evolution
   {
