Index: lib/make/configure.in
===================================================================
--- lib/make/configure.in	(revision 4850)
+++ lib/make/configure.in	(working copy)
@@ -943,6 +943,9 @@
 dnl Do we have isinf(3)
 CCTK_CHECK_FUNCS(isinf, , CCTK_CHECK_LIB_FUNC(m, isinf))
 
+dnl Do we have shmget(3)
+CCTK_CHECK_FUNCS(shmget, , )
+
 dnl Do we have mkstemp(3)
 CCTK_CHECK_FUNCS(mkstemp, , CCTK_CHECK_LIB_FUNC(c, mkstemp))
 
Index: lib/make/cctk_Config.h.in
===================================================================
--- lib/make/cctk_Config.h.in	(revision 4850)
+++ lib/make/cctk_Config.h.in	(working copy)
@@ -26,6 +26,7 @@
 #undef HAVE_FINITE
 #undef HAVE_ISNAN
 #undef HAVE_ISINF
+#undef HAVE_SHMGET
 #undef HAVE_MKSTEMP
 #undef HAVE_VA_COPY
 
