Index: configuration.ccl
===================================================================
--- configuration.ccl	(revision 128)
+++ configuration.ccl	(working copy)
@@ -4,7 +4,7 @@
 {
   SCRIPT src/detect.sh
   LANG bash
-  OPTIONS HDF5 HDF5_DIR HDF5_INSTALL_DIR HDF5_ENABLE_CXX HDF5_ENABLE_FORTRAN ZLIB_DIR LIBSZ_DIR LIBZ_DIR
+  OPTIONS HDF5 HDF5_DIR HDF5_INSTALL_DIR HDF5_ENABLE_CXX HDF5_ENABLE_FORTRAN ZLIB_DIR LIBSZ_DIR LIBZ_DIR HDF5_INC_DIRS HDF5_LIB_DIRS HDF5_LIBS
 }
 
 REQUIRES HDF5
Index: src/detect.sh
===================================================================
--- src/detect.sh	(revision 128)
+++ src/detect.sh	(working copy)
@@ -56,7 +56,7 @@
 HDF5_REQ_LIBS="${HDF5_CXX_LIBS} ${HDF5_FORTRAN_LIBS} ${HDF5_C_LIBS}"
 
 # Try to find the library if build isn't explicitly requested
-if [ -z "${HDF5_BUILD}" ]; then
+if [ -z "${HDF5_BUILD}" -a -z "${HDF5_INC_DIRS}" -a -z "${HDF_LIB_DIRS}" -a "${HDF5_LIBS}" ]; then
     find_lib HDF5 hdf5 1 1.0 "$HDF5_REQ_LIBS" "hdf5.h" "$HDF5_DIR"
 
     # Sadly, pkg-config for HDF5 is good for paths, but bad for the list of
