Index: OpenSSL/configure.sh
===================================================================
--- OpenSSL.orig/configure.sh	2014-01-11 18:20:20.029009958 -0800
+++ OpenSSL/configure.sh	2014-01-11 18:21:33.229583223 -0800
@@ -137,9 +137,9 @@
         unset options # OpenSSL's 'config' script uses $options itself
         # OpenSSL doesn't want to link with -fopenmp (can't pass
         # LDFLAGS?), so instead we remove all OpenMP flags
-        export CPPFLAGS=$(echo '' $CPPFLAGS | sed -e 's/-fopenmp//')
-        export CFLAGS=$(echo '' $CFLAGS | sed -e 's/-fopenmp//')
-        export LDFLAGS=$(echo '' $LDFLAGS | sed -e 's/-fopenmp//')
+        export CPPFLAGS=$(echo '' $CPPFLAGS | sed -e 's/-f?openmp//')
+        export CFLAGS=$(echo '' $CFLAGS | sed -e 's/-f?openmp//')
+        export LDFLAGS=$(echo '' $LDFLAGS | sed -e 's/-f?openmp//')
         # OpenSSL does not automatically build a 64bit version on 64bit Macs
         # setting KERNEL_BITS=64 tells it we want one
         if uname -v | grep >/dev/null '^Darwin.*RELEASE_X86_64' ; then
