Index: RunTestUtils.pl
===================================================================
--- RunTestUtils.pl	(revision 4723)
+++ RunTestUtils.pl	(working copy)
@@ -154,8 +154,8 @@
   {
     my $testdir = $arrangement_dir . $testdata->{"$thorn ARRANGEMENT"} .
                   "${sep}$thorn${sep}test";
-    my $old_config_file = "$testdir${sep}test.ccl";
-    my $new_config_file = "$testdir${sep}config";
+    my $old_config_file = "$testdir${sep}config";
+    my $new_config_file = "$testdir${sep}test.ccl";
     my $config_file = '';
 
     if (-r $old_config_file and -r $new_config_file)
