Index: configuration.ccl
===================================================================
--- configuration.ccl	(revision 46)
+++ configuration.ccl	(working copy)
@@ -4,7 +4,7 @@
 {
   SCRIPT src/detect.sh
   LANG bash
-  OPTIONS FFTW3_DIR FFTW3_INSTALL_DIR FFTW3_LIBS
+  OPTIONS FFTW3_DIR FFTW3_INSTALL_DIR FFTW3_LIBS FFTW3_INC_DIRS FFTW3_LIB_DIRS
 }
 
 # Pass configuration options to build script
Index: src/detect.sh
===================================================================
--- src/detect.sh	(revision 46)
+++ src/detect.sh	(working copy)
@@ -130,10 +130,10 @@
 
 # Set options
 if [ "${FFTW3_DIR}" != 'NO_BUILD' ]; then
-    : ${FFTW3_INC_DIRS="${FFTW3_DIR}/include"}
-    : ${FFTW3_LIB_DIRS="${FFTW3_DIR}/lib"}
+    : ${FFTW3_INC_DIRS:="${FFTW3_DIR}/include"}
+    : ${FFTW3_LIB_DIRS:="${FFTW3_DIR}/lib"}
 fi
-: ${FFTW3_LIBS='fftw3'}
+: ${FFTW3_LIBS:='fftw3'}
 
 FFTW3_INC_DIRS="$(${CCTK_HOME}/lib/sbin/strip-incdirs.sh ${FFTW3_INC_DIRS})"
 FFTW3_LIB_DIRS="$(${CCTK_HOME}/lib/sbin/strip-libdirs.sh ${FFTW3_LIB_DIRS})"
