diff --git a/GRHydro/schedule.ccl b/GRHydro/schedule.ccl
index a32c20d..40ac22b 100644
--- a/GRHydro/schedule.ccl
+++ b/GRHydro/schedule.ccl
@@ -382,7 +382,7 @@ else if (CCTK_IsThornActive("Carpet"))
 }
 
 
-schedule GRHydro_SqrtSpatialDeterminant AT CCTK_INITIAL AFTER (HydroBase_Initial,GRHydroTransformADMToLocalBasis) BEFORE HydroBase_Prim2ConInitial
+schedule GRHydro_SqrtSpatialDeterminant AT CCTK_INITIAL AFTER (HydroBase_Initial,GRHydroTransformADMToLocalBasis,ADMBase_PostInitial) BEFORE HydroBase_Prim2ConInitial
 {
   LANG: Fortran
 } "Calculate sdetg"
