Index: make/configure.in
===================================================================
--- make/configure.in	(revision 4655)
+++ make/configure.in	(working copy)
@@ -270,6 +270,12 @@
 
 # Set the warning flags if they've not been set by now
 # (using GNU compiler warning flags as a reasonable default)
+AC_SUBST(CPP_WARN_FLAGS)
+: ${CPP_WARN_FLAGS='-Wall'}
+
+AC_SUBST(FPP_WARN_FLAGS)
+: ${FPP_WARN_FLAGS='-Wall'}
+
 AC_SUBST(C_WARN_FLAGS)
 if test "$CC" = 'gcc' ; then
   : ${C_WARN_FLAGS='-Wall -Wshadow -Wpointer-arith -Wcast-qual -Wcast-align -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wbad-function-cast -Wsign-compare'}
@@ -307,6 +313,12 @@
 
 # Set the debug flags if they've not been set by now
 # (using '-g' as a reasonable default)
+AC_SUBST(CPP_DEBUG_FLAGS)
+: ${CPP_DEBUG_FLAGS='-g'}
+
+AC_SUBST(FPP_DEBUG_FLAGS)
+: ${FPP_DEBUG_FLAGS='-g'}
+
 AC_SUBST(C_DEBUG_FLAGS)
 : ${C_DEBUG_FLAGS='-g'}
 
@@ -324,6 +336,12 @@
 # (using '-O2' as a reasonable default)
 AC_SUBST(OPTIMISE_MODE)
 
+AC_SUBST(CPP_OPTIMISE_FLAGS)
+: ${CPP_OPTIMISE_FLAGS=''}
+
+AC_SUBST(FPP_OPTIMISE_FLAGS)
+: ${FPP_OPTIMISE_FLAGS=''}
+
 AC_SUBST(C_OPTIMISE_FLAGS)
 : ${C_OPTIMISE_FLAGS='-O2'}
 
@@ -345,6 +363,12 @@
 # (using '-pg' as a reasonable default)
 AC_SUBST(PROFILE_MODE)
 
+AC_SUBST(CPP_PROFILE_FLAGS)
+: ${CPP_PROFILE_FLAGS=''}
+
+AC_SUBST(FPP_PROFILE_FLAGS)
+: ${FPP_PROFILE_FLAGS=''}
+
 AC_SUBST(C_PROFILE_FLAGS)
 : ${C_PROFILE_FLAGS='-pg'}
 
@@ -361,6 +385,8 @@
 # Set the OpenMP flags if they've not been set by now
 AC_SUBST(OPENMP_MODE)
 
+AC_SUBST(CPP_OPENMP_FLAGS)
+AC_SUBST(FPP_OPENMP_FLAGS)
 AC_SUBST(C_OPENMP_FLAGS)
 AC_SUBST(CXX_OPENMP_FLAGS)
 AC_SUBST(F77_OPENMP_FLAGS)
Index: make/make.config.defn.in
===================================================================
--- make/make.config.defn.in	(revision 4655)
+++ make/make.config.defn.in	(working copy)
@@ -61,30 +61,40 @@
 export LIBLINK_PREFIX = @LIBLINK_PREFIX@
 
 # Debug flags
+export CPP_DEBUG_FLAGS = @CPP_DEBUG_FLAGS@
+export FPP_DEBUG_FLAGS = @FPP_DEBUG_FLAGS@
 export C_DEBUG_FLAGS   = @C_DEBUG_FLAGS@
 export CXX_DEBUG_FLAGS = @CXX_DEBUG_FLAGS@
 export F90_DEBUG_FLAGS = @F90_DEBUG_FLAGS@
 export F77_DEBUG_FLAGS = @F77_DEBUG_FLAGS@
 
 # Optimisation flags
+export CPP_OPTIMISE_FLAGS = @CPP_OPTIMISE_FLAGS@
+export FPP_OPTIMISE_FLAGS = @FPP_OPTIMISE_FLAGS@
 export C_OPTIMISE_FLAGS   = @C_OPTIMISE_FLAGS@
 export CXX_OPTIMISE_FLAGS = @CXX_OPTIMISE_FLAGS@
 export F90_OPTIMISE_FLAGS = @F90_OPTIMISE_FLAGS@
 export F77_OPTIMISE_FLAGS = @F77_OPTIMISE_FLAGS@
 
 # Profiling flags
+export CPP_PROFILE_FLAGS = @CPP_PROFILE_FLAGS@
+export FPP_PROFILE_FLAGS = @FPP_PROFILE_FLAGS@
 export C_PROFILE_FLAGS   = @C_PROFILE_FLAGS@
 export CXX_PROFILE_FLAGS = @CXX_PROFILE_FLAGS@
 export F90_PROFILE_FLAGS = @F90_PROFILE_FLAGS@
 export F77_PROFILE_FLAGS = @F77_PROFILE_FLAGS@
 
 # OpenMP flags
+export CPP_OPENMP_FLAGS = @CPP_OPENMP_FLAGS@
+export FPP_OPENMP_FLAGS = @FPP_OPENMP_FLAGS@
 export C_OPENMP_FLAGS   = @C_OPENMP_FLAGS@
 export CXX_OPENMP_FLAGS = @CXX_OPENMP_FLAGS@
 export F90_OPENMP_FLAGS = @F90_OPENMP_FLAGS@
 export F77_OPENMP_FLAGS = @F77_OPENMP_FLAGS@
 
 # Warning flags
+export CPP_WARN_FLAGS = @CPP_WARN_FLAGS@
+export FPP_WARN_FLAGS = @FPP_WARN_FLAGS@
 export C_WARN_FLAGS   = @C_WARN_FLAGS@
 export CXX_WARN_FLAGS = @CXX_WARN_FLAGS@
 export F90_WARN_FLAGS = @F90_WARN_FLAGS@
@@ -200,6 +210,8 @@
 export CCTK_DEBUG_MODE = @DEBUG_MODE@
 
 ifeq ($(strip $(CCTK_DEBUG_MODE)),yes)
+  CPPFLAGS += $(CPP_DEBUG_FLAGS)
+  FPPFLAGS += $(FPP_DEBUG_FLAGS)
   CFLAGS   += $(C_DEBUG_FLAGS)
   CXXFLAGS += $(CXX_DEBUG_FLAGS)
   F77FLAGS += $(F77_DEBUG_FLAGS)
@@ -210,6 +222,8 @@
 CCTK_OPTIMISE_MODE = @OPTIMISE_MODE@
 
 ifeq ($(strip $(CCTK_OPTIMISE_MODE)),yes)
+  CPPFLAGS += $(CPP_OPTIMISE_FLAGS)
+  FPPFLAGS += $(FPP_OPTIMISE_FLAGS)
   CFLAGS   += $(C_OPTIMISE_FLAGS)
   CXXFLAGS += $(CXX_OPTIMISE_FLAGS)
   F77FLAGS += $(F77_OPTIMISE_FLAGS)
@@ -220,6 +234,8 @@
 CCTK_PROFILE_MODE = @PROFILE_MODE@
 
 ifeq ($(strip $(CCTK_PROFILE_MODE)),yes)
+  CPPFLAGS += $(CPP_PROFILE_FLAGS)
+  FPPFLAGS += $(FPP_PROFILE_FLAGS)
   CFLAGS   += $(C_PROFILE_FLAGS)
   CXXFLAGS += $(CXX_PROFILE_FLAGS)
   F77FLAGS += $(F77_PROFILE_FLAGS)
@@ -230,6 +246,8 @@
 CCTK_OPENMP_MODE = @OPENMP_MODE@
 
 ifeq ($(strip $(CCTK_OPENMP_MODE)),yes)
+  CPPFLAGS += $(CPP_OPENMP_FLAGS)
+  FPPFLAGS += $(FPP_OPENMP_FLAGS)
   CFLAGS   += $(C_OPENMP_FLAGS)
   CXXFLAGS += $(CXX_OPENMP_FLAGS)
   F77FLAGS += $(F77_OPENMP_FLAGS)
@@ -246,6 +264,8 @@
 endif
 
 ifeq ($(strip $(CCTK_WARN_MODE)),yes)
+  CPPFLAGS += $(CPP_WARN_FLAGS)
+  FPPFLAGS += $(FPP_WARN_FLAGS)
   CFLAGS   += $(C_WARN_FLAGS)
   CXXFLAGS += $(CXX_WARN_FLAGS)
   F77FLAGS += $(F77_WARN_FLAGS)
