Index: src/piraha/Call.cc
===================================================================
--- src/piraha/Call.cc	(revision 5066)
+++ src/piraha/Call.cc	(working copy)
@@ -872,8 +872,8 @@
                         check_types(thorn.c_str(),aexpr->line(),smv,data->type);
                     }
                     set_function(
-                            strdup(key.c_str()),
-                            strdup(val.c_str()),
+                            key.c_str(),
+                            val.c_str(),
                             gr->group(0)->line());
                 } else {
                     smart_ptr<Group> arr = gr->group("array");
@@ -891,8 +891,8 @@
                             check_types(thorn.c_str(),aexpr->line(),smv,data->type);
                         }
                         set_function(
-                                strdup(keyi.str().c_str()),
-                                strdup(val.c_str()),
+                                keyi.str().c_str(),
+                                val.c_str(),
                                 aexpr->line());
                     }
                 }
