Index: ScheduleCreater.c
===================================================================
--- ScheduleCreater.c	(revision 4808)
+++ ScheduleCreater.c	(working copy)
@@ -790,7 +790,14 @@
     printf("\n");
 #endif
 
-    CCTKi_ScheduleAddRow(group->n_scheditems, array, order, item, thisorders);
+    errcode = CCTKi_ScheduleAddRow(group->n_scheditems, array, order, item, thisorders);
+    if(errcode)
+    {
+      CCTK_VWarn(CCTK_WARN_ABORT, __LINE__, __FILE__, "Cactus",
+      "Adding item %s to group %s failed due to a circular dependency on %s.\n",
+      group->scheditems[item].name, group->name,
+      group->scheditems[-errcode-1].name );  /*NOTREACHED*/
+    }
 
     /* Clear the array for the next item. */
     for(i=0; i < group->n_scheditems; i++)
@@ -1047,9 +1054,30 @@
 
 #ifdef TEST_SCHEDULECREATOR
 
+#include <stdarg.h>
+
 #define func_x(x) \
 int func_ ## x (void) { return printf("I'm func " #x "\n"); }
 
+int CCTK_VWarn(int level, int line, const char *file, 
+    const char *thorn, const char *fmt, ...)
+{
+  va_list ap;
+
+  va_start(ap, fmt);
+
+  fprintf(stdout, "WARNING[L%d,P0] (%s) in line %d of %s: ", level, thorn, line, file);
+  vfprintf(stdout, fmt, ap);
+  fputc('\n', stdout);
+
+  va_end(ap);
+
+  if(level == 0)
+    exit(1);
+
+  return 0;
+}
+
 func_x(a)
 func_x(b)
 func_x(c)
Index: ScheduleSorter.c
===================================================================
--- ScheduleSorter.c	(revision 4808)
+++ ScheduleSorter.c	(working copy)
@@ -166,7 +166,8 @@
 
    @returntype int
    @returndesc 
-   0 - success
+   0  - success
+   <0 - -error_column
    @endreturndesc
 @@*/
 int CCTKi_ScheduleAddRow(int size, 
@@ -190,7 +191,11 @@
   {
     if(thisorders[column])
     {
-      if(array[row][column] && array[row][column] != (signed char)thisorders[column]) retval--;
+      if(array[row][column] && array[row][column] != (signed char)thisorders[column])
+      {
+        retval = -(1+column);
+        break;
+      }
       array[row][column] = (signed char)(  thisorders[column]);
       array[column][row] = (signed char)( -thisorders[column]);
     }
