From 6160341f02efce4b9c1e50b101e7684605a1c84a Mon Sep 17 00:00:00 2001
From: Roland Haas <roland.haas@physics.gatech.edu>
Date: Mon, 25 Jun 2012 10:41:30 -0700
Subject: [PATCH 1/2] make anonymous checkout the default

---
 GetComponents |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/GetComponents b/GetComponents
index 79ae294..065f806 100755
--- a/GetComponents
+++ b/GetComponents
@@ -78,7 +78,7 @@ my $DEBUG              = 0;
 my $HELP               = 0;
 my $MAN                = 0;
 my $EXPERIMENTAL       = 0;
-my $ANONYMOUS_CHECKOUT = 0;
+my $ANONYMOUS_CHECKOUT = 1;
 my $DO_UPDATES         = 0;
 my $STATUS             = 0;
 my $DIFF               = 0;
@@ -138,7 +138,7 @@ GetOptions(
     'help|?'               => \$HELP,
     'man'                  => \$MAN,
     'debug'                => \$DEBUG,
-    'anonymous'            => \$ANONYMOUS_CHECKOUT,
+    'anonymous!'           => \$ANONYMOUS_CHECKOUT,
     'update'               => \$DO_UPDATES,
     'root=s'               => \$ROOT,
     'date=s'               => \$DATE,
-- 
1.7.10

