Index: src/piraha/Call.cc
===================================================================
--- src/piraha/Call.cc	(revision 5068)
+++ src/piraha/Call.cc	(working copy)
@@ -424,6 +424,56 @@
                 val->ddata = val->idata;
                 return val;
             }
+        } else if(val->type == PIR_STRING) {
+            if(fn == "int") {
+                val->type = PIR_INT;
+                std::istringstream buf(val->sdata);
+                if(!(buf >> val->idata)) {
+                    std::ostringstream msg;
+                    msg << "Invalid numerical value: " << val->sdata << std::endl;
+                    std::string par = get_parfile();
+                    CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
+                }
+                std::string extra;
+                if(buf >> extra) {
+                    std::ostringstream msg;
+                    msg << "Trailing input in numerical value: " << val->sdata << std::endl;
+                    std::string par = get_parfile();
+                    CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
+                }
+                return val;
+            } else if(fn == "real") {
+                val->type = PIR_REAL;
+                std::istringstream buf(val->sdata);
+                if(!(buf >> val->ddata)) {
+                    std::ostringstream msg;
+                    msg << "Invalid numerical value: " << val->sdata << std::endl;
+                    std::string par = get_parfile();
+                    CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
+                }
+                std::string extra;
+                if(buf >> extra) {
+                    std::ostringstream msg;
+                    msg << "Trailing input in numerical value: " << val->sdata << std::endl;
+                    std::string par = get_parfile();
+                    CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
+                }
+                return val;
+            } else if(fn == "bool") {
+                val->type = PIR_BOOL;
+                std::string s = mklower(val->sdata);
+                if(s == "no" || s == "false") {
+                    ret->idata = 0;
+                } else if(s == "yes" || s == "true") {
+                    ret->idata = 1;
+                } else {
+                    std::ostringstream msg;
+                    msg << "Invalid boolean value: " << val->sdata << std::endl;
+                    std::string par = get_parfile();
+                    CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
+                }
+                return val;
+            }
         }
         std::ostringstream msg;
         msg << "Unknown func: " << fn << "(" << val->type << ")" << std::endl;
