Index: arrangements/CactusArchive/ADM/test/test_ADM_1.par
===================================================================
--- arrangements/CactusArchive/ADM/test/test_ADM_1.par	(revision 384)
+++ arrangements/CactusArchive/ADM/test/test_ADM_1.par	(working copy)
@@ -12,7 +12,7 @@
 # @@*/           
 
 # Required thorns
-ActiveThorns = "ADMCoupling Boundary Time ADMconstraints SpaceMask pugh pughreduce CartGrid3D CoordBase SymBase admbase ADM admanalysis staticconformal coordgauge idanalyticbh ioutil iobasic ADMMacros localreduce"
+ActiveThorns = "ADMCoupling Boundary Time ADMconstraints SpaceMask pugh pughreduce CartGrid3D CoordBase SymBase admbase tmunubase ADM admanalysis staticconformal coordgauge idanalyticbh ioutil iobasic ADMMacros localreduce"
 
 # GENERAL 
 
Index: arrangements/CactusArchive/ADM/test/test_ADM_2.par
===================================================================
--- arrangements/CactusArchive/ADM/test/test_ADM_2.par	(revision 384)
+++ arrangements/CactusArchive/ADM/test/test_ADM_2.par	(working copy)
@@ -1,7 +1,7 @@
 !DESC "Kleing 1+log sliced black hole, ADM, double leapfrog"
 
 #Required thorns
-ActiveThorns = "ADMCoupling Boundary Time ADMBase ADMConstraints SpaceMask StaticConformal CoordGauge ADMMacros CartGrid3D CoordBase SymBase pugh pughslab pughreduce ADM IDanalyticBH IOASCII IOUtil iobasic admanalysis localreduce"
+ActiveThorns = "ADMCoupling Boundary Time TmunuBase ADMBase ADMConstraints SpaceMask StaticConformal CoordGauge ADMMacros CartGrid3D CoordBase SymBase pugh pughslab pughreduce ADM IDanalyticBH IOASCII IOUtil iobasic admanalysis localreduce"
 
 # GENERAL 
 
Index: arrangements/CactusArchive/ADM/test/test_ADM_3.par
===================================================================
--- arrangements/CactusArchive/ADM/test/test_ADM_3.par	(revision 384)
+++ arrangements/CactusArchive/ADM/test/test_ADM_3.par	(working copy)
@@ -1,7 +1,7 @@
 
 !DESC "1+log sliced black hole, ADM,  iterative CN"
 
-ActiveThorns = "ADMCoupling Boundary Time ADMConstraints SpaceMask PUGH pughreduce CartGrid3D CoordBase SymBase ADMBase StaticConformal ADMAnalysis CoordGauge ADM IDAnalyticBH IOUtil IOBasic ADMMacros localreduce"
+ActiveThorns = "ADMCoupling Boundary Time ADMConstraints SpaceMask PUGH pughreduce CartGrid3D CoordBase SymBase TmunuBase ADMBase StaticConformal ADMAnalysis CoordGauge ADM IDAnalyticBH IOUtil IOBasic ADMMacros localreduce"
 
 admbase::evolution_method  = "ADM"     
 driver::global_nsize        = 20     
Index: arrangements/CactusArchive/ADM/test/test_ADM_4.par
===================================================================
--- arrangements/CactusArchive/ADM/test/test_ADM_4.par	(revision 384)
+++ arrangements/CactusArchive/ADM/test/test_ADM_4.par	(working copy)
@@ -12,7 +12,7 @@
 # @@*/           
 
 # Required thorns
-ActiveThorns = "ADMCoupling Boundary time ADMconstraints SpaceMask pugh pughreduce CartGrid3D CoordBase SymBase ADMBase ADM ADMMacros StaticConformal CoordGauge ADMAnalysis idanalyticbh ioutil iobasic localreduce"
+ActiveThorns = "ADMCoupling Boundary time ADMconstraints SpaceMask pugh pughreduce CartGrid3D CoordBase SymBase TmunuBase ADMBase ADM ADMMacros StaticConformal CoordGauge ADMAnalysis idanalyticbh ioutil iobasic localreduce"
 
 # GENERAL 
 
Index: arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Shift.par
===================================================================
--- arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Shift.par	(revision 153)
+++ arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Shift.par	(working copy)
@@ -12,7 +12,7 @@
 # @@*/           
 
 # Required thorns
-ActiveThorns = "ADMCoupling Boundary Time ADMconstraints ADM CoordGauge pugh pughreduce CartGrid3D CoordBase SymBase ADMBase StaticConformal admmacros spacemask idanalyticbh ioutil iobasic localreduce"
+ActiveThorns = "ADMCoupling Boundary Time ADMconstraints ADM CoordGauge pugh pughreduce CartGrid3D CoordBase SymBase TmunuBase ADMBase StaticConformal admmacros spacemask idanalyticbh ioutil iobasic localreduce"
 
 # GENERAL
 
Index: arrangements/EinsteinAnalysis/ADMConstraints/test/test_ADM.par
===================================================================
--- arrangements/EinsteinAnalysis/ADMConstraints/test/test_ADM.par	(revision 153)
+++ arrangements/EinsteinAnalysis/ADMConstraints/test/test_ADM.par	(working copy)
@@ -12,7 +12,7 @@
 # @@*/           
 
 # Required thorns
-ActiveThorns = "ADMCoupling Boundary Time ADMconstraints pugh pughreduce CartGrid3D CoordBase SymBase ADMBase StaticConformal admmacros coordgauge spacemask ADM idanalyticbh ioutil iobasic localreduce"
+ActiveThorns = "ADMCoupling Boundary Time ADMconstraints pugh pughreduce CartGrid3D CoordBase SymBase TmunuBase ADMBase StaticConformal admmacros coordgauge spacemask ADM idanalyticbh ioutil iobasic localreduce"
 
 # GENERAL
 
Index: arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Mask.par
===================================================================
--- arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Mask.par	(revision 153)
+++ arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Mask.par	(working copy)
@@ -12,7 +12,7 @@
 # @@*/           
 
 # Required thorns
-ActiveThorns = "ADMCoupling Boundary Time ADMconstraints pugh ADM CoordGauge pughreduce CartGrid3D CoordBase SymBase ADMBase StaticConformal admmacros spacemask idanalyticbh ioutil iobasic localreduce"
+ActiveThorns = "ADMCoupling Boundary Time ADMconstraints pugh ADM CoordGauge pughreduce CartGrid3D CoordBase SymBase TmunuBase ADMBase StaticConformal admmacros spacemask idanalyticbh ioutil iobasic localreduce"
 
 # GENERAL
 
Index: arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Physical.par
===================================================================
--- arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Physical.par	(revision 153)
+++ arrangements/EinsteinAnalysis/ADMConstraints/test/ADMConstraints_Physical.par	(working copy)
@@ -12,7 +12,7 @@
 # @@*/           
 
 # Required thorns
-ActiveThorns = "ADMCoupling Boundary Time ADMconstraints ADM CoordGauge pugh pughreduce CartGrid3D CoordBase SymBase ADMBase StaticConformal admmacros spacemask idanalyticbh ioutil iobasic localreduce"
+ActiveThorns = "ADMCoupling Boundary Time ADMconstraints ADM CoordGauge pugh pughreduce CartGrid3D CoordBase SymBase TmunuBase ADMBase StaticConformal admmacros spacemask idanalyticbh ioutil iobasic localreduce"
 
 # GENERAL
 
Index: arrangements/EinsteinInitialData/DistortedBHIVP/test/E2xeon_test_dbh.par
===================================================================
--- arrangements/EinsteinInitialData/DistortedBHIVP/test/E2xeon_test_dbh.par	(revision 59)
+++ arrangements/EinsteinInitialData/DistortedBHIVP/test/E2xeon_test_dbh.par	(working copy)
@@ -2,7 +2,7 @@
 #!DESC "Initial data for axisymmetric distored black hole"
 ######################################################################
 
-ActiveThorns = "coordgauge admanalysis admbase admmacros staticconformal time boundary ADMCoupling ADMconstraints pugh pughreduce pughslab pughinterp LocalInterp SymBase CoordBase cartgrid3d SpaceMask DistortedBHIVP ioascii iobasic ioutil LocalReduce"
+ActiveThorns = "coordgauge admanalysis tmunubase admbase admmacros staticconformal time boundary ADMCoupling ADMconstraints pugh pughreduce pughslab pughinterp LocalInterp SymBase CoordBase cartgrid3d SpaceMask DistortedBHIVP ioascii iobasic ioutil LocalReduce"
 
 # GENERAL 
 
Index: arrangements/EinsteinInitialData/DistortedBHIVP/test/test_dbh.par
===================================================================
--- arrangements/EinsteinInitialData/DistortedBHIVP/test/test_dbh.par	(revision 59)
+++ arrangements/EinsteinInitialData/DistortedBHIVP/test/test_dbh.par	(working copy)
@@ -2,7 +2,7 @@
 # single rotating black hole
 #==============================================================================
 
-ActiveThorns = " DistortedBHIVP adm coordbase coordgauge staticconformal spacemask admbase admcoupling admmacros admanalysis ellbase cartgrid3d time pugh pughreduce pughslab ioascii ioutil iobasic adm admconstraints symbase boundary pughinterp LocalInterp flexio nanchecker LocalReduce"
+ActiveThorns = " DistortedBHIVP adm coordbase coordgauge staticconformal spacemask tmunubase admbase admcoupling admmacros admanalysis ellbase cartgrid3d time pugh pughreduce pughslab ioascii ioutil iobasic adm admconstraints symbase boundary pughinterp LocalInterp flexio nanchecker LocalReduce"
 IOBasic::outScalar_vars = " admbase::metric admbase::curv admbase::lapse admbase::shift admconstraints::momy admconstraints::ham "
 IOBasic::outInfo_vars = " admconstraints::ham admbase::alp "
 IOASCII::out1D_vars = " admbase::metric admbase::curv admbase::lapse admbase::shift admconstraints::momy admconstraints::ham "
Index: arrangements/EinsteinInitialData/Exact/test/de_Sitter.par
===================================================================
--- arrangements/EinsteinInitialData/Exact/test/de_Sitter.par	(revision 287)
+++ arrangements/EinsteinInitialData/Exact/test/de_Sitter.par	(working copy)
@@ -3,7 +3,7 @@
 # $Header$
 
 # Required thorns
-ActiveThorns = "CartGrid3D CoordBase SymBase Boundary Time   PUGH PUGHReduce PUGHSlab   IOAscii IOUtil IOBasic LocalReduce   ADMBase ADMMacros ADMConstraints ADMCoupling StaticConformal CoordGauge SpaceMask  ReflectionSymmetry Exact"
+ActiveThorns = "CartGrid3D CoordBase SymBase Boundary Time   PUGH PUGHReduce PUGHSlab   IOAscii IOUtil IOBasic LocalReduce  TmunuBase  ADMBase ADMMacros ADMConstraints ADMCoupling StaticConformal CoordGauge SpaceMask  ReflectionSymmetry Exact"
 
 # Grid
 driver::global_nx = 10
Index: arrangements/EinsteinInitialData/IDAxiBrillBH/test/test_axibrill_nostagger.par
===================================================================
--- arrangements/EinsteinInitialData/IDAxiBrillBH/test/test_axibrill_nostagger.par	(revision 122)
+++ arrangements/EinsteinInitialData/IDAxiBrillBH/test/test_axibrill_nostagger.par	(working copy)
@@ -2,7 +2,7 @@
 !DESC "Initial data for axisymmetric distored black hole"
 ######################################################################
 
-ActiveThorns = "LocalInterp ADMCoupling Boundary time iobasic ADMconstraints pugh pughslab pughreduce CartGrid3D CoordBase SymBase ADMBase StaticConformal ADM admmacros coordgauge ADMAnalysis IDAxiBrillBH ioascii ioutil spacemask localreduce"
+ActiveThorns = "LocalInterp ADMCoupling Boundary time iobasic ADMconstraints pugh pughslab pughreduce CartGrid3D CoordBase SymBase TmunuBase ADMBase StaticConformal ADM admmacros coordgauge ADMAnalysis IDAxiBrillBH ioascii ioutil spacemask localreduce"
 
 #pugh::enable_all_storage = "yes"
 #cactus::cctk_show_banners = 0
Index: arrangements/EinsteinInitialData/IDAxiBrillBH/test/test_axibrill.par
===================================================================
--- arrangements/EinsteinInitialData/IDAxiBrillBH/test/test_axibrill.par	(revision 122)
+++ arrangements/EinsteinInitialData/IDAxiBrillBH/test/test_axibrill.par	(working copy)
@@ -2,7 +2,7 @@
 !DESC "Initial data for axisymmetric distored black hole"
 ######################################################################
 
-ActiveThorns = "LocalInterp ADMCoupling Boundary time iobasic ADMconstraints pugh pughslab pughreduce CartGrid3D CoordBase SymBase ADMBase StaticConformal ADM admmacros coordgauge ADMAnalysis IDAxiBrillBH ioascii ioutil spacemask localreduce"
+ActiveThorns = "LocalInterp ADMCoupling Boundary time iobasic ADMconstraints pugh pughslab pughreduce CartGrid3D CoordBase SymBase TmunuBase ADMBase StaticConformal ADM admmacros coordgauge ADMAnalysis IDAxiBrillBH ioascii ioutil spacemask localreduce"
 
 #pugh::enable_all_storage = "yes"
 #cactus::cctk_show_banners = 0
Index: arrangements/EinsteinInitialData/IDAxiOddBrillBH/test/E2xeon_test_axioddbh.par
===================================================================
--- arrangements/EinsteinInitialData/IDAxiOddBrillBH/test/E2xeon_test_axioddbh.par	(revision 66)
+++ arrangements/EinsteinInitialData/IDAxiOddBrillBH/test/E2xeon_test_axioddbh.par	(working copy)
@@ -2,7 +2,7 @@
 #!DESC "Initial data for axisymmetric distored black hole"
 ######################################################################
 
-ActiveThorns = "coordgauge admanalysis admbase admmacros staticconformal time boundary ADMconstraints pugh pughreduce pughslab pughinterp cartgrid3d ADM IDAxiOddBrillBH ioascii iobasic ioutil LocalReduce ADMCoupling SpaceMask CoordBase SymBase LocalInterp"
+ActiveThorns = "coordgauge admanalysis tmunubase admbase admmacros staticconformal time boundary ADMconstraints pugh pughreduce pughslab pughinterp cartgrid3d ADM IDAxiOddBrillBH ioascii iobasic ioutil LocalReduce ADMCoupling SpaceMask CoordBase SymBase LocalInterp"
 
 # GENERAL 
 
Index: arrangements/EinsteinInitialData/IDAxiOddBrillBH/test/test_axioddbh.par
===================================================================
--- arrangements/EinsteinInitialData/IDAxiOddBrillBH/test/test_axioddbh.par	(revision 66)
+++ arrangements/EinsteinInitialData/IDAxiOddBrillBH/test/test_axioddbh.par	(working copy)
@@ -2,7 +2,7 @@
 # single rotating black hole
 #==============================================================================
 
-ActiveThorns = " IDAxiOddBrillBH adm coordbase coordgauge staticconformal spacemask admbase admcoupling admmacros admanalysis ellbase cartgrid3d time pugh pughreduce pughslab ioascii ioutil iobasic adm admconstraints boundary pughinterp LocalInterp nanchecker LocalInterp LocalReduce SymBase"
+ActiveThorns = " IDAxiOddBrillBH adm coordbase coordgauge staticconformal spacemask tmunubase admbase admcoupling admmacros admanalysis ellbase cartgrid3d time pugh pughreduce pughslab ioascii ioutil iobasic adm admconstraints boundary pughinterp LocalInterp nanchecker LocalInterp LocalReduce SymBase"
 IOBasic::outScalar_vars = " admbase::metric admbase::curv admbase::lapse admbase::shift admconstraints::momy admconstraints::ham "
 IOBasic::outInfo_vars = " admconstraints::ham admbase::alp "
 IOASCII::out1D_vars = " admbase::metric admbase::curv admbase::lapse admbase::shift admconstraints::momy admconstraints::ham "
Index: arrangements/EinsteinInitialData/IDConstraintViolate/test/ConstraintViolate.par
===================================================================
--- arrangements/EinsteinInitialData/IDConstraintViolate/test/ConstraintViolate.par	(revision 23)
+++ arrangements/EinsteinInitialData/IDConstraintViolate/test/ConstraintViolate.par	(working copy)
@@ -1,7 +1,7 @@
 
 !DESC "Constraint violating Gaussian initial data"
 
-ActiveThorns = "SpaceMask Boundary ADMBase ADM ADMMacros ADMCoupling ADMConstraints CoordGauge StaticConformal CartGrid3D CoordBase SymBase IOBasic IOUtil IOASCII IDConstraintViolate Pugh PughReduce PughSlab Time LocalReduce"
+ActiveThorns = "SpaceMask Boundary TmunuBase ADMBase ADM ADMMacros ADMCoupling ADMConstraints CoordGauge StaticConformal CartGrid3D CoordBase SymBase IOBasic IOUtil IOASCII IDConstraintViolate Pugh PughReduce PughSlab Time LocalReduce"
  
 #------------------------------------------------------------------------------
 
Index: arrangements/EinsteinInitialData/RotatingDBHIVP/test/E2xeon_test_rdbh.par
===================================================================
--- arrangements/EinsteinInitialData/RotatingDBHIVP/test/E2xeon_test_rdbh.par	(revision 54)
+++ arrangements/EinsteinInitialData/RotatingDBHIVP/test/E2xeon_test_rdbh.par	(working copy)
@@ -2,7 +2,7 @@
 #!DESC "Initial data for axisymmetric distored black hole"
 ######################################################################
 
-ActiveThorns = "coordgauge admanalysis admbase admmacros staticconformal time boundary ellbase ADMconstraints pugh pughinterp pughreduce pughslab cartgrid3d ADM RotatingDBHIVP ioascii iobasic ioutil ADMCoupling CoordBase SpaceMask LocalReduce LocalInterp SymBase"
+ActiveThorns = "coordgauge admanalysis tmunubase admbase admmacros staticconformal time boundary ellbase ADMconstraints pugh pughinterp pughreduce pughslab cartgrid3d ADM RotatingDBHIVP ioascii iobasic ioutil ADMCoupling CoordBase SpaceMask LocalReduce LocalInterp SymBase"
 
 # GENERAL 
 
Index: arrangements/EinsteinInitialData/RotatingDBHIVP/test/test_rdbh.par
===================================================================
--- arrangements/EinsteinInitialData/RotatingDBHIVP/test/test_rdbh.par	(revision 54)
+++ arrangements/EinsteinInitialData/RotatingDBHIVP/test/test_rdbh.par	(working copy)
@@ -2,7 +2,7 @@
 # single rotating black hole
 #==============================================================================
 
-ActiveThorns = " adm RotatingDBHIVP coordbase coordgauge staticconformal spacemask admbase admcoupling admmacros admanalysis ellbase cartgrid3d time pugh pughreduce pughslab ioascii ioutil iobasic adm admconstraints symbase boundary pughinterp LocalInterp nanchecker LocalReduce"
+ActiveThorns = " adm RotatingDBHIVP coordbase coordgauge staticconformal spacemask tmunubase admbase admcoupling admmacros admanalysis ellbase cartgrid3d time pugh pughreduce pughslab ioascii ioutil iobasic adm admconstraints symbase boundary pughinterp LocalInterp nanchecker LocalReduce"
 IOBasic::outScalar_vars = " admbase::metric admbase::curv admbase::lapse admbase::shift admconstraints::momy admconstraints::ham "
 IOBasic::outInfo_vars = " admconstraints::ham admbase::alp "
 IOASCII::out1D_vars = " admbase::metric admbase::curv admbase::lapse admbase::shift admconstraints::momy admconstraints::ham "
Index: arrangements/EinsteinInitialData/TwoPunctures/test/bam.par
===================================================================
--- arrangements/EinsteinInitialData/TwoPunctures/test/bam.par	(revision 129)
+++ arrangements/EinsteinInitialData/TwoPunctures/test/bam.par	(working copy)
@@ -3,7 +3,7 @@
 ActiveThorns = "Boundary CartGrid3D Time CoordBase SymBase LocalReduce"
 ActiveThorns = "PUGH PUGHReduce PUGHSlab"
 ActiveThorns = "IOASCII IOBasic IOUtil LocalReduce"
-ActiveThorns = "ADMBase StaticConformal"
+ActiveThorns = "TmunuBase ADMBase StaticConformal"
 ActiveThorns = "ADMConstraints ADMCoupling ADMMacros SpaceMask"
 ActiveThorns = "BAM_Elliptic EllBase"
 
Index: arrangements/EinsteinInitialData/TwoPunctures/test/idpuncture.par
===================================================================
--- arrangements/EinsteinInitialData/TwoPunctures/test/idpuncture.par	(revision 129)
+++ arrangements/EinsteinInitialData/TwoPunctures/test/idpuncture.par	(working copy)
@@ -3,7 +3,7 @@
 ActiveThorns = "Boundary CartGrid3D Time CoordBase SymBase"
 ActiveThorns = "PUGH PUGHReduce PUGHSlab"
 ActiveThorns = "IOASCII IOBasic IOUtil"
-ActiveThorns = "ADMBase StaticConformal"
+ActiveThorns = "TmunuBase ADMBase StaticConformal"
 ActiveThorns = "ADMConstraints ADMCoupling ADMMacros SpaceMask"
 ActiveThorns = "IDPuncture TATelliptic TATJacobi TATMG TATPETSc"
 
Index: arrangements/EinsteinInitialData/TwoPunctures/test/twopunctures.par
===================================================================
--- arrangements/EinsteinInitialData/TwoPunctures/test/twopunctures.par	(revision 129)
+++ arrangements/EinsteinInitialData/TwoPunctures/test/twopunctures.par	(working copy)
@@ -4,7 +4,7 @@
 ActiveThorns = "PUGH PUGHReduce PUGHSlab LocalReduce"
 ActiveThorns = "IOASCII IOBasic IOUtil"
 ActiveThorns = "ADMBase StaticConformal"
-ActiveThorns = "ADMConstraints ADMCoupling ADMMacros SpaceMask"
+ActiveThorns = "TmunuBase ADMConstraints ADMCoupling ADMMacros SpaceMask"
 ActiveThorns = "TwoPunctures"
 
 Cactus::cctk_itlast		= 0
Index: arrangements/PITTNullCode/NullInterp/test/news_test_c.par
===================================================================
--- arrangements/PITTNullCode/NullInterp/test/news_test_c.par	(revision 8)
+++ arrangements/PITTNullCode/NullInterp/test/news_test_c.par	(working copy)
@@ -5,7 +5,7 @@
 
 ActiveThorns = "SymBase CoordBase CartGrid3D IOASCII IOBasic IOUtil Pugh PughReduce PughSlab PUGHInterp Time LocalReduce"
 ActiveThorns = "Fortran MoL SpaceMask"
-ActiveThorns = "ADMBase"
+ActiveThorns = "TmunuBase ADMBase"
 ActiveThorns = "HarmBase HarmGauge HarmExact HarmEvol HarmIData HarmExactEvol HarmBdry"
 ActiveThorns = "Boundary ADMCoupling ADMMacros ADMConstraints StaticConformal Harm2ADM"
 ActiveThorns = "AEILocalInterp WT_Extraction NullDecomp NullGrid NullInterp NullVars NullEvolve NullNews"
