From 546ad01bd3453780ce3c9f4414ed71ca5690b5d4 Mon Sep 17 00:00:00 2001
From: Tanja Bode <tanja.bode@physics.gatech.edu>
Date: Fri, 8 Mar 2013 12:52:23 -0500
Subject: [PATCH 07/12] GRHydro_InitData: Allow n_p=0 again for poloidal
 magnetic field

---
 param.ccl |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/param.ccl b/param.ccl
index 9b741b2..77f88eb 100644
--- a/param.ccl
+++ b/param.ccl
@@ -424,7 +424,7 @@ CCTK_REAL poloidal_A_b  "Vector potential strength"
 
 CCTK_INT poloidal_n_p  "Vector potential strength"
 {
-  (0:*  :: "Any positive integer."
+  0:*  :: "Any positive integer."
 } 3
 
 CCTK_REAL poloidal_P_cut  "Pressure used to confine the B field inside a star"
-- 
1.7.10.4

