Index: src/Output.c
===================================================================
--- src/Output.c	(revision 45)
+++ src/Output.c	(working copy)
@@ -309,9 +309,14 @@
         }
         else
         {
+          const char *name = CCTK_TimerName(i);
+          
+          if (name == NULL)
+            name = "(null)";
+
           CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
                      "Clock \"%s\" not found for timer #%d \"%s\"",
-                     all_timers_clock, i, CCTK_TimerName(i));
+                     all_timers_clock, i, name);
           timer_secs = -1;
         }
       }
@@ -673,10 +678,16 @@
   }
   
   /* Determine local timer names and their values */
-  char my_timernames[my_ntimers][TIMERNAME_LENGTH];
+  char *my_timernames = malloc(my_ntimers*TIMERNAME_LENGTH); /* these arrays can be too large for the stack */
+  assert(my_timernames);
   for (int n=0; n<my_ntimers; ++n) {
-    strncpy(my_timernames[n], CCTK_TimerName(n), TIMERNAME_LENGTH-1);
-    my_timernames[n][TIMERNAME_LENGTH-1] = '\0';
+    const char *name = CCTK_TimerName(n);
+    
+    if (name == NULL)
+      snprintf(&my_timernames[n*TIMERNAME_LENGTH], TIMERNAME_LENGTH-1, "DESTROYED TIMER %5d", n);
+    else
+      strncpy(&my_timernames[n*TIMERNAME_LENGTH], name, TIMERNAME_LENGTH-1);
+    my_timernames[n*TIMERNAME_LENGTH+TIMERNAME_LENGTH-1] = '\0';
   }
   double my_timervalues[my_ntimers];
   {
@@ -685,9 +696,14 @@
       CCTK_TimerI(n, td);
       cTimerVal const *const tv = CCTK_GetClockValue(all_timers_clock, td);
       if (!tv) {
+        const char *name = CCTK_TimerName(n);
+    
+        if (name == NULL)
+          name = "(null)";
+
         CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
                    "Clock \"%s\" not found for timer #%d \"%s\"",
-                   all_timers_clock, n, CCTK_TimerName(n));
+                   all_timers_clock, n, name);
         my_timervalues[n] = -1;
       } else {
         my_timervalues[n] = CCTK_TimerClockSeconds(tv);
@@ -697,7 +713,8 @@
   }
   
   /* Gather timer names and values from each process */
-  char all_timernames[total_ntimers][TIMERNAME_LENGTH];
+  char *all_timernames = malloc(total_ntimers*TIMERNAME_LENGTH); /* these arrays can likely be too large for the stack */
+  assert(all_timernames);
   double all_timervalues[total_ntimers];
   int name_displacements[nprocs], value_displacements[nprocs];
   int name_counts[nprocs];
@@ -729,6 +746,8 @@
     timers->secs_avg = NULL;
     timers->secs_min = NULL;
     timers->secs_max = NULL;
+    free(all_timernames);
+    free(my_timernames);
     
     return 1;
   }
@@ -741,15 +760,16 @@
     sort_index[i] = i;
   }
   assert(!compare_string_array);
-  compare_string_array = all_timernames;
+  /* the funny type is: pointer to array of char of length TIMERNAME_LENGTH */
+  compare_string_array = (char (*)[TIMERNAME_LENGTH])all_timernames;
   qsort(sort_index, total_ntimers, sizeof *sort_index, compare_string);
   compare_string_array = NULL;
   int unique_timers = 0;
   if (total_ntimers > 0) {
     unique_timers = 1;        /* first timer is always unique */
     for (int i=1; i<total_ntimers; ++i) {
-      if (strcmp(all_timernames[sort_index[i]],
-                 all_timernames[sort_index[unique_timers-1]]) != 0)
+      if (strcmp(&all_timernames[sort_index[i]*TIMERNAME_LENGTH],
+                 &all_timernames[sort_index[unique_timers-1]*TIMERNAME_LENGTH]) != 0)
       {
         sort_index[unique_timers++] = sort_index[i];
       }
@@ -770,7 +790,7 @@
   assert(timers->ntimers==0 || timers->secs_max);
   
   for (int n=0; n<unique_timers; ++n) {
-    timers->names[n] = strdup(all_timernames[sort_index[n]]);
+    timers->names[n] = strdup(&all_timernames[sort_index[n]*TIMERNAME_LENGTH]);
   }
   
   /* Reduce timer values */
@@ -786,7 +806,7 @@
       /* TODO: sort, then use bsearch */
       int i;
       for (i=0; i<all_ntimers[p]; ++i) {
-        if (strcmp(timers->names[n], all_timernames[name_offset+i]) == 0) break;
+        if (strcmp(timers->names[n], &all_timernames[(name_offset+i)*TIMERNAME_LENGTH]) == 0) break;
       }
       if (i < all_ntimers[p]) {
         /* Found the timer */
@@ -810,6 +830,9 @@
       timers->secs_max[n] = maxval;
     }
   }
+
+  free(all_timernames);
+  free(my_timernames);
   
   return 1;
 }
