Index: src/nuc_eos/bisection.F90
===================================================================
--- src/nuc_eos/bisection.F90	(revision 68)
+++ src/nuc_eos/bisection.F90	(working copy)
@@ -31,8 +31,8 @@
   ltmin=logtemp(1)
 
   lt = lt0
-  lt1 = dlog10(min(10.0d0**ltmax,1.10d0*(10.0d0**lt0)))
-  lt2 = dlog10(max(10.0d0**ltmin,0.90d0*(10.0d0**lt0)))
+  lt1 = min(ltmax,lt0+0.05d0)
+  lt2 = max(ltmin,lt0-0.05d0)
 
   call findthis(lr,lt1,y,f1a,bivar,d1,d2,d3)
   call findthis(lr,lt2,y,f2a,bivar,d1,d2,d3)
@@ -43,35 +43,35 @@
   keyerrt=0
   
 
-  do while(f1*f2.ge.0.0d0)
+  do while(f1*f2>=0.0d0)
      bcount=bcount+1
-     lt1=dlog10(min(10.0d0**ltmax,1.2d0*(10.0d0**lt1)))
-     lt2=dlog10(max(10.0d0**ltmin,0.8d0*(10.0d0**lt2)))
+     lt1=min(ltmax,lt1+0.1d0)
+     lt2=max(ltmin,lt2-0.1d0)
      call findthis(lr,lt1,y,f1a,bivar,d1,d2,d3)
      call findthis(lr,lt2,y,f2a,bivar,d1,d2,d3)
      f1=f1a-eps0
      f2=f2a-eps0
-     if(bcount.ge.maxbcount) then
+     if(bcount>=maxbcount) then
         keyerrt=667
         return
      endif
   enddo
 
-  if(f1.lt.0.0d0) then
+  if(f1<0.0d0) then
      lt=lt1
-     dlt=dlog10((10.0D0**lt2)-(10.0d0**lt1))
+     dlt=lt2-lt1
   else
      lt=lt2
-     dlt=dlog10((10.0D0**lt1)-(10.0d0**lt2))
+     dlt=lt1-lt2
   endif
 
   do i=1,itmax
-     dlt=dlog10((10.0d0**dlt)*0.5d0)
-     ltmid=dlog10(10.0d0**lt+10.0d0**dlt)
+     dlt=dlt*0.5d0
+     ltmid=lt+dlt
      call findthis(lr,ltmid,y,f2a,bivar,d1,d2,d3)
      fmid=f2a-eps0
-     if(fmid.le.0.0d0) lt=ltmid
-     if(abs(1.0d0-f2a/eps0).lt.tol) then
+     if(fmid<=0.0d0) lt=ltmid
+     if(abs(1.0d0-f2a/eps0)<tol) then
         lt=ltmid
         return
      endif
@@ -113,8 +113,8 @@
   lrmin=logrho(1)
 
   lr = lr0
-  lr1 = dlog10(min(10.0d0**lrmax,1.10d0*(10.0d0**lr0)))
-  lr2 = dlog10(max(10.0d0**lrmin,0.90d0*(10.0d0**lr0)))
+  lr1 = min(lrmax,lr0+0.05d0)
+  lr2 = max(lrmin,lr0-0.05d0)
 
   call findthis(lr1,lt,y,f1a,bivar,d1,d2,d3)
   call findthis(lr2,lt,y,f2a,bivar,d1,d2,d3)
@@ -124,35 +124,35 @@
 
   keyerrr=0
   
-  do while(f1*f2.ge.0.0d0)
+  do while(f1*f2>=0.0d0)
      bcount=bcount+1
-     lr1=dlog10(min(10.0d0**lrmax,1.2d0*(10.0d0**lr1)))
-     lr2=dlog10(max(10.0d0**lrmin,0.8d0*(10.0d0**lr2)))
+     lr1=min(lrmax,lr1+0.1d0)
+     lr2=max(lrmin,lr2-0.1d0)
      call findthis(lr1,lt,y,f1a,bivar,d1,d2,d3)
      call findthis(lr2,lt,y,f2a,bivar,d1,d2,d3)
      f1=f1a-lpressin
      f2=f2a-lpressin
-     if(bcount.ge.maxbcount) then
+     if(bcount>=maxbcount) then
         keyerrr=667
         return
      endif
   enddo
 
-  if(f1.lt.0.0d0) then
+  if(f1<0.0d0) then
      lr=lr1
-     dlr=dlog10((10.0D0**lr2)-(10.0d0**lr1))
+     dlr=lr2-lr1
   else
      lr=lr2
-     dlr=dlog10((10.0D0**lr1)-(10.0d0**lr2))
+     dlr=lr1-lr2
   endif
 
   do i=1,itmax
-     dlr=dlog10((10.0d0**dlr)*0.5d0)
-     lrmid=dlog10(10.0d0**lr+10.0d0**dlr)
+     dlr=dlr*0.5d0
+     lrmid=lr+dlr
      call findthis(lrmid,lt,y,f2a,bivar,d1,d2,d3)
      fmid=f2a-lpressin
-     if(fmid.le.0.0d0) lr=lrmid
-     if(abs(1.0d0-f2a/lpressin).lt.tol) then
+     if(fmid<=0.0d0) lr=lrmid
+     if(abs(1.0d0-f2a/lpressin)<tol) then
         lr=lrmid
         return
      endif
