From 56d7b09e3cd185a6bae1ba161ef72655aadf399c Mon Sep 17 00:00:00 2001
From: Ian Hinder <ian.hinder@aei.mpg.de>
Date: Sat, 27 Aug 2011 13:30:55 +0200
Subject: [PATCH] Add dylib extension to autodetect on Mac OS

---
 OpenSSL.sh |   21 +++++++++++++--------
 1 files changed, 13 insertions(+), 8 deletions(-)

diff --git a/OpenSSL.sh b/OpenSSL.sh
index b0a5fb9..6f001ce 100644
--- a/OpenSSL.sh
+++ b/OpenSSL.sh
@@ -18,14 +18,19 @@ if [ -z "${OPENSSL_DIR}" ]; then
     echo "BEGIN MESSAGE"
     echo "OpenSSL selected, but OPENSSL_DIR not set. Checking some places..."
     echo "END MESSAGE"
-    
-    FILES="include/openssl/ssl.h lib/libssl.a lib/libcrypto.a"
-    DIRS="/usr /usr/local /opt/local"
-    for dir in $DIRS; do
-        OPENSSL_DIR="$dir"
-        for file in $FILES; do
-            if [ ! -r "$dir/$file" ]; then
-                unset OPENSSL_DIR
+
+    for libext in a dylib; do
+        FILES="include/openssl/ssl.h lib/libssl.${libext} lib/libcrypto.${libext}"
+        DIRS="/usr /usr/local /opt/local"
+        for dir in $DIRS; do
+            OPENSSL_DIR="$dir"
+            for file in $FILES; do
+                if [ ! -r "$dir/$file" ]; then
+                    unset OPENSSL_DIR
+                    break
+                fi
+            done
+            if [ -n "$OPENSSL_DIR" ]; then
                 break
             fi
         done
-- 
1.7.5.4

