From 5a8d40c7367dc5df16c6d4eca36ef6131bde43cf Mon Sep 17 00:00:00 2001
From: Roland Haas <roland.haas@physics.gatech.edu>
Date: Sun, 22 Apr 2012 20:07:50 -0700
Subject: [PATCH 05/17] GRHydro_InitData: REQUIRE EOS_Omni

---
 configuration.ccl |    2 ++
 1 file changed, 2 insertions(+)
 create mode 100644 configuration.ccl

diff --git a/configuration.ccl b/configuration.ccl
new file mode 100644
index 0000000..15805b5
--- /dev/null
+++ b/configuration.ccl
@@ -0,0 +1,2 @@
+REQUIRES EOS_Omni
+
-- 
1.7.10.4

