Index: sbin/CreateConfigurationBindings.pl
===================================================================
--- sbin/CreateConfigurationBindings.pl	(revision 4701)
+++ sbin/CreateConfigurationBindings.pl	(working copy)
@@ -136,6 +136,23 @@
     }
   }
 
+  # turn optional capabilities into required capabilities, if the
+  # capability is provided
+  foreach $thorn (sort keys %thorns)
+  {
+    if ($cfg->{"\U$thorn\E OPTIONAL"})
+    {
+      foreach $cap (split (' ', $cfg->{"\U$thorn\E OPTIONAL"}))
+      {
+        # TODO: prevent matches of partial words
+        if ($providedcaplist =~ m/$cap/i)
+        {
+          $cfg->{"\U$thorn\E REQUIRES"} .= "$cap ";
+        }
+      }
+    }
+  }
+
   # here we add the files to the thorns that require capabilities
   foreach $thorn (sort keys %thorns)
   {
@@ -159,19 +176,19 @@
       }
     }
 
-   if ($cfg->{"\U$thorn\E OPTIONAL"})
-    {
-      foreach $providedcap (split (' ', $cfg->{"\U$thorn\E OPTIONAL"}))
-      {
-        if ($providedcaplist =~ m/$providedcap/i)
-        {
-          $defs .= $providedcap . " = 1\n";
-          $defs .= "include $bindings_dir/Configuration/Capabilities/make.\U$providedcap\E.defn\n";
-          $incs .= "#include \"../Capabilities/cctki_\U$providedcap\E.h\"\n";
-          $deps .= "include $bindings_dir/Configuration/Capabilities/make.\U$providedcap\E.deps\n";
-        }
-      }
-    }
+#   if ($cfg->{"\U$thorn\E OPTIONAL"})
+#    {
+#      foreach $providedcap (split (' ', $cfg->{"\U$thorn\E OPTIONAL"}))
+#      {
+#        if ($providedcaplist =~ m/$providedcap/i)
+#        {
+#          $defs .= $providedcap . " = 1\n";
+#          $defs .= "include $bindings_dir/Configuration/Capabilities/make.\U$providedcap\E.defn\n";
+#          $incs .= "#include \"../Capabilities/cctki_\U$providedcap\E.h\"\n";
+#          $deps .= "include $bindings_dir/Configuration/Capabilities/make.\U$providedcap\E.deps\n";
+#        }
+#      }
+#    }
 
     if ($cfg->{"\U$thorn\E REQUIRES"} || $cfg->{"\U$thorn\E OPTIONAL"})
     {
Index: sbin/CST
===================================================================
--- sbin/CST	(revision 4701)
+++ sbin/CST	(working copy)
@@ -331,14 +331,15 @@
 
   foreach $thorn (sort keys %$thorns)
   {
-    if($config->{"\U$thorn\E OPTIONS"} ne 'NO_SOURCE' &&
-      $thorn ne 'Cactus')
+    next if $thorn eq 'Cactus';
+
+    if($config->{"\U$thorn\E OPTIONS"} ne 'NO_SOURCE')
     {
       $thorns->{$thorn} =~ m:.*/(.*/.*):;
       $thornlist .= " $1";
     }
 
-    if( -r "$thorns->{$thorn}/configuration.ccl")
+    if(-r "$thorns->{$thorn}/configuration.ccl")
     {
       $thorns->{$thorn} =~ m:.*/(.*/.*):;
       $config_thornlist .= " $1";
Index: sbin/ConfigurationParser.pl
===================================================================
--- sbin/ConfigurationParser.pl	(revision 4701)
+++ sbin/ConfigurationParser.pl	(working copy)
@@ -42,6 +42,8 @@
 #        if ($cfg{"\U$thorn\E PROVIDES"});
 #      print "           Requires: ", $cfg{"\U$thorn\E REQUIRES"}, "\n"
 #        if ($cfg{"\U$thorn\E REQUIRES"});
+#      print "           Optional: ", $cfg{"\U$thorn\E OPTIONAL"}, "\n"
+#        if ($cfg{"\U$thorn\E OPTIONAL"});
 #    }
 
     $cfg{"\U$thorn\E USES THORNS"} = '';
Index: make/make.configuration
===================================================================
--- make/make.configuration	(revision 4701)
+++ make/make.configuration	(working copy)
@@ -212,7 +212,7 @@
 
 rebuild: $(CONFIG)/make.thornlist
 
-$(CONFIG)/make.thornlist: $(TOP)/ThornList $(foreach lib, $(CCTK_HOME)/src $(THORNS:%=$(PACKAGE_DIR)/%), $(lib)/param.ccl $(lib)/interface.ccl $(lib)/schedule.ccl) $(CONFIG_THORNS:%=$(PACKAGE_DIR)/%/configuration.ccl) $(CCTK_HOME)/lib/make/force-rebuild
+$(CONFIG)/make.thornlist: $(TOP)/ThornList $(foreach lib, $(CCTK_HOME)/src $(THORNS:%=$(PACKAGE_DIR)/%), $(lib)/param.ccl $(lib)/interface.ccl $(lib)/schedule.ccl) $(CCTK_HOME)/src/configuration.ccl $(CONFIG_THORNS:%=$(PACKAGE_DIR)/%/configuration.ccl) $(CCTK_HOME)/lib/make/force-rebuild
 	if [ -r $@ ] ; then echo Reconfiguring thorns ; rm $@ ;\
 	else echo Processing CCL files ; fi
 	+$(PERL) -s $(CST) -config_dir=$(CONFIG) -cctk_home=$(CCTK_HOME) -top=$(TOP) $<
@@ -244,7 +244,7 @@
 endif
 
 # Catch the special case of a thorn being removed from the disk entirely.
-$(foreach lib, $(CCTK_HOME)/src $(THORNS:%=$(PACKAGE_DIR)/%), $(lib)/param.ccl $(lib)/interface.ccl $(lib)/schedule.ccl) $(CONFIG_THORNS:%=$(PACKAGE_DIR)/%/configuration.ccl):
+$(foreach lib, $(CCTK_HOME)/src $(THORNS:%=$(PACKAGE_DIR)/%), $(lib)/param.ccl $(lib)/interface.ccl $(lib)/schedule.ccl) $(CCTK_HOME)/src/configuration.ccl $(CONFIG_THORNS:%=$(PACKAGE_DIR)/%/configuration.ccl):
 	@echo "Missing file $@"
 	@echo "Possibly from a thorn deleted from an arrangement"
 	@echo "Deleting make.thornlist"
