Index: sim
===================================================================
--- sim	(revision 1314)
+++ sim	(working copy)
@@ -1085,7 +1085,7 @@
     # seem to be required, but is probably useful).
     # This expects the shell variable $SHELL to be set on the remote
     # machine.
-    my $cmd = "cd $path && \$SHELL -l";
+    my $cmd = "cd $path || echo \"could not change to directory $path\"; \$SHELL -l";
     
     # ssh to target
     $cmd = get_sshcmd $machine, $cmd, '-t';
