From bc674c2e01f790e3eae292dc77a484081041803b Mon Sep 17 00:00:00 2001
From: Bruno Coutinho Mundim <bcmsma@astro.rit.edu>
Date: Sun, 21 Oct 2012 12:20:32 -0500
Subject: [PATCH 12/13] GRHydro_InitData: fix typo in error message.

---
 src/CheckParam.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/CheckParam.c b/src/CheckParam.c
index e90abcf..d3a40df 100644
--- a/src/CheckParam.c
+++ b/src/CheckParam.c
@@ -62,7 +62,7 @@ void GRHydro_InitData_CheckParameters(CCTK_ARGUMENTS)
   if((CCTK_Equals(initial_hydro,"magnetized_bondi_solution") || CCTK_Equals(initial_hydro,"magnetized_bondi_solution_iso")) &&
      !CCTK_Equals(initial_Bvec, "magnetized Bondi") )
     {
-      CCTK_PARAMWARN("Please set initial_Bec='magnetized Bondi' to properly initialize the magnetic field for Bondi solutions!");
+      CCTK_PARAMWARN("Please set initial_Bvec='magnetized Bondi' to properly initialize the magnetic field for Bondi solutions!");
     }
 
   if( CCTK_Equals(initial_hydro,"magnetized_bondi_solution") || CCTK_Equals(initial_hydro,"hydro_bondi_solution")  ) { 
-- 
1.7.10.4

