Index: src/util/Misc.c
===================================================================
--- src/util/Misc.c	(revision 4828)
+++ src/util/Misc.c	(working copy)
@@ -19,8 +19,6 @@
 #include <math.h>
 #include <float.h>
 
-#include "cctk_GNU.h"
-
 #include "cctk_Flesh.h"
 #include "cctk_Misc.h"
 #include "cctk_FortranString.h"
@@ -40,10 +38,6 @@
  ********************* Local Routine Prototypes *********************
  ********************************************************************/
 
-int CCTK_RegexMatch(const char *string,
-                    const char *pattern,
-                    const int nmatch,
-                    regmatch_t *pmatch);
 int CCTK_SetStringInRegexList(char **data, const char *value,
                               int n_elements, ...);
 void CCTK_PrintString(const char *data);
@@ -367,6 +361,7 @@
   regmatch_t pmatch[8];
   int start_closed, end_closed;
   int start, end, step;
+  int matched;
 
   /* Valid ranges are of the form start:end:step
    * possibly preceeded by a [ or ( and ended by a ) or ] to indicate
@@ -407,7 +402,8 @@
 
 #undef R_MAYBE
 
-  if(CCTK_RegexMatch(range, pattern, 8, pmatch) != 0)
+  matched = CCTK_RegexMatch(range, pattern, 8, pmatch);
+  if(matched > 0)
   {
     /* First work out if the range is closed at the lower end. */
     if(pmatch[1].rm_so != -1)
@@ -528,10 +524,15 @@
     }
 
   }
-  else
+  else if(!matched)
   {
     CCTK_Warn(1, __LINE__, __FILE__, "Flesh", "Invalid range");
   }
+  else
+  {
+    CCTK_VWarn(0, __LINE__, __FILE__, "Flesh",
+               "Invalid patten '%s' used to parse range", pattern);
+  }
 
   return retval;
 }
@@ -609,7 +610,7 @@
 
 #undef R_MAYBE
 
-  if(CCTK_RegexMatch(range, pattern, 8, pmatch) != 0)
+  if(CCTK_RegexMatch(range, pattern, 8, pmatch) > 0)
   {
     /* First work out if the range is closed at the lower end. */
     if(pmatch[1].rm_so != -1)
@@ -1170,7 +1171,7 @@
   {
     element = va_arg(ap, char *);
 
-    if(CCTK_RegexMatch(value, element, 0, NULL))
+    if(CCTK_RegexMatch(value, element, 0, NULL) > 0)
     {
       retval = CCTK_SetString(data, value);
       break;
@@ -1312,6 +1313,7 @@
    @returndesc
    1 - pattern matches
    0 - pattern doesn't match
+   <0 - pattern is invalid
    @endreturndesc
 @@*/
 
@@ -1347,7 +1349,7 @@
     }
     else
     {
-      retval = 0;
+      retval = -1;       /* repost error */
     }
   }
 
Index: src/include/cctk_Misc.h
===================================================================
--- src/include/cctk_Misc.h	(revision 4828)
+++ src/include/cctk_Misc.h	(working copy)
@@ -13,6 +13,8 @@
 #ifndef _CCTK_MISC_H_
 #define _CCTK_MISC_H_
 
+#include "cctk_GNU.h"
+
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -33,6 +35,11 @@
 int Util_IntInRangeList(int inval, int n_elements, ...);
 int Util_DoubleInRangeList(double inval, int n_elements, ...);
 
+int CCTK_RegexMatch (const char *string,
+                     const char *pattern,
+                     const int nmatch,
+                     regmatch_t *pmatch);
+
 int CCTK_SetDoubleInRangeList(CCTK_REAL *data, const char *value, 
                               int n_elements, ...);
 int CCTK_SetIntInRangeList(CCTK_INT *data, const char *value, 
Index: src/main/Parameters.c
===================================================================
--- src/main/Parameters.c	(revision 4828)
+++ src/main/Parameters.c	(working copy)
@@ -203,11 +203,6 @@
  ********************* Other Routine Prototypes *********************
  ********************************************************************/
 
-int CCTK_RegexMatch (const char *string,
-                     const char *pattern,
-                     const int nmatch,
-                     regmatch_t *pmatch);
-
 extern void CCTKi_SetParameterSetMask (int mask);
 
 /********************************************************************
@@ -2107,13 +2102,20 @@
         CCTK_Equals (param->props->thorn, range->origin))
     {
 #ifndef CCTK_PARAMUNCHECKED
-      if (CCTK_RegexMatch (value, range->range, 0, NULL))
+      const int matched = CCTK_RegexMatch (value, range->range, 0, NULL);
+      if (matched > 0)
       {
 #endif
         retval = CCTK_SetString (param->data, value);
         break;
 #ifndef CCTK_PARAMUNCHECKED
       }
+      else if (matched < 0)
+      {
+        CCTK_VWarn(CCTK_WARN_ALERT, __LINE__, __FILE__, "Cactus",
+                   "Invalid regular expression '%s' used as range for string %s::%s",
+                   range->range, param->props->thorn, param->props->name);
+      }
 #endif
     }
   }
@@ -2151,13 +2153,20 @@
         CCTK_Equals (param->props->thorn, range->origin))
     {
 #ifndef CCTK_PARAMUNCHECKED
-      if (CCTK_RegexMatch (value, range->range, 0, NULL))
+      const int matched = CCTK_RegexMatch (value, range->range, 0, NULL);
+      if (matched > 0)
       {
 #endif
         retval = CCTK_SetString (param->data, value);
         break;
 #ifndef CCTK_PARAMUNCHECKED
       }
+      else if (matched < 0)
+      {
+        CCTK_VWarn(CCTK_WARN_ALERT, __LINE__, __FILE__, "Cactus",
+                   "Invalid regular expression '%s' used as range for sequence %s::%s",
+                   range->range, param->props->thorn, param->props->name);
+      }
 #endif
     }
   }
Index: doc/ReferenceManual/CCTKReference.tex
===================================================================
--- doc/ReferenceManual/CCTKReference.tex	(revision 4828)
+++ doc/ReferenceManual/CCTKReference.tex	(working copy)
@@ -659,6 +659,9 @@
 \item[\code{CCTK\_ReductionHandle}] [\pageref{CCTK-ReductionHandle}]
   Get the handle for a registered reduction operator
 
+\item[\code{CCTK\_RegexMatch}] [\pageref{CCTK-RegexMatch}]
+  Perform a regular expression match of string against pattern
+
 \item[\code{CCTK\_RegisterBanner}] [\pageref{CCTK-RegisterBanner}]
   Register a banner for a thorn
 
@@ -10647,6 +10650,72 @@
 \end{ExampleSection}
 \end{FunctionDescription}
 
+% Misc.c
+\begin{FunctionDescription}{CCTK\_RegexMatch}{Perform a regular expression match of string against pattern}
+\label{CCTK-RegexMatch}
+\begin{SynopsisSection}
+\begin{Synopsis}{C}
+\begin{verbatim}success = CCTK_RegexMatch( const char *string, const char *pattern,
+                          const int nmatch, regmatch_t *pmatch)\end{verbatim}
+\end{Synopsis}
+\end{SynopsisSection}
+\begin{ParameterSection}
+\begin{Parameter}{string}
+String to match against
+\end{Parameter}
+\begin{Parameter}{pattern}
+Regex pattern
+\end{Parameter}
+\begin{Parameter}{nmatch}
+The size of the pmatch array
+\end{Parameter}
+\begin{Parameter}{pmatch}
+Array in which to place the matches
+\end{Parameter}
+\end{ParameterSection}
+\begin{ResultSection}
+\begin{Result}{0} pattern does not match\end{Result}
+\begin{Result}{1} pattern matches\end{Result}
+\begin{Result}{< 0} indicates an error condition (pattern did not compile as a
+    regular expression) \end{Result}
+\end{ResultSection}
+\begin{Discussion}
+Perform a regular expression match of string against pattern.
+Also returns the specified number of matched substrings as
+give by regexec.
+This is a modified form of the example routine given in the SGI
+man page for regcomp.
+\end{Discussion}
+\begin{ExampleSection}
+\begin{Example}{C}
+\begin{verbatim}
+#define R_BEGIN "(\\[|\\()?"
+#define R_VALUE "([^]):]*)"
+#define R_SEP   ":"
+#define R_END   "(\\]|\\))?"
+#define R_MAYBE(x) "(" x ")?"
+
+  int matched;
+  const char pattern[] =
+    R_BEGIN
+    R_VALUE
+    R_MAYBE(R_SEP R_VALUE R_MAYBE(R_SEP R_VALUE))
+    R_END;
+
+  if( (matched = CCTK_RegexMatch(range, pattern, 8, pmatch)) > 0) {
+      CCTK_VInfo(CCTK_THORNSTRING, "'%s' is a valid range specifier",
+                 range);
+  } else if(!matched) {
+      CCTK_VInfo(CCTK_THORNSTRING, "'%s' is not a valid range specifier",
+                 range);
+  } else {
+      CCTK_VInfo(CCTK_THORNSTRING, "invalid pattern '%s'", pattern);
+  }
+\end{verbatim}
+\end{Example}
+\end{ExampleSection}
+\end{FunctionDescription}
+
 % Coord.c
 \begin{FunctionDescription}{CCTK\_RegisterBanner}{Register a banner for a thorn}
 \label{CCTK-RegisterBanner}
Index: doc/ReferenceManual.pdf
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/pdf
