Index: sim.py
===================================================================
--- sim.py	(revision 1234)
+++ sim.py	(working copy)
@@ -137,5 +137,6 @@
 
     CommandDispatch()
 
-simenv.init(base=BASE_PATH, callingExecutable=os.path.join(BASE_PATH, "bin/sim"), usageString=buildUsageString(), optionGroups=optionGroups)
-main()
+if __name__ == '__main__':
+    simenv.init(base=BASE_PATH, callingExecutable=os.path.join(BASE_PATH, "bin/sim"), usageString=buildUsageString(), optionGroups=optionGroups)
+    main()
