Index: bin/RunTestSuite
===================================================================
--- bin/RunTestSuite	(revision 2099)
+++ bin/RunTestSuite	(working copy)
@@ -16,6 +16,6 @@
 # Do not produce core files
 ulimit -c 0
 
-MAKE=$(@SOURCEDIR@/simfactory/bin/sim print-mdb-entry @MACHINE@ | grep '^make' | sed -e 's/^.*= *//;s/ *#.*$//')
+MAKE=$(@SOURCEDIR@/simfactory/bin/sim print-mdb-entry @MACHINE@ | grep '^make' | sed -e 's/^make *= *//;s/ *#.*$//')
 
 ${MAKE} @CONFIGURATION@-testsuite PROMPT=no
