Index: configuration.ccl
===================================================================
--- configuration.ccl	(revision 2)
+++ configuration.ccl	(working copy)
@@ -4,7 +4,7 @@
 {
   SCRIPT configure.sh
   LANG bash
-  OPTIONS HWLOC_DIR
+  OPTIONS HWLOC_DIR HWLOC_INSTALL_DIR
 }
 
 REQUIRES hwloc
Index: configure.sh
===================================================================
--- configure.sh	(revision 2)
+++ configure.sh	(working copy)
@@ -5,40 +5,111 @@
 ################################################################################
 
 # Set up shell
-set -x                          # Output commands
+if [ "$(echo ${VERBOSE} | tr '[:upper:]' '[:lower:]')" = 'yes' ]; then
+    set -x                          # Output commands
+fi
 set -e                          # Abort on errors
 
 
+################################################################################
+# Search
+################################################################################
 
+if [ -z "${HWLOC_DIR}" ]; then
+    echo "BEGIN MESSAGE"
+    echo "hwloc selected, but HWLOC_DIR not set. Checking some places..."
+    echo "END MESSAGE"
+
+    DIRS="/usr /usr/local /usr/local/packages /usr/local/apps /opt/local ${HOME} c:/packages"
+    for dir in $DIRS; do
+        # libraries might have different file extensions
+        for libext in a so dylib; do
+            # libraries can be in /lib or /lib64
+            for libdir in lib64 lib; do
+                FILES="include/hwloc.h $libdir/libhwloc.$libext"
+                # assume this is the one and check all needed files
+                HWLOC_DIR="$dir"
+                for file in $FILES; do
+                    # discard this directory if one file was not found
+                    if [ ! -r "$dir/$file" ]; then
+                        unset HWLOC_DIR
+                        break
+                    fi
+                done
+                # don't look further if all files have been found
+                if [ -n "$HWLOC_DIR" ]; then
+                    break
+                fi
+            done
+            # don't look further if all files have been found
+            if [ -n "$HWLOC_DIR" ]; then
+                break
+            fi
+        done
+        # don't look further if all files have been found
+        if [ -n "$HWLOC_DIR" ]; then
+            break
+        fi
+    done
+
+    if [ -z "$HWLOC_DIR" ]; then
+        echo "BEGIN MESSAGE"
+        echo "hwloc not found"
+        echo "END MESSAGE"
+    else
+        echo "BEGIN MESSAGE"
+        echo "Found hwloc in ${HWLOC_DIR}"
+        echo "END MESSAGE"
+    fi
+fi
+
 ################################################################################
 # Build
 ################################################################################
 
-echo "BEGIN MESSAGE"
-echo "Building hwloc..."
-echo "END MESSAGE"
+if [ -z "${HWLOC_DIR}"                                                   \
+     -o "$(echo "${HWLOC_DIR}" | tr '[a-z]' '[A-Z]')" = 'BUILD' ]
+then
+    echo "BEGIN MESSAGE"
+    echo "Building hwloc..."
+    echo "END MESSAGE"
 
-# Set locations
-THORN=hwloc
-NAME=hwloc-1.4
-SRCDIR=$(dirname $0)
-BUILD_DIR=${SCRATCH_BUILD}/build/${THORN}
-INSTALL_DIR=${SCRATCH_BUILD}/external/${THORN}
-DONE_FILE=${SCRATCH_BUILD}/done/${THORN}
-HWLOC_DIR=${INSTALL_DIR}
+    # Set locations
+    THORN=hwloc
+    NAME=hwloc-1.4
+    SRCDIR=$(dirname $0)
+    BUILD_DIR=${SCRATCH_BUILD}/build/${THORN}
+    if [ -z "${HWLOC_INSTALL_DIR}" ]; then
+        INSTALL_DIR=${SCRATCH_BUILD}/external/${THORN}
+    else
+        echo "BEGIN MESSAGE"
+        echo "Installing hwloc into ${HWLOC_INSTALL_DIR}"
+        echo "END MESSAGE"
+        INSTALL_DIR=${HWLOC_INSTALL_DIR}
+    fi
+    DONE_FILE=${SCRATCH_BUILD}/done/${THORN}
+    HWLOC_DIR=${INSTALL_DIR}
 
-(
-    exec >&2                    # Redirect stdout to stderr
-    set -x                      # Output commands
-    set -e                      # Abort on errors
-    cd ${SCRATCH_BUILD}
     if [ -e ${DONE_FILE} -a ${DONE_FILE} -nt ${SRCDIR}/dist/${NAME}.tar.gz \
                          -a ${DONE_FILE} -nt ${SRCDIR}/configure.sh ]
     then
-        echo "hwloc: The enclosed hwloc library has already been built; doing nothing"
+        echo "BEGIN MESSAGE"
+        echo "hwloc has already been built; doing nothing"
+        echo "END MESSAGE"
     else
-        echo "hwloc: Building enclosed hwloc library"
-        
+        echo "BEGIN MESSAGE"
+        echo "Building hwloc"
+        echo "END MESSAGE"
+
+        # Build in a subshell
+        (
+        exec >&2                    # Redirect stdout to stderr
+        if [ "$(echo ${VERBOSE} | tr '[:upper:]' '[:lower:]')" = 'yes' ]; then
+            set -x                  # Output commands
+        fi
+        set -e                      # Abort on errors
+        cd ${SCRATCH_BUILD}
+
         export LDFLAGS
         unset LIBS
         if echo '' ${ARFLAGS} | grep 64 > /dev/null 2>&1; then
@@ -74,18 +145,19 @@
         
         date > ${DONE_FILE}
         echo "hwloc: Done."
+
+        )
+        if (( $? )); then
+            echo 'BEGIN ERROR'
+            echo 'Error while building hwloc. Aborting.'
+            echo 'END ERROR'
+            exit 1
+        fi
     fi
-)
 
-if (( $? )); then
-    echo 'BEGIN ERROR'
-    echo 'Error while building hwloc. Aborting.'
-    echo 'END ERROR'
-    exit 1
 fi
 
 
-
 #HWLOC_INC_DIRS='$(HWLOC_DIR)/include '"${BUILD_DIR}/${NAME}/include"
 HWLOC_INC_DIRS='$(HWLOC_DIR)/include'
 HWLOC_LIB_DIRS='$(HWLOC_DIR)/lib'
