Index: include/cctk.h
===================================================================
--- include/cctk.h	(revision 4721)
+++ include/cctk.h	(working copy)
@@ -78,8 +78,10 @@
 #define CCTK_ORIGIN_SPACE(x) (cctk_origin_space(x)+cctk_delta_space(x)/cctk_levfac(x)*cctk_levoff(x)/cctk_levoffdenom(x))
 #define CCTK_DELTA_SPACE(x) (cctk_delta_space(x)/cctk_levfac(x))
 #define CCTK_DELTA_TIME (cctk_delta_time/cctk_timefac)
-#define CCTK_LSSH(stag,dim) cctk_lssh((stag)+CCTK_NSTAGGER*(dim))
-#define CCTK_LSSH_IDX(stag,dim) ((stag)+CCTK_NSTAGGER*(dim))
+/* The "stagger index" stag is zero-based (0,1,...), the direction dim
+   is one-based (1,2,3,...) */
+#define CCTK_LSSH(stag,dim) cctk_lssh(CCTK_LSSH_IDX(stag,dim))
+#define CCTK_LSSH_IDX(stag,dim) (1+(stag)+CCTK_NSTAGGER*((dim)-1))
 
 #ifdef F90CODE
 
@@ -404,7 +406,6 @@
   ((i) + (GH)->cctk_lsh[0] *                    \
    ((j) + (GH)->cctk_lsh[1] *                   \
     ((k) + (GH)->cctk_lsh[2] * (l))))
-
 #define CCTK_VECTGFINDEX0D(GH, n)               \
   (n)
 #define CCTK_VECTGFINDEX1D(GH, i, n)            \
@@ -436,27 +437,27 @@
 #define _DECLARE_CCTK_ARGUMENTS _DECLARE_CCTK_CARGUMENTS
 #define _DECLARE_CCTK_CARGUMENTS \
         CCTK_DECLARE_INIT(ptrdiff_t,cctki_dummy_int,0);\
-        CCTK_DECLARE_INIT(int,cctk_dim,cctkGH->cctk_dim);\
-        CCTK_DECLARE_INIT(int,*cctk_gsh,cctkGH->cctk_gsh);\
-        CCTK_DECLARE_INIT(int,*cctk_lsh,cctkGH->cctk_lsh);\
-        CCTK_DECLARE_INIT(int,*cctk_lbnd,cctkGH->cctk_lbnd);\
-        CCTK_DECLARE_INIT(int,*cctk_ubnd,cctkGH->cctk_ubnd);\
-        CCTK_DECLARE_INIT(int,*cctk_lssh,cctkGH->cctk_lssh);\
-        CCTK_DECLARE_INIT(int,*cctk_from,cctkGH->cctk_from);\
-        CCTK_DECLARE_INIT(int,*cctk_to,cctkGH->cctk_to);\
-        CCTK_DECLARE_INIT(int,*cctk_bbox,cctkGH->cctk_bbox);\
-        CCTK_DECLARE_INIT(CCTK_REAL,cctk_delta_time,cctkGH->cctk_delta_time);\
-        CCTK_DECLARE_INIT(CCTK_REAL,cctk_time,cctkGH->cctk_time);\
-        CCTK_DECLARE_INIT(CCTK_REAL,*cctk_delta_space,cctkGH->cctk_delta_space);\
-        CCTK_DECLARE_INIT(CCTK_REAL,*cctk_origin_space,cctkGH->cctk_origin_space);\
-        CCTK_DECLARE_INIT(int,*cctk_levfac,cctkGH->cctk_levfac);\
-        CCTK_DECLARE_INIT(int,*cctk_levoff,cctkGH->cctk_levoff);\
-        CCTK_DECLARE_INIT(int,*cctk_levoffdenom,cctkGH->cctk_levoffdenom);\
-        CCTK_DECLARE_INIT(int,cctk_timefac,cctkGH->cctk_timefac);\
-        CCTK_DECLARE_INIT(int,cctk_convlevel,cctkGH->cctk_convlevel);\
-        CCTK_DECLARE_INIT(int,cctk_convfac,cctkGH->cctk_convfac);\
-        CCTK_DECLARE_INIT(int,*cctk_nghostzones,cctkGH->cctk_nghostzones);\
-        CCTK_DECLARE_INIT(int,cctk_iteration,cctkGH->cctk_iteration);\
+        CCTK_DECLARE_INIT(int const,cctk_dim,cctkGH->cctk_dim);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_gsh,cctkGH->cctk_gsh);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_lsh,cctkGH->cctk_lsh);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_lbnd,cctkGH->cctk_lbnd);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_ubnd,cctkGH->cctk_ubnd);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_lssh,cctkGH->cctk_lssh);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_from,cctkGH->cctk_from);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_to,cctkGH->cctk_to);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_bbox,cctkGH->cctk_bbox);\
+        CCTK_DECLARE_INIT(CCTK_REAL const,cctk_delta_time,cctkGH->cctk_delta_time);\
+        CCTK_DECLARE_INIT(CCTK_REAL const,cctk_time,cctkGH->cctk_time);\
+        CCTK_DECLARE_INIT(CCTK_REAL const *CCTK_RESTRICT const,cctk_delta_space,cctkGH->cctk_delta_space);\
+        CCTK_DECLARE_INIT(CCTK_REAL const *CCTK_RESTRICT const,cctk_origin_space,cctkGH->cctk_origin_space);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_levfac,cctkGH->cctk_levfac);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_levoff,cctkGH->cctk_levoff);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_levoffdenom,cctkGH->cctk_levoffdenom);\
+        CCTK_DECLARE_INIT(int const,cctk_timefac,cctkGH->cctk_timefac);\
+        CCTK_DECLARE_INIT(int const,cctk_convlevel,cctkGH->cctk_convlevel);\
+        CCTK_DECLARE_INIT(int const,cctk_convfac,cctkGH->cctk_convfac);\
+        CCTK_DECLARE_INIT(int const *CCTK_RESTRICT const,cctk_nghostzones,cctkGH->cctk_nghostzones);\
+        CCTK_DECLARE_INIT(int const,cctk_iteration,cctkGH->cctk_iteration);\
 
 #define _INITIALISE_CCTK_C2F
 #define _DECLARE_CCTK_C2F
@@ -466,9 +467,8 @@
                             (xGH)->cctk_lssh,\
                             (xGH)->cctk_from,(xGH)->cctk_to,\
                             (xGH)->cctk_bbox,\
-                            &((xGH)->cctk_delta_time),\
-                            &((xGH)->cctk_time), (xGH)->cctk_delta_space,\
-                            (xGH)->cctk_origin_space,\
+                            &((xGH)->cctk_delta_time), &((xGH)->cctk_time),\
+                            (xGH)->cctk_delta_space, (xGH)->cctk_origin_space,\
                             (xGH)->cctk_levfac,\
                             (xGH)->cctk_levoff,\
                             (xGH)->cctk_levoffdenom,\
@@ -478,20 +478,23 @@
                             (xGH)->cctk_nghostzones,\
                             &((xGH)->cctk_iteration),\
                             &(xGH)
-#define _CCTK_C2F_PROTO     int *,\
-                            int *,\
-                            int *,int *, int *, int *, int *,int *,int *,\
-                            CCTK_REAL *, CCTK_REAL *, CCTK_REAL *,\
-                            CCTK_REAL *,\
-                            int *,\
-                            int *,\
-                            int *,\
-                            int *,\
-                            int *,\
-                            int *,\
-                            int *,\
-                            int *,\
-                            cGH **
+#define _CCTK_C2F_PROTO     int const *,\
+                            int const *, int const *,\
+                            int const *, int const *,\
+                            int const *,\
+                            int const *, int const *,\
+                            int const *,\
+                            CCTK_REAL const *, CCTK_REAL const *,\
+                            CCTK_REAL const *, CCTK_REAL const *,\
+                            int const *,\
+                            int const *,\
+                            int const *,\
+                            int const *,\
+                            int const *,\
+                            int const *,\
+                            int const *,\
+                            int const *,\
+                            cGH const *const *
 
 #define CCTK_EQUALS(a,b) (CCTK_Equals((a),(b)))
 
