Index: src/CartGrid3D.c
===================================================================
--- src/CartGrid3D.c	(revision 255)
+++ src/CartGrid3D.c	(working copy)
@@ -379,8 +379,8 @@
   /* Register the coordinate ranges */
   for (i = 0; i < 3; i++)
   {
-    cctk_delta_space[i] = this_delta[i];
-    cctk_origin_space[i] = origin[i];
+    cctkGH->cctk_delta_space[i] = this_delta[i];
+    cctkGH->cctk_origin_space[i] = origin[i];
 
     lower[i] = origin[i];
     upper[i] = origin[i] + this_delta[i]*(cctk_gsh[i] - 1);
