Index: configure.sh
===================================================================
--- configure.sh	(revision 42)
+++ configure.sh	(working copy)
@@ -60,6 +60,23 @@
     echo "Using bundled GSL..."
     echo "END MESSAGE"
     
+    # check for required tools. Do this here so that we don't require them when
+    # using the system library
+    if [ x$TAR = x ] ; then
+      echo 'BEGIN ERROR'
+      echo 'Could not find tar command. Please make sure that (gnu) tar is present'
+      echo 'and that the TAR variable is set to its location.'
+      echo 'END ERROR'
+      exit 1
+    fi
+    #if [ x$PATCH = x ] ; then
+    #  echo 'BEGIN ERROR'
+    #  echo 'Could not find patch command. Please make sure that patch is present'
+    #  echo 'and that the PATCH variable is set to its location.'
+    #  echo 'END ERROR'
+    #  exit 1
+    #fi
+
     # Set locations
     THORN=GSL
     NAME=gsl-1.16
