Index: src/piraha/smart_ptr.cc
===================================================================
--- src/piraha/smart_ptr.cc	(revision 5057)
+++ src/piraha/smart_ptr.cc	(working copy)
@@ -2,6 +2,6 @@
 
 namespace cctki_piraha {
 
-std::set<void*> *ptrs = new std::set<void*>();
+std::set<void*> *ptrs = 0;
 
 }
Index: src/piraha/smart_ptr.hpp
===================================================================
--- src/piraha/smart_ptr.hpp	(revision 5057)
+++ src/piraha/smart_ptr.hpp	(working copy)
@@ -10,29 +10,40 @@
 
 namespace cctki_piraha {
 
+// This global debug variable
+// is used to detect the case
+// where the same pointer is
+// accidentally tracked by two
+// independent smart_ptr_guts
+// objects.
+//
+// This can come about if the
+// programmer does this:
+// foo *f = new foo();
+// smart_ptr<foo> f1(f);
+// smart_ptr<foo> f2(f);
 extern std::set<void*> *ptrs;
 
-// TODO: This code is disabled because it leads to segfaults during
-// startup. Most likely, this code is used during initialisation of
-// global variables, but also implicitly assumes that all global
-// variables have already been initialised.
-#if 0
-//#ifndef NDEBUG
-inline void add(std::set<void*>& v,void *t) {
+#ifdef NDEBUG
+inline void add(void *t) {
     if(t == NULL)
         return;
+    if(ptrs == 0)
+        ptrs = new std::set<void*>();
     // TODO: Don't separate finding and inserting; do it in one go to
     // save a lookup.
-    assert(v.find(t) == v.end());
-    v.insert(t);
+    assert(ptrs->find(t) == ptrs->end());
+    ptrs->insert(t);
 }
 
-inline void remove(std::set<void*>& v,void* t) {
+inline void remove(void* t) {
         // TODO: Don't separate finding and erasing; do it in one go
         // to save a lookup.
-	std::set<void*>::iterator it = v.find(t);
-	assert(it != v.end());
-	v.erase(it);
+    if(t == NULL)
+        return;
+	std::set<void*>::iterator it = ptrs->find(t);
+	assert(it != ptrs->end());
+	ptrs->erase(it);
 }
 #else
 inline void add(std::set<void*>& v,void* t) {
@@ -52,12 +63,12 @@
     bool array;
     smart_ptr_guts(int rc,T *p,bool array_) : ref_count(rc), ptr(p), array(array_) {
         if(ptr != NULL) {
-            add(*ptrs,(void*)ptr);
+            add((void*)ptr);
         }
     }
     ~smart_ptr_guts() {
         if(ptr != NULL) {
-            remove(*ptrs,(void*)ptr);
+            remove((void*)ptr);
             if(array)
                 delete[] ptr;
             else
