Index: tov_carpet.par
===================================================================
--- tov_carpet.par	(revision 55)
+++ tov_carpet.par	(working copy)
@@ -68,8 +68,8 @@
 ADMBase::initial_shift = "tov"
 
 tovsolver::TOV_Rho_Central[0] = 1.28e-3
-tovsolver::TOV_Gamma[0]       = 2.0
-tovsolver::TOV_K[0]           = 100.0
+tovsolver::TOV_Gamma          = 2.0
+tovsolver::TOV_K              = 100.0
 tovsolver::TOV_Combine_Method = "maximum"
 
 ActiveThorns = "ADMMass"
Index: tov.par
===================================================================
--- tov.par	(revision 55)
+++ tov.par	(working copy)
@@ -39,8 +39,8 @@
 ADMBase::initial_shift = "tov"
 
 tovsolver::TOV_Rho_Central[0] = 1.28e-3
-tovsolver::TOV_Gamma[0]       = 2.0
-tovsolver::TOV_K[0]           = 100.0
+tovsolver::TOV_Gamma          = 2.0
+tovsolver::TOV_K              = 100.0
 tovsolver::TOV_Combine_Method = "maximum"
 
 ActiveThorns = "ADMMass"
