Index: test/qlm-bl.par
===================================================================
--- test/qlm-bl.par	(revision 30)
+++ test/qlm-bl.par	(working copy)
@@ -215,6 +215,8 @@
 ActiveThorns = "AHFinderDirect"
 
 AHFinderDirect::find_every = 1
+AHFinderDirect::run_at_CCTK_ANALYSIS = true
+AHFinderDirect::run_at_CCTK_POSTSTEP = false
 
 AHFinderDirect::geometry_interpolator_name = "Lagrange polynomial interpolation"
 AHFinderDirect::geometry_interpolator_pars = "order=4"
