diff --git a/Boundary/src/CopyBoundary.c b/Boundary/src/CopyBoundary.c
index 179982d..d2343c7 100644
--- a/Boundary/src/CopyBoundary.c
+++ b/Boundary/src/CopyBoundary.c
@@ -170,7 +170,7 @@ CCTK_INT BndCopy(const cGH *GH, CCTK_INT num_vars, CCTK_INT *vars,
     /* fill it with values, either from table or the boundary_width
        parameter */
     if (widths[i] < 0) {
-      err = Util_TableGetIntArray(tables[i], gdim, width_alldirs,
+      err = Util_TableGetIntArray(tables[i], 2 * gdim, width_alldirs,
                                   "BOUNDARY_WIDTH");
       if (err < 0) {
         CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
diff --git a/Boundary/src/FlatBoundary.c b/Boundary/src/FlatBoundary.c
index e20fdb7..d5ad109 100644
--- a/Boundary/src/FlatBoundary.c
+++ b/Boundary/src/FlatBoundary.c
@@ -131,7 +131,7 @@ CCTK_INT BndFlat(const cGH *GH, CCTK_INT num_vars, CCTK_INT *vars,
     /* fill it with values, either from table or the boundary_width
        parameter */
     if (widths[i] < 0) {
-      err = Util_TableGetIntArray(tables[i], gdim, width_alldirs,
+      err = Util_TableGetIntArray(tables[i], 2 * gdim, width_alldirs,
                                   "BOUNDARY_WIDTH");
       if (err < 0) {
         CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
diff --git a/Boundary/src/RadiationBoundary.c b/Boundary/src/RadiationBoundary.c
index d4329eb..8a1e139 100644
--- a/Boundary/src/RadiationBoundary.c
+++ b/Boundary/src/RadiationBoundary.c
@@ -238,7 +238,7 @@ CCTK_INT BndRadiative(const cGH *GH, CCTK_INT num_vars, CCTK_INT *vars,
     /* fill it with values, either from table or the boundary_width
        parameter */
     if (widths[i] < 0) {
-      err = Util_TableGetIntArray(tables[i], gdim, width_alldirs,
+      err = Util_TableGetIntArray(tables[i], 2 * gdim, width_alldirs,
                                   "BOUNDARY_WIDTH");
       if (err < 0) {
         CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
diff --git a/Boundary/src/RobinBoundary.c b/Boundary/src/RobinBoundary.c
index 96acb15..c203626 100644
--- a/Boundary/src/RobinBoundary.c
+++ b/Boundary/src/RobinBoundary.c
@@ -162,7 +162,7 @@ CCTK_INT BndRobin(const cGH *GH, CCTK_INT num_vars, CCTK_INT *vars,
     /* fill it with values, either from table or the boundary_width
        parameter */
     if (widths[i] < 0) {
-      err = Util_TableGetIntArray(tables[i], gdim, width_alldirs,
+      err = Util_TableGetIntArray(tables[i], 2 * gdim, width_alldirs,
                                   "BOUNDARY_WIDTH");
       if (err < 0) {
         CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
diff --git a/Boundary/src/ScalarBoundary.c b/Boundary/src/ScalarBoundary.c
index f7b90cb..53fabaa 100644
--- a/Boundary/src/ScalarBoundary.c
+++ b/Boundary/src/ScalarBoundary.c
@@ -147,7 +147,7 @@ CCTK_INT BndScalar(const cGH *GH, CCTK_INT num_vars, CCTK_INT *vars,
     /* fill it with values, either from table or the boundary_width
        parameter */
     if (widths[i] < 0) {
-      err = Util_TableGetIntArray(tables[i], gdim, width_alldirs,
+      err = Util_TableGetIntArray(tables[i], 2 * gdim, width_alldirs,
                                   "BOUNDARY_WIDTH");
       if (err < 0) {
         CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
diff --git a/Boundary/src/StaticBoundary.c b/Boundary/src/StaticBoundary.c
index a5755d0..d4ce0bf 100644
--- a/Boundary/src/StaticBoundary.c
+++ b/Boundary/src/StaticBoundary.c
@@ -128,7 +128,7 @@ CCTK_INT BndStatic(const cGH *GH, CCTK_INT num_vars, CCTK_INT *vars,
     /* fill it with values, either from table or the boundary_width
        parameter */
     if (widths[i] < 0) {
-      err = Util_TableGetIntArray(tables[i], gdim, width_alldirs,
+      err = Util_TableGetIntArray(tables[i], 2 * gdim, width_alldirs,
                                   "BOUNDARY_WIDTH");
       if (err < 0) {
         CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
