Index: configure.sh
===================================================================
--- configure.sh	(revision 38)
+++ configure.sh	(working copy)
@@ -129,6 +129,8 @@
         
         echo "OpenSSL: Configuring..."
         cd ${NAME}
+        # OpenSSL's 'config' script uses $options itself, so we unset it
+        unset options
         ./config --prefix=${OPENSSL_DIR}
         
         echo "OpenSSL: Building..."
