Index: src/Mag_NS.cc
===================================================================
--- src/Mag_NS.cc	(revision 14)
+++ src/Mag_NS.cc	(working copy)
@@ -1,6 +1,7 @@
 #include <cassert>
 #include <cstdio>
 #include <vector>
+#include <ios>
 
 #include <cctk.h>
 #include <cctk_Arguments.h>
@@ -69,6 +70,7 @@
   
   CCTK_VInfo (CCTK_THORNSTRING, "Reading from file \"%s\"", filename);
   
+  try {
   Mag_NS mag_ns (npoints, &xx[0], &yy[0], &zz[0], filename);
   
   CCTK_VInfo (CCTK_THORNSTRING, "omega [rad/s]:     %g", mag_ns.omega);
@@ -194,4 +196,8 @@
   
   
   CCTK_INFO ("Done.");
+  } catch (ios::failure e) {
+    CCTK_VWarn (CCTK_WARN_ABORT, __LINE__, __FILE__, CCTK_THORNSTRING,
+                "Could not read initial data from file '%s': %s", filename, e.what());
+  }
 }
