From 2888d221261da25e7802b76c46ee3e10ca578418 Mon Sep 17 00:00:00 2001
From: Barry Wardell <barry.wardell@gmail.com>
Date: Thu, 18 Aug 2011 00:05:43 +0200
Subject: [PATCH 1/2] Enable Vectorisation.

---
 m/McLachlan_BSSN.m |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/m/McLachlan_BSSN.m b/m/McLachlan_BSSN.m
index 927b894..a999fb6 100644
--- a/m/McLachlan_BSSN.m
+++ b/m/McLachlan_BSSN.m
@@ -1348,6 +1348,7 @@ CreateKrancThornTT [groups, ".", BSSN,
   DefaultEvolutionTimelevels -> 3,
   UseJacobian -> useJacobian,
   UseLoopControl -> True,
+  UseVectors -> True,
   InheritedImplementations -> inheritedImplementations,
   InheritedKeywordParameters -> inheritedKeywordParameters,
   ExtendedKeywordParameters -> extendedKeywordParameters,
-- 
1.7.5.1

