Index: interface.ccl
===================================================================
--- interface.ccl	(revision 143)
+++ interface.ccl	(working copy)
@@ -4,10 +4,6 @@
 implements: IOJpeg
 inherits:   IO
 
-USES INCLUDE HEADER: jconfig.h
-USES INCLUDE HEADER: jpeglib.h
-
-
 CCTK_INT FUNCTION \
   Hyperslab_Get (CCTK_POINTER_TO_CONST IN cctkGH, \
                  CCTK_INT IN mapping_handle, \
