From 18f14fbcce590e75d93d2ca2afe292b2d0a83d45 Mon Sep 17 00:00:00 2001
From: Roland Haas <roland.haas@physics.gatech.edu>
Date: Sun, 15 Jul 2012 14:39:39 -0700
Subject: [PATCH 01/25] GRHydro: make more paramters steerable

---
 param.ccl |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/param.ccl b/param.ccl
index c2c5c0d..6b808a2 100644
--- a/param.ccl
+++ b/param.ccl
@@ -659,7 +659,7 @@ int GRHydro_c2p_warn_from_reflevel "Start warning on given refinement level and
 # This can improve the performance on many cores, but may be dangerous 
 # (it seems to work in simple testcases, but morde complicated settings may fail)
 # If you encounter strange behavior, switch this OFF
-BOOLEAN sync_conserved_only "Only sync evolved conserved quantities during evolution."
+BOOLEAN sync_conserved_only "Only sync evolved conserved quantities during evolution." STEERABLE = recover
 {
 } no
 
@@ -677,7 +677,7 @@ private:
 
 
 
-BOOLEAN verbose "Some debug output"
+BOOLEAN verbose "Some debug output" STEERABLE = always
 {
 } no
 
-- 
1.7.10.4

