Index: make.code.deps
===================================================================
--- make.code.deps	(revision 506)
+++ make.code.deps	(working copy)
@@ -1,8 +0,0 @@
-# Module dependencies for thorn GRHydro
-
-GRHydro_ENOReconstruct.F90.o:		GRHydro_ENOScalars.F90.o
-GRHydro_Eigenproblem.F90.o:		GRHydro_Scalars.F90.o
-GRHydro_Eigenproblem_Marquina.F90.o:	GRHydro_Scalars.F90.o
-GRHydro_PreLoop.F90.o:			GRHydro_Scalars.F90.o
-GRHydro_TVDReconstruct.F90.o:		GRHydro_Scalars.F90.o
-GRHydro_SlopeLimiter.F90.o:		GRHydro_Scalars.F90.o
