Index: interface.ccl
===================================================================
--- interface.ccl	(revision 7)
+++ interface.ccl	(working copy)
@@ -39,12 +39,12 @@
   eth4_maskn, eth4_masks
 } "dissipation masks, to protect the startup from taking eth4"
 
-CCTK_COMPLEX Jrad TYPE=ARRAY dim=1 size=NullGrid::N_radial_pts
+CCTK_COMPLEX Jrad TYPE=ARRAY dim=1 size=NullGrid::N_radial_pts DISTRIB=CONSTANT
 {
  jcn_rad, jcs_rad
 }
 
-CCTK_COMPLEX dxJrad TYPE=ARRAY dim=1 size=NullGrid::N_radial_pts
+CCTK_COMPLEX dxJrad TYPE=ARRAY dim=1 size=NullGrid::N_radial_pts DISTRIB=CONSTANT
 {
  dxjcn_rad, dxjcs_rad
 }
