From 3786e47951e7f0ac3c96d731d79471f5dd032ecb Mon Sep 17 00:00:00 2001
From: Barry Wardell <barry.wardell@gmail.com>
Date: Wed, 12 Feb 2014 11:26:44 -0500
Subject: [PATCH 1/2] Make sure to use lib64 and other alternative library
 directories when appropriate

---
 configure.sh | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/configure.sh b/configure.sh
index 17a1f32..f877ac6 100755
--- a/configure.sh
+++ b/configure.sh
@@ -38,6 +38,7 @@ if [ -z "${HWLOC_DIR}" ]; then
                 done
                 # don't look further if all files have been found
                 if [ -n "$HWLOC_DIR" ]; then
+                    HWLOC_LIB_DIR="$HWLOC_DIR/$libdir"
                     break
                 fi
             done
@@ -89,6 +90,7 @@ then
     fi
     DONE_FILE=${SCRATCH_BUILD}/done/${THORN}
     HWLOC_DIR=${INSTALL_DIR}
+    HWLOC_LIB_DIR=${INSTALL_DIR}/lib
     
     if [ -e ${DONE_FILE} -a ${DONE_FILE} -nt ${SRCDIR}/dist/${NAME}.tar.gz \
                          -a ${DONE_FILE} -nt ${SRCDIR}/configure.sh ]
@@ -181,14 +183,14 @@ fi
 
 
 # Check that pkg-config works
-export PKG_CONFIG_PATH=${HWLOC_DIR}/lib/pkgconfig:${PCIUTILS_DIR}/lib/pkgconfig:${PKG_CONFIG_PATH}
+export PKG_CONFIG_PATH=${HWLOC_LIB_DIR}/pkgconfig:${PCIUTILS_DIR}/lib/pkgconfig:${PKG_CONFIG_PATH}
 if ! pkg-config hwloc; then
     echo "BEGIN MESSAGE"
     echo "pkg-config not found; attempting to use reasonable defaults"
     echo "END MESSAGE"
 
     HWLOC_INC_DIRS="${HWLOC_DIR}/include"
-    HWLOC_LIB_DIRS="${HWLOC_DIR}/lib"
+    HWLOC_LIB_DIRS="${HWLOC_LIB_DIR}"
     HWLOC_LIBS='hwloc'
 else
     HWLOC_INC_DIRS="$(echo '' $(pkg-config hwloc --static --cflags 2>/dev/null || pkg-config hwloc --cflags) '' | sed -e 's+ -I/include + +g;s+ -I/usr/include + +g;s+ -I/usr/local/include + +g' | sed -e 's/ -I/ /g')"
@@ -197,7 +199,7 @@ else
 fi
 
 # Add libnuma manually, if necessary
-if grep -q '[-]lnuma' ${HWLOC_DIR}/lib/libhwloc.la; then
+if grep -q '[-]lnuma' ${HWLOC_LIB_DIR}/lib/libhwloc.la; then
     if ! echo '' ${HWLOC_LIBS} '' | grep -q ' numa '; then
         HWLOC_LIBS="${HWLOC_LIBS} numa"
     fi
-- 
1.8.4.1

