Index: util/Math.c
===================================================================
--- util/Math.c	(revision 4883)
+++ util/Math.c	(working copy)
@@ -11,6 +11,7 @@
 
 #include <math.h>
 #include <cctk_Config.h>
+#include <cctk_Math.h>
 
 
 
@@ -23,6 +24,14 @@
 #endif
 }
 
+double CCTK_FNAME(CCTK_copysign)(double *x, double *y);
+double CCTK_FNAME(CCTK_copysign)(double *x, double *y)
+{
+  return CCTK_copysign(*x, *y);
+}
+
+
+
 int CCTK_fpclassify(double x)
 {
 #ifdef HAVE_FPCLASSIFY
@@ -32,6 +41,14 @@
 #endif
 }
 
+int CCTK_FNAME(CCTK_fpclassify)(double *x);
+int CCTK_FNAME(CCTK_fpclassify)(double *x)
+{
+  return CCTK_fpclassify(*x);
+}
+
+
+
 int CCTK_isfinite(double x)
 {
 #ifdef HAVE_ISFINITE
@@ -41,6 +58,14 @@
 #endif
 }
 
+int CCTK_FNAME(CCTK_isfinite)(double *x);
+int CCTK_FNAME(CCTK_isfinite)(double *x)
+{
+  return CCTK_isfinite(*x);
+}
+
+
+
 int CCTK_isinf(double x)
 {
 #ifdef HAVE_ISINF
@@ -50,6 +75,14 @@
 #endif
 }
 
+int CCTK_FNAME(CCTK_isinf)(double *x);
+int CCTK_FNAME(CCTK_isinf)(double *x)
+{
+  return CCTK_isinf(*x);
+}
+
+
+
 int CCTK_isnan(double x)
 {
 #ifdef HAVE_ISNAN
@@ -59,6 +92,14 @@
 #endif
 }
 
+int CCTK_FNAME(CCTK_isnan)(double *x);
+int CCTK_FNAME(CCTK_isnan)(double *x)
+{
+  return CCTK_isnan(*x);
+}
+
+
+
 int CCTK_isnormal(double x)
 {
 #ifdef HAVE_ISNORMAL
@@ -68,6 +109,14 @@
 #endif
 }
 
+int CCTK_FNAME(CCTK_isnormal)(double *x);
+int CCTK_FNAME(CCTK_isnormal)(double *x)
+{
+  return CCTK_isnormal(*x);
+}
+
+
+
 int CCTK_signbit(double x)
 {
 #ifdef HAVE_SIGNBIT
@@ -76,3 +125,9 @@
   return x < 0.0;
 #endif
 }
+
+int CCTK_FNAME(CCTK_signbit)(double *x);
+int CCTK_FNAME(CCTK_signbit)(double *x)
+{
+  return CCTK_signbit(*x);
+}
