Index: src/driver/setup.cc
===================================================================
--- src/driver/setup.cc	(revision 1568)
+++ src/driver/setup.cc	(working copy)
@@ -683,6 +683,13 @@
       printf ("AHF setup %d found_flag=%d\n",               n+1, (int) AH_data.found_flag);
     }
   }
+
+  // Save in grid array in case a recovery only recovers some horizons
+  for (int n = 0; n < N_horizons; ++ n) {
+    struct AH_data& AH_data = *state.AH_data_array[n+1];
+    const struct BH_diagnostics& BH_diagnostics = AH_data.BH_diagnostics;
+    BH_diagnostics.save(cctkGH, n+1);
+  }
 }
 
 
