Index: mdb/submitscripts/carver.sub
===================================================================
--- mdb/submitscripts/carver.sub	(revision 2417)
+++ mdb/submitscripts/carver.sub	(working copy)
@@ -10,4 +10,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/pandora.sub
===================================================================
--- mdb/submitscripts/pandora.sub	(revision 2417)
+++ mdb/submitscripts/pandora.sub	(working copy)
@@ -22,4 +22,4 @@
 #@ error = @RUNDIR@/@SIMULATION_NAME@.err
 #@ queue
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/tianhe1a.sub
===================================================================
--- mdb/submitscripts/tianhe1a.sub	(revision 2417)
+++ mdb/submitscripts/tianhe1a.sub	(working copy)
@@ -9,4 +9,4 @@
 #SBATCH -o @RUNDIR@/@SIMULATION_NAME@.out
 #SBATCH -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/queenbee-mvapich2.sub
===================================================================
--- mdb/submitscripts/queenbee-mvapich2.sub	(revision 2417)
+++ mdb/submitscripts/queenbee-mvapich2.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/stampede.sub
===================================================================
--- mdb/submitscripts/stampede.sub	(revision 2417)
+++ mdb/submitscripts/stampede.sub	(working copy)
@@ -10,4 +10,4 @@
 #SBATCH -o @RUNDIR@/@SIMULATION_NAME@.out
 #SBATCH -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/fermi.sub
===================================================================
--- mdb/submitscripts/fermi.sub	(revision 2417)
+++ mdb/submitscripts/fermi.sub	(working copy)
@@ -16,4 +16,4 @@
 # @ output = @RUNDIR@/@SIMULATION_NAME@.out
 # @ error = @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/supermuc-ibmmpi.sub
===================================================================
--- mdb/submitscripts/supermuc-ibmmpi.sub	(revision 2417)
+++ mdb/submitscripts/supermuc-ibmmpi.sub	(working copy)
@@ -20,4 +20,4 @@
 #@ minimize_time_to_solution = yes
 #@ queue
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --basedir=@BASEDIR@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@  --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/titan.sub
===================================================================
--- mdb/submitscripts/titan.sub	(revision 2417)
+++ mdb/submitscripts/titan.sub	(working copy)
@@ -9,4 +9,4 @@
 #BSUB -o @RUNDIR@/@SIMULATION_NAME@.out
 #BSUB -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=titan --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=titan --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/supermuc-intelmpi.sub
===================================================================
--- mdb/submitscripts/supermuc-intelmpi.sub	(revision 2417)
+++ mdb/submitscripts/supermuc-intelmpi.sub	(working copy)
@@ -20,4 +20,4 @@
 #@ minimize_time_to_solution = yes
 #@ queue
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/angel.sub
===================================================================
--- mdb/submitscripts/angel.sub	(revision 2417)
+++ mdb/submitscripts/angel.sub	(working copy)
@@ -1,3 +1,3 @@
 #! /bin/bash
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/saw.sub
===================================================================
--- mdb/submitscripts/saw.sub	(revision 2417)
+++ mdb/submitscripts/saw.sub	(working copy)
@@ -1,3 +1,3 @@
 #! /bin/bash
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/tezpur-mvapich2.sub
===================================================================
--- mdb/submitscripts/tezpur-mvapich2.sub	(revision 2417)
+++ mdb/submitscripts/tezpur-mvapich2.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/stampede-mic.sub
===================================================================
--- mdb/submitscripts/stampede-mic.sub	(revision 2417)
+++ mdb/submitscripts/stampede-mic.sub	(working copy)
@@ -9,4 +9,4 @@
 #SBATCH -o @RUNDIR@/@SIMULATION_NAME@.out
 #SBATCH -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/generic.sub
===================================================================
--- mdb/submitscripts/generic.sub	(revision 2417)
+++ mdb/submitscripts/generic.sub	(working copy)
@@ -6,4 +6,4 @@
         sleep 60
     done
 fi
-exec @SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@ < /dev/null > @RUNDIR@/@SIMULATION_NAME@.out 2> @RUNDIR@/@SIMULATION_NAME@.err
+exec @SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@ < /dev/null > @RUNDIR@/@SIMULATION_NAME@.out 2> @RUNDIR@/@SIMULATION_NAME@.err
Index: mdb/submitscripts/bluewaters.sub
===================================================================
--- mdb/submitscripts/bluewaters.sub	(revision 2417)
+++ mdb/submitscripts/bluewaters.sub	(working copy)
@@ -10,4 +10,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/datura.sub
===================================================================
--- mdb/submitscripts/datura.sub	(revision 2417)
+++ mdb/submitscripts/datura.sub	(working copy)
@@ -12,4 +12,4 @@
 #$ -cwd
 #$ -R y
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=datura --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=datura --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/stampede-hybrid.sub
===================================================================
--- mdb/submitscripts/stampede-hybrid.sub	(revision 2417)
+++ mdb/submitscripts/stampede-hybrid.sub	(working copy)
@@ -8,4 +8,4 @@
 #SBATCH -o @RUNDIR@/@SIMULATION_NAME@.out
 #SBATCH -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/vesta.sub
===================================================================
--- mdb/submitscripts/vesta.sub	(revision 2417)
+++ mdb/submitscripts/vesta.sub	(working copy)
@@ -1,4 +1,4 @@
 #! /bin/bash
 # Queueing options are set via qsub options
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/shc-openmpi.sub
===================================================================
--- mdb/submitscripts/shc-openmpi.sub	(revision 2417)
+++ mdb/submitscripts/shc-openmpi.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/debian.sub
===================================================================
--- mdb/submitscripts/debian.sub	(revision 2417)
+++ mdb/submitscripts/debian.sub	(working copy)
@@ -6,4 +6,4 @@
         sleep 60
     done
 fi
-exec @SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@ < /dev/null > @RUNDIR@/@SIMULATION_NAME@.out 2> @RUNDIR@/@SIMULATION_NAME@.err
+exec @SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@ < /dev/null > @RUNDIR@/@SIMULATION_NAME@.out 2> @RUNDIR@/@SIMULATION_NAME@.err
Index: mdb/submitscripts/philip-mpich.sub
===================================================================
--- mdb/submitscripts/philip-mpich.sub	(revision 2417)
+++ mdb/submitscripts/philip-mpich.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/gordon.sub
===================================================================
--- mdb/submitscripts/gordon.sub	(revision 2417)
+++ mdb/submitscripts/gordon.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/intrepid-xlc.sub
===================================================================
--- mdb/submitscripts/intrepid-xlc.sub	(revision 2417)
+++ mdb/submitscripts/intrepid-xlc.sub	(working copy)
@@ -1,3 +1,3 @@
 #! /bin/bash
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/requin.sub
===================================================================
--- mdb/submitscripts/requin.sub	(revision 2417)
+++ mdb/submitscripts/requin.sub	(working copy)
@@ -1,3 +1,3 @@
 #! /bin/bash
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/datura-gpu.sub
===================================================================
--- mdb/submitscripts/datura-gpu.sub	(revision 2417)
+++ mdb/submitscripts/datura-gpu.sub	(working copy)
@@ -12,4 +12,4 @@
 #$ -cwd
 #$ -R y
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=datura-gpu --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=datura-gpu --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/s-kraken.sub
===================================================================
--- mdb/submitscripts/s-kraken.sub	(revision 2417)
+++ mdb/submitscripts/s-kraken.sub	(working copy)
@@ -1,3 +1,3 @@
 #! /bin/bash
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/loewe.sub
===================================================================
--- mdb/submitscripts/loewe.sub	(revision 2417)
+++ mdb/submitscripts/loewe.sub	(working copy)
@@ -11,4 +11,4 @@
 #SBATCH --output=@RUNDIR@/@SIMULATION_NAME@.out
 #SBATCH --error=@RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@  --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/compute.sub
===================================================================
--- mdb/submitscripts/compute.sub	(revision 2417)
+++ mdb/submitscripts/compute.sub	(working copy)
@@ -6,4 +6,4 @@
         sleep 60
     done
 fi
-exec @SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@ < /dev/null > @RUNDIR@/@SIMULATION_NAME@.out 2> @RUNDIR@/@SIMULATION_NAME@.err
+exec @SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@ < /dev/null > @RUNDIR@/@SIMULATION_NAME@.out 2> @RUNDIR@/@SIMULATION_NAME@.err
Index: mdb/submitscripts/shelob.sub
===================================================================
--- mdb/submitscripts/shelob.sub	(revision 2417)
+++ mdb/submitscripts/shelob.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/orca.sub
===================================================================
--- mdb/submitscripts/orca.sub	(revision 2417)
+++ mdb/submitscripts/orca.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 ##PBS -W x=nmatchpolicy:exactnode
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/mike.sub
===================================================================
--- mdb/submitscripts/mike.sub	(revision 2417)
+++ mdb/submitscripts/mike.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/zwicky-openmpi.sub
===================================================================
--- mdb/submitscripts/zwicky-openmpi.sub	(revision 2417)
+++ mdb/submitscripts/zwicky-openmpi.sub	(working copy)
@@ -10,4 +10,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/bethe.sub
===================================================================
--- mdb/submitscripts/bethe.sub	(revision 2417)
+++ mdb/submitscripts/bethe.sub	(working copy)
@@ -6,4 +6,4 @@
         sleep 60
     done
 fi
-exec @SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@ < /dev/null > @RUNDIR@/@SIMULATION_NAME@.out 2> @RUNDIR@/@SIMULATION_NAME@.err
+exec @SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@ < /dev/null > @RUNDIR@/@SIMULATION_NAME@.out 2> @RUNDIR@/@SIMULATION_NAME@.err
Index: mdb/submitscripts/lonestar.sub
===================================================================
--- mdb/submitscripts/lonestar.sub	(revision 2417)
+++ mdb/submitscripts/lonestar.sub	(working copy)
@@ -12,4 +12,4 @@
 #$ -o @RUNDIR@/@SIMULATION_NAME@.out
 #$ -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/hopper.sub
===================================================================
--- mdb/submitscripts/hopper.sub	(revision 2417)
+++ mdb/submitscripts/hopper.sub	(working copy)
@@ -10,7 +10,7 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
 
 # These two options are not recommended any more:
 # -l mppnppn=@(@PPN@/@NUM_THREADS@)@
Index: mdb/submitscripts/trestles.sub
===================================================================
--- mdb/submitscripts/trestles.sub	(revision 2417)
+++ mdb/submitscripts/trestles.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/gpc.sub
===================================================================
--- mdb/submitscripts/gpc.sub	(revision 2417)
+++ mdb/submitscripts/gpc.sub	(working copy)
@@ -10,4 +10,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/fionn.sub
===================================================================
--- mdb/submitscripts/fionn.sub	(revision 2417)
+++ mdb/submitscripts/fionn.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=fionn --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=fionn --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
Index: mdb/submitscripts/pbs-normal.sub
===================================================================
--- mdb/submitscripts/pbs-normal.sub	(revision 2417)
+++ mdb/submitscripts/pbs-normal.sub	(working copy)
@@ -11,4 +11,4 @@
 #PBS -o @RUNDIR@/@SIMULATION_NAME@.out
 #PBS -e @RUNDIR@/@SIMULATION_NAME@.err
 cd @SOURCEDIR@
-@SIMFACTORY@ run @SIMULATION_NAME@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
+@SIMFACTORY@ run @SIMULATION_NAME@ --basedir=@BASEDIR@ --machine=@MACHINE@ --restart-id=@RESTART_ID@ @FROM_RESTART_COMMAND@
