diff --git a/lib/sbin/ScheduleParser.pl b/lib/sbin/ScheduleParser.pl
index f9105e3..eceee73 100644
--- a/lib/sbin/ScheduleParser.pl
+++ b/lib/sbin/ScheduleParser.pl
@@ -43,7 +43,8 @@ our @schedule_bins = (
     'SHUTDOWN');
 # A regular expression matching all possible schedule bins, including
 # a CCTK prefix and in upper case
-our $schedule_bin_regexp = 'CCTK_(' . join ('|', @schedule_bins) . ')';
+our $schedule_bin_regexp = '\bCCTK_(' . join ('|', @schedule_bins) . ')\b';
+our $time_bin_info = {};
 
 # Check that the schedule bin exists
 my %schedule_bins = ();
@@ -185,8 +186,22 @@ sub parse_schedule_statement
             } elsif($prep_name eq "at") {
               $where = uc($prep->group(1,"pararg")->group(0,"vname")->substring());
               $where =~ s/^(CCTK_|)/CCTK_/gi;
+              $time_bin_info->{$thorn}->{$name}->{$where} = {
+                level => 0,
+                line  => $prep->linenum(),
+                file  => $ccl_file,
+              };
+              if($where !~ $schedule_bin_regexp) {
+   						  my $hint="If this routine should be scheduled check the spelling of the group or timebin name. Note that scheduling IN must be used to schedule a routine to run in a thorn-defined schedule group, whereas scheduling AT is used for a usual timebin. (Schedule IN may also be used with the usual timebins, but in this case the full name of the bin must be used, e.g. CCTK_EVOL and not EVOL";
+  						  &CST_error(0,"Scheduling routine $name from thorn $thorn in non-existent group or timebin $where",$hint,$prep->linenum(),$ccl_file);
+              }
             } elsif($prep_name eq "in") {
               $where = $prep->group(1,"pararg")->group(0,"vname")->substring();
+                $time_bin_info->{$thorn}->{$name}->{$where} = {
+                  level => 1,
+                  line  => $prep->linenum(),
+                  file  => $ccl_file,
+                };
             } elsif($prep_name eq "while") {
               $while_list = "";
               for my $w (@{$prep->group(1)->{children}}) {
@@ -948,10 +963,16 @@ sub check_schedule_database
     {
       if(!defined($allgroups{$rhschedule_db->{"\U$thorn\E BLOCK_$block WHERE"}}))
       {
-        my $message = "Scheduling routine $rhschedule_db->{\"\U$thorn\E BLOCK_$block NAME\"} from thorn $thorn in non-existent group or timebin $rhschedule_db->{\"\U$thorn\E BLOCK_$block WHERE\"}";
+        my $name  = $rhschedule_db->{"\U$thorn\E BLOCK_$block NAME"};
+        my $where = $rhschedule_db->{"\U$thorn\E BLOCK_$block WHERE"};
+        my $message = "Scheduling routine $name from thorn $thorn in non-existent group or timebin $where";
         print "    $message\n";
         my $hint = "If this routine should be scheduled check the spelling of the group or timebin name. Note that scheduling IN must be used to schedule a routine to run in a thorn-defined schedule group, whereas scheduling AT is used for a usual timebin. (Schedule IN may also be used with the usual timebins, but in this case the full name of the bin must be used, e.g. CCTK_EVOL and not EVOL)";
-        &CST_error(1,$message,$hint,__LINE__,__FILE__);
+        my $info = $time_bin_info->{$thorn}->{$name}->{$where};
+        my $level = $info->{level};
+        if($level != 0) {
+          &CST_error($level,$message,$hint,$info->{line},$info->{file});
+        }
       }
     }
   }
