diff --git a/lib/sbin/parameter_parser.pl b/lib/sbin/parameter_parser.pl
index 371cefb..61740b9 100644
--- a/lib/sbin/parameter_parser.pl
+++ b/lib/sbin/parameter_parser.pl
@@ -396,7 +396,7 @@ sub parse_param_ccl
             if ($type eq 'REAL' && ! (
                 $new_ranges eq '*' ||
                 $new_ranges =~ /^$real$/ ||
-                $new_ranges =~ /^["(]?(\*|\(?$real)?:(\*|$real\)?)?(:$real)?[)"]?$/)) {
+                $new_ranges =~ /^["(\[]?(\*|\(?$real)?:(\*|$real\)?)?(:$real)?[\])"]?$/)) {
               # this will become a level 0 error in the fture
               &CST_error(1, "Invalid range '$new_ranges' for real " .
                          "parameter '$variable' of thorn '$thorn'. ".
