Index: configure.sh
===================================================================
--- configure.sh	(revision 18)
+++ configure.sh	(working copy)
@@ -15,7 +15,9 @@
 # Search
 ################################################################################
 
-if [ -z "${HWLOC_DIR}" ]; then
+if [ -z "${HWLOC_DIR}" \
+     -o "$(echo "${HWLOC_DIR}" | tr '[a-z]' '[A-Z]')" = 'NO_BUILD' ]
+then
     echo "BEGIN MESSAGE"
     echo "hwloc selected, but HWLOC_DIR not set. Checking some places..."
     echo "END MESSAGE"
