Index: src/StaticBoundary.c
===================================================================
--- src/StaticBoundary.c	(revision 306)
+++ src/StaticBoundary.c	(working copy)
@@ -95,7 +95,7 @@
   int dir;            /* direction in which to apply bc */
 
 #ifdef DEBUG
-  printf("BndStatic(): got passed GH=%p, num_vars=%d, var_indices[0]=%d, table_handles[0]=%d\n", (const void *) GH, num_vars, var_indices[0], table_handles[0]);
+  printf("BndStatic(): got passed GH=%p, num_vars=%d, vars[0]=%d, tables[0]=%d\n", (const void *) GH, num_vars, vars[0], tables[0]);
 #endif
 
   retval = 0; width_alldirs = NULL; max_gdim = 0;
Index: src/RobinBoundary.c
===================================================================
--- src/RobinBoundary.c	(revision 306)
+++ src/RobinBoundary.c	(working copy)
@@ -106,7 +106,7 @@
   CCTK_INT npow;           /* decay rate */
 
 #ifdef DEBUG
-  printf("BndRobin(): got passed GH=%p, num_vars=%d, var_indices[0]=%d, table_handles[0]=%d\n", (const void *) GH, num_vars, var_indices[0], table_handles[0]);
+  printf("BndRobin(): got passed GH=%p, num_vars=%d, vars[0]=%d, tables[0]=%d\n", (const void *) GH, num_vars, vars[0], tables[0]);
 #endif
 
   retval = 0; width_alldirs = NULL; max_gdim = 0;
