From 106aa55b152e4c9e697f42d4c501b4957a4fd9d4 Mon Sep 17 00:00:00 2001
From: Roland Haas <roland.haas@physics.gatech.edu>
Date: Sun, 22 Apr 2012 20:07:50 -0700
Subject: [PATCH 3/3] GRHydro_InitData: REQUIRE EOS_Omni and Fortran

---
 configuration.ccl |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/configuration.ccl b/configuration.ccl
index 15805b5..aa25d0e 100644
--- a/configuration.ccl
+++ b/configuration.ccl
@@ -1,2 +1 @@
-REQUIRES EOS_Omni
-
+REQUIRES EOS_Omni Fortran
-- 
1.7.10.4

