Index: configure.sh
===================================================================
--- configure.sh	(revision 19)
+++ configure.sh	(working copy)
@@ -29,7 +29,9 @@
 # Search
 ################################################################################
 
-if [ -z "${MPI_DIR}" ]; then
+if [ -z "${MPI_DIR}" \
+     -o "$(echo "${MPI_DIR}" | tr '[a-z]' '[A-Z]')" = 'NO_BUILD' ] 
+then
     echo "BEGIN MESSAGE"
     echo "MPI selected, but MPI_DIR not set. Checking some places..."
     echo "END MESSAGE"
