Index: GRHydro_Prim2Con.F90
===================================================================
--- GRHydro_Prim2Con.F90	(revision 586)
+++ GRHydro_Prim2Con.F90	(working copy)
@@ -478,7 +478,7 @@
                    g12(i,j,k),g13(i,j,k),&
                    g22(i,j,k),g23(i,j,k),g33(i,j,k),&
                    det, dens(i,j,k),sx(i,j,k),sy(i,j,k),sz(i,j,k),&
-                   tau(i,j,k),rho(i,j,k),velx(i,j,k),vely(i,j,k),velz(i,j,k),&
+                   tau(i,j,k),rho(i,j,k),vup(i,j,k,1),vup(i,j,k,2),vup(i,j,k,3),&
                    eps(i,j,k),press(i,j,k),w_lorentz(i,j,k))
            end do
         end do
@@ -501,7 +501,7 @@
                    g12(i,j,k),g13(i,j,k),&
                    g22(i,j,k),g23(i,j,k),g33(i,j,k),&
                    det, dens(i,j,k),sx(i,j,k),sy(i,j,k),sz(i,j,k),&
-                   tau(i,j,k),rho(i,j,k),velx(i,j,k),vely(i,j,k),velz(i,j,k),&
+                   tau(i,j,k),rho(i,j,k),vup(i,j,k,1),vup(i,j,k,2),vup(i,j,k,3),&
                    eps(i,j,k),press(i,j,k),w_lorentz(i,j,k), &
                    temperature(i,j,k),y_e(i,j,k))
            end do
@@ -771,7 +771,7 @@
         call prim2conpolytype(GRHydro_eos_handle,g11(i,j,k),g12(i,j,k),&
              g13(i,j,k),g22(i,j,k),g23(i,j,k),g33(i,j,k),&
              det, dens(i,j,k),sx(i,j,k),sy(i,j,k),sz(i,j,k),&
-             tau(i,j,k),rho(i,j,k),velx(i,j,k),vely(i,j,k),velz(i,j,k),&
+             tau(i,j,k),rho(i,j,k),vup(i,j,k,1),vup(i,j,k,2),vup(i,j,k,3),&
              eps(i,j,k),press(i,j,k),w_lorentz(i,j,k))
 
       end do
