diff --git a/ADMBase/schedule.ccl b/ADMBase/schedule.ccl
index ae1c1b8..e9d1a35 100644
--- a/ADMBase/schedule.ccl
+++ b/ADMBase/schedule.ccl
@@ -173,6 +173,7 @@ if (CCTK_Equals(initial_shift, "zero"))
   SCHEDULE ADMBase_ShiftZero in ADMBase_InitialGauge
   {
     LANG: C
+    WRITES: ADMBASE::shift(everywhere)
   } "Set the shift to 0 at all points"
 }
 
@@ -182,6 +183,7 @@ if (CCTK_Equals(initial_dtlapse, "zero"))
   SCHEDULE ADMBase_DtLapseZero in ADMBase_InitialGauge
   {
     LANG: C
+    WRITES: ADMBase::dtalp(EVERYWHERE)
   } "Set the dtlapse to 0 at all points"
 }
 
@@ -191,6 +193,7 @@ if (CCTK_Equals(initial_dtshift, "zero"))
   SCHEDULE ADMBase_DtShiftZero in ADMBase_InitialGauge
   {
     LANG: C
+    WRITES: ADMBase::dtshift(EVERYWHERE)
   } "Set the dtshift to 0 at all points"
 }
 
