diff --git a/m/WaveToy.m b/m/WaveToy.m
index 7163269..e0df292 100644
--- a/m/WaveToy.m
+++ b/m/WaveToy.m
@@ -17,11 +17,9 @@ derivatives =
    StandardCenteredDifferenceOperator[1,derivOrder/2,i]
    StandardCenteredDifferenceOperator[1,derivOrder/2,j]};
 
-FD = PDstandardNth;
-ResetJacobians;
-DefineJacobian[PD, FD, KD, Zero3];
+PD = PDstandardNth;
 
-evolutionTimelevels = 2;
+evolutionTimelevels = 4;
 
 KD = KroneckerDelta;
 
