Index: arrangements/EinsteinAnalysis/AHFinderDirect/schedule.ccl
===================================================================
--- arrangements/EinsteinAnalysis/AHFinderDirect/schedule.ccl	(revision 1567)
+++ arrangements/EinsteinAnalysis/AHFinderDirect/schedule.ccl	(working copy)
@@ -105,8 +105,7 @@
      } "import the excision mask"
 
   schedule AHFinderDirect_store at CCTK_ANALYSIS \
-     after AHFinderDirect_find_horizons \
-     before SphericalSurface_HasBeenSet
+     after AHFinderDirect_find_horizons
      {
      lang: C
      options: global
@@ -189,8 +188,7 @@
      } "import the excision mask"
 
   schedule AHFinderDirect_store at CCTK_POSTINITIAL \
-     after AHFinderDirect_find_horizons \
-     before SphericalSurface_HasBeenSet
+     after AHFinderDirect_find_horizons
      {
      lang: C
      options: global
@@ -231,8 +229,7 @@
      } "import the excision mask"
 
   schedule AHFinderDirect_store at CCTK_POSTPOSTINITIAL \
-     after AHFinderDirect_find_horizons \
-     before SphericalSurface_HasBeenSet
+     after AHFinderDirect_find_horizons
      {
      lang: C
      options: global
@@ -273,15 +270,14 @@
      } "import the excision mask"
 
   schedule AHFinderDirect_store at CCTK_POST_RECOVER_VARIABLES \
-     after AHFinderDirect_find_horizons \
-     before SphericalSurface_HasBeenSet
+     after AHFinderDirect_find_horizons
      {
      lang: C
      options: global
      } "store apparent horizon(s) into spherical surface(s)"
 
   schedule AHFinderDirect_save at CCTK_POST_RECOVER_VARIABLES \
-     after AHFinderDirect_find_horizons \
+     after AHFinderDirect_find_horizons
      {
      lang: C
      options: global

