Index: param.ccl
===================================================================
--- param.ccl	(revision 73)
+++ param.ccl	(working copy)
@@ -33,59 +33,59 @@
 
 
 
-KEYWORD evolution_method "The metric an extrinsic curvature evolution method"
+KEYWORD evolution_method "The metric an extrinsic curvature evolution method" STEERABLE=recover
 {
   "none"   :: "The metric and extrinsic curvature are not evolved"
   "static" :: "The metric and extrinsic curvature are not evolved"
 } "static"
 
-KEYWORD lapse_evolution_method "The lapse evolution method"
+KEYWORD lapse_evolution_method "The lapse evolution method" STEERABLE=recover
 {
   "static" :: "lapse is not evolved"
 } "static"
 
-KEYWORD shift_evolution_method "The shift evolution method"
+KEYWORD shift_evolution_method "The shift evolution method" STEERABLE=recover
 { 
   "static" :: "shift is not evolved"
 } "static"
 
-KEYWORD dtlapse_evolution_method "The dtlapse evolution method"
+KEYWORD dtlapse_evolution_method "The dtlapse evolution method" STEERABLE=recover
 {
   "static" :: "dtlapse is not evolved"
 } "static"
 
-KEYWORD dtshift_evolution_method "The dtshift evolution method"
+KEYWORD dtshift_evolution_method "The dtshift evolution method" STEERABLE=recover
 { 
   "static" :: "shift is not evolved"
 } "static"
 
 
 
-STRING admbase_boundary_condition "Boundary condition for ADMBase variables"
+STRING admbase_boundary_condition "Boundary condition for ADMBase variables" STEERABLE=recover
 {
   "" :: "must be a registered boundary condition"
 } "flat"
 
 
 
-KEYWORD metric_type "The semantics of the metric variables (physical, static conformal, etc)"
+KEYWORD metric_type "The semantics of the metric variables (physical, static conformal, etc)" STEERABLE=recover
 {
   "physical" :: "metric and extrinsic curvature are the physical ones"
 } "physical"
 
-KEYWORD lapse_prolongation_type "The kind of boundary prolongation for the lapse"
+KEYWORD lapse_prolongation_type "The kind of boundary prolongation for the lapse" STEERABLE=recover
 {
   "Lagrange" :: "standard prolongation (requires several time levels)"
   "none"     :: "no prolongation (use this if you do not have enough time levels active)"
 } "Lagrange"
 
-KEYWORD shift_prolongation_type "The kind of boundary prolongation for the shift"
+KEYWORD shift_prolongation_type "The kind of boundary prolongation for the shift" STEERABLE=recover
 {
   "Lagrange" :: "standard prolongation (requires several time levels)"
   "none"     :: "no prolongation (use this if you do not have enough time levels active)"
 } "Lagrange"
 
-KEYWORD metric_prolongation_type "The kind of boundary prolongation for the metric and extrinsic curvature"
+KEYWORD metric_prolongation_type "The kind of boundary prolongation for the metric and extrinsic curvature" STEERABLE=recover
 {
   "Lagrange" :: "standard prolongation (requires several time levels)"
   "none"     :: "no prolongation (use this if you do not have enough time levels active)"
