Index: doc/MaintGuide.pdf
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/pdf
Index: doc/ReferenceManual.pdf
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/pdf
Index: src/main/Parameters.c
===================================================================
--- src/main/Parameters.c	(revision 4919)
+++ src/main/Parameters.c	(working copy)
@@ -2244,6 +2244,7 @@
 
   if (!retval)
   {
+    retval = -1;
     for (range = param->props->range; range; range = range->next)
     {
       if (CCTK_IsThornActive (range->origin) ||
