diff --git a/AHFinderDirect/schedule.ccl b/AHFinderDirect/schedule.ccl
index d83a8af..d695a39 100644
--- a/AHFinderDirect/schedule.ccl
+++ b/AHFinderDirect/schedule.ccl
@@ -101,6 +101,7 @@ if (run_at_CCTK_ANALYSIS != 0)
      before AHFinderDirect_find_horizons
      {
      lang: C
+     WRITES: AHFINDERDIRECT::ahmask(EVERYWHERE)
      options: global-early loop-local
      } "import the excision mask"
 
@@ -142,6 +143,7 @@ if (run_at_CCTK_POSTSTEP != 0)
      before AHFinderDirect_find_horizons
      {
      lang: C
+     WRITES: AHFINDERDIRECT::ahmask(EVERYWHERE)
      options: global-early loop-local
      } "import the excision mask"
 
@@ -183,6 +185,7 @@ if (run_at_CCTK_POSTINITIAL != 0)
      before AHFinderDirect_find_horizons
      {
      lang: C
+     WRITES: AHFINDERDIRECT::ahmask(EVERYWHERE)
      options: global-early loop-local
      } "import the excision mask"
 
@@ -224,6 +227,7 @@ if (run_at_CCTK_POSTPOSTINITIAL != 0)
      before AHFinderDirect_find_horizons
      {
      lang: C
+     WRITES: AHFINDERDIRECT::ahmask(EVERYWHERE)
      options: global-early loop-local
      } "import the excision mask"
 
@@ -264,6 +268,7 @@ if (run_at_CCTK_POST_RECOVER_VARIABLES != 0)
      before AHFinderDirect_find_horizons
      {
      lang: C
+     WRITES: AHFINDERDIRECT::ahmask(EVERYWHERE)
      options: global-early loop-local
      } "import the excision mask"
 
