From c98e91c7ec4a58df727faa456512ba3795265be3 Mon Sep 17 00:00:00 2001
From: Josh Faber <jfaberrit@jfaber3.local>
Date: Thu, 18 Oct 2012 11:14:26 -0400
Subject: [PATCH 5/9] ng the interface for magentized con2prim for consistency

---
 interface.ccl |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/interface.ccl b/interface.ccl
index 06d9a63..8a4c78a 100644
--- a/interface.ccl
+++ b/interface.ccl
@@ -68,10 +68,10 @@ void FUNCTION Con2PrimGen(CCTK_INT INOUT handle, CCTK_REAL INOUT dens, \
                            CCTK_INT INOUT GRHydro_reflevel, \
                            CCTK_REAL OUT retval)
 
-void FUNCTION Con2PrimGenM(CCTK_INT INOUT handle, CCTK_REAL INOUT gamma_eos, CCTK_REAL INOUT dens, \
+void FUNCTION Con2PrimGenM(CCTK_INT INOUT handle, CCTK_INT INOUT keytemp, CCTK_REAL INOUT prec, CCTK_REAL INOUT gamma_eos, CCTK_REAL INOUT dens, \
                            CCTK_REAL INOUT sx, CCTK_REAL INOUT sy, CCTK_REAL INOUT sz, \
                            CCTK_REAL INOUT tau, CCTK_REAL INOUT Bconsx, CCTK_REAL INOUT Bconsy, CCTK_REAL INOUT Bconsz, \
-			   CCTK_REAL INOUT rho, \
+			   CCTK_REAL INOUT y_e, CCTK_REAL INOUT temp, CCTK_REAL INOUT rho, \
                            CCTK_REAL INOUT velx, CCTK_REAL INOUT vely, CCTK_REAL INOUT velz, \
                            CCTK_REAL INOUT epsilon, CCTK_REAL INOUT pressure, \
                            CCTK_REAL INOUT Bvecx, CCTK_REAL INOUT Bvecy, CCTK_REAL INOUT Bvecz, \
-- 
1.7.10.4

