diff -r 33df321567d2 Carpet/CarpetEvolutionMask/src/evolution_mask.cc
--- a/Carpet/CarpetEvolutionMask/src/evolution_mask.cc	Tue Mar 20 09:46:24 2012 -0700
+++ b/Carpet/CarpetEvolutionMask/src/evolution_mask.cc	Mon Apr 30 01:42:44 2012 -0700
@@ -22,20 +22,6 @@
   {
     DECLARE_CCTK_PARAMETERS;
     
-    // ask the flesh what the value of carpet::buffer_width is...
-    // this should be fixed at some point...
-
-    CCTK_INT use_outer_buffer_zones = *(CCTK_INT*) CCTK_ParameterGet
-      ("use_outer_buffer_zones","Carpet",0);
-
-    CCTK_INT buffer_width;
-    if (use_outer_buffer_zones) {
-      buffer_width = 0;
-    } else {
-      buffer_width = *(CCTK_INT*) CCTK_ParameterGet
-        ("buffer_width","Carpet",0);
-    }
-    // cout << "buffer width: " << buffer_width << endl;
     
     if (! is_singlemap_mode()) {
       CCTK_WARN (0, "This routine may only be called in singlemap mode");
@@ -57,6 +43,8 @@
       ivect const reffact
         = spacereffacts.at(reflevel) / spacereffacts.at(reflevel-1);
       assert (all (reffact == 2));
+
+      const i2vect buffer_widths = dd.buffer_widths.at(reflevel);
       
       // cout << "base: " << base << endl;
       // cout << "coarsebase: " << coarsebase << endl;
@@ -96,12 +84,14 @@
 
       // now make antibase larger
       // make refined regions SMALLER
-      ivect antishrinkby;
-      for (int d=0; d<dim;d++) {
-	antishrinkby[d] = cctkGH->cctk_nghostzones[d] + buffer_width;
+      i2vect antishrinkby;
+      for (int f=0; f<2;f++) {
+        for (int d=0; d<dim;d++) {
+	  antishrinkby[f][d] = cctkGH->cctk_nghostzones[d] + buffer_widths[f][d];
+        }
       }
       ibset antishrunk
-        (antirefined.expand(antishrinkby, antishrinkby).
+        (antirefined.expand(antishrinkby).
          expanded_for(coarsebase));
       
       //      cout << "antishrunk1: " << antishrunk << endl;
@@ -129,13 +119,15 @@
 
       //      cout << "notrefined: " << notrefined << endl;
 
-      ivect enlargeby;
+      i2vect enlargeby;
       int stencil_size = dd.prolongation_stencil_size(reflevel);
-      for (int d=0; d<dim;d++) {
-	enlargeby[d] = cctkGH->cctk_nghostzones[d] + buffer_width +
-	  stencil_size;
+      for (int f=0; f<2;f++) {
+        for (int d=0; d<dim;d++) {
+	  enlargeby[f][d] = cctkGH->cctk_nghostzones[d] + buffer_widths[f][d] +
+	    stencil_size;
+        }
       }
-      ibset const enlarged (notrefined.expand(enlargeby, enlargeby));
+      ibset const enlarged (notrefined.expand(enlargeby));
 
       //      cout << "enlarged: " << enlarged << endl;
       
diff -r 33df321567d2 Carpet/CarpetEvolutionMask/test/CarpetEvolutionMask_test.par
--- a/Carpet/CarpetEvolutionMask/test/CarpetEvolutionMask_test.par	Tue Mar 20 09:46:24 2012 -0700
+++ b/Carpet/CarpetEvolutionMask/test/CarpetEvolutionMask_test.par	Mon Apr 30 01:42:44 2012 -0700
@@ -1,14 +1,14 @@
 
-!DESC "1+log sliced black hole, ADM_BSSN, Iterative CN, Carpet with 3 levels"
+!DESC "1+log sliced black hole, ML_BSSN, Iterative CN, Carpet with 3 levels"
 
 # Required thorns
 
-ActiveThorns = "CoordBase SymBase time carpet carpetlib carpetregrid GSL LoopControl carpetreduce carpetslab boundary cartgrid3d bssn_mol mol nanchecker idanalyticbh carpetioascii ioutil carpetiobasic carpetioscalar admcoupling admbase admmacros coordgauge spacemask staticconformal carpetevolutionmask"
+ActiveThorns = "CoordBase SymBase InitBase time carpet carpetlib carpetregrid GSL LoopControl carpetreduce carpetslab boundary cartgrid3d ml_bssn ml_bssn_helper genericfd NewRad TmunuBase mol nanchecker idanalyticbh carpetioascii ioutil carpetiobasic carpetioscalar admcoupling admbase admmacros coordgauge spacemask staticconformal carpetevolutionmask ReflectionSymmetry"
 
 # CarpetEvolutionMask 
 
 carpetevolutionmask::verbose = "yes"
-carpetevolutionmask::enforce_vars = "adm_bssn::adm_bs_gxx adm_bssn::adm_bs_Axx adm_bssn::adm_bs_phi"
+carpetevolutionmask::enforce_vars = "ml_bssn::gt11 ml_bssn::At11 ml_bssn::phi"
 carpetevolutionmask::enforce_mask = "yes"
 carpetevolutionmask::writeNaNs = "yes"
 
@@ -24,7 +24,8 @@
 Carpet::prolongation_order_time = 2
 
 Carpet::init_each_timelevel = yes
-Carpet::buffer_width = 6
+InitBase::initial_data_setup_method="init_single_level"
+Carpet::use_buffer_zones = yes
 
 
 coordbase::xmin					= 0.0
@@ -41,15 +42,25 @@
 CarpetRegrid::regrid_every = 0
 
 cartgrid3d::type                        	= "coordbase"
-cartgrid3d::domain                      	= "bitant"
+cartgrid3d::domain                      	= "full"
 cartgrid3d::avoid_origin			= "no"
 
+ReflectionSymmetry::reflection_x                = "yes"
+ReflectionSymmetry::reflection_y                = "yes"
+ReflectionSymmetry::reflection_z                = "yes"
+ReflectionSymmetry::avoid_origin_x              = "no"
+ReflectionSymmetry::avoid_origin_y              = "no"
+ReflectionSymmetry::avoid_origin_z              = "no"
+
 Carpet::domain_from_coordbase 			= "yes"
 CoordBase::boundary_size_z_lower     		= 2
+CoordBase::boundary_size_z_upper     		= 2
 CoordBase::boundary_shiftout_z_lower 		= 1
 CoordBase::boundary_size_y_lower     		= 2
+CoordBase::boundary_size_y_upper     		= 2
 CoordBase::boundary_shiftout_y_lower 		= 1
 CoordBase::boundary_size_x_lower     		= 2
+CoordBase::boundary_size_x_upper     		= 2
 CoordBase::boundary_shiftout_x_lower 		= 1
 
 
@@ -68,21 +79,33 @@
 
 # Initial data 
 
-admbase::metric_type = "static conformal"
+admbase::metric_type = "physical"
 admbase::initial_data = "schwarzschild"
 
 # Evolution
 
-admbase::evolution_method = "adm_bssn" 
+admbase::evolution_method = "ML_BSSN" 
 
 MoL::ODE_Method = "ICN"
 MoL::MoL_Intermediate_Steps = 3
 
-adm_bssn::bound  = "flat"
+ML_BSSN::my_initial_boundary_condition = "extrapolate-gammas"
+ML_BSSN::ML_log_confac_bound = "flat"
+ML_BSSN::ML_metric_bound     = "flat"
+ML_BSSN::ML_Gamma_bound      = "flat"
+ML_BSSN::ML_trace_curv_bound = "flat"
+ML_BSSN::ML_curv_bound       = "flat"
+ML_BSSN::ML_lapse_bound      = "flat"
+ML_BSSN::ML_dtlapse_bound    = "flat"
+ML_BSSN::ML_shift_bound      = "flat"
+ML_BSSN::ML_dtshift_bound    = "flat"
+ML_BSSN::fdOrder             = 2
 
 # Gauge
 
-admbase::lapse_evolution_method = "1+log"
+ADMBase::lapse_evolution_method   = "ML_BSSN"
+ADMBase::dtlapse_evolution_method = "ML_BSSN"
+ML_BSSN::harmonicN = 1   # 1+log
 
 # Output
 
@@ -97,13 +120,13 @@
 carpetioascii::out2D_every = 1
 carpetioascii::out3D_every = -1
 
-carpetioascii::out1D_vars = "adm_bssn::adm_bs_gxx carpetevolutionmask::evolution_mask"
-carpetioascii::out2D_vars = "adm_bssn::adm_bs_gxx carpetevolutionmask::evolution_mask"
-carpetioascii::out3D_vars = "adm_bssn::adm_bs_gxx carpetevolutionmask::evolution_mask"
+carpetioascii::out1D_vars = "ml_bssn::gt11 carpetevolutionmask::evolution_mask"
+carpetioascii::out2D_vars = "ml_bssn::gt11 carpetevolutionmask::evolution_mask"
+carpetioascii::out3D_vars = "ml_bssn::gt11 carpetevolutionmask::evolution_mask"
 
 
 IOScalar::outScalar_every      =  4
-IOScalar::outScalar_vars       = "adm_bssn::adm_bs_gxx adm_bssn::adm_bs_Axx adm_bssn::adm_bs_phi adm_bssn::adm_bs_k adm_bssn::adm_bs_Gx admbase::gxx admbase::kxx admbase::alp"
+IOScalar::outScalar_vars       = "ml_bssn::gt11 ml_bssn::At11 ml_bssn::phi ml_bssn::trK ml_bssn::Xt1 admbase::gxx admbase::kxx admbase::alp"
 IOScalar::outScalar_reductions = "count minimum maximum average norm1 norm2 norm_inf"
 
 
diff -r 33df321567d2 Carpet/CarpetEvolutionMask/test/CarpetEvolutionMask_test_off.par
--- a/Carpet/CarpetEvolutionMask/test/CarpetEvolutionMask_test_off.par	Tue Mar 20 09:46:24 2012 -0700
+++ b/Carpet/CarpetEvolutionMask/test/CarpetEvolutionMask_test_off.par	Mon Apr 30 01:42:44 2012 -0700
@@ -1,14 +1,14 @@
 
-!DESC "1+log sliced black hole, ADM_BSSN, Iterative CN, Carpet with 3 levels"
+!DESC "1+log sliced black hole, ML_BSSN, Iterative CN, Carpet with 3 levels"
 
 # Required thorns
 
-ActiveThorns = "CoordBase SymBase time carpet carpetlib carpetregrid GSL LoopControl carpetreduce carpetslab boundary cartgrid3d bssn_mol mol nanchecker idanalyticbh carpetioascii ioutil carpetiobasic carpetioscalar admcoupling admbase admmacros coordgauge spacemask staticconformal carpetevolutionmask"
+ActiveThorns = "CoordBase SymBase InitBase time carpet carpetlib carpetregrid GSL LoopControl carpetreduce carpetslab boundary cartgrid3d ml_bssn ml_bssn_helper genericfd NewRad TmunuBase mol nanchecker idanalyticbh carpetioascii ioutil carpetiobasic carpetioscalar admcoupling admbase admmacros coordgauge spacemask staticconformal carpetevolutionmask ReflectionSymmetry"
 
 # CarpetEvolutionMask 
 
 carpetevolutionmask::verbose = "yes"
-carpetevolutionmask::enforce_vars = "adm_bssn::adm_bs_gxx adm_bssn::adm_bs_Axx adm_bssn::adm_bs_phi"
+carpetevolutionmask::enforce_vars = "ml_bssn::gt11 ml_bssn::At11 ml_bssn::phi"
 carpetevolutionmask::enforce_mask = "no"
 carpetevolutionmask::writeNaNs = "no"
 
@@ -24,7 +24,8 @@
 Carpet::prolongation_order_time = 2
 
 Carpet::init_each_timelevel = yes
-Carpet::buffer_width = 6
+InitBase::initial_data_setup_method="init_single_level"
+Carpet::use_buffer_zones = yes
 
 
 coordbase::xmin					= 0.0
@@ -41,15 +42,25 @@
 CarpetRegrid::regrid_every = 0
 
 cartgrid3d::type                        	= "coordbase"
-cartgrid3d::domain                      	= "bitant"
+cartgrid3d::domain                      	= "full"
 cartgrid3d::avoid_origin			= "no"
 
+ReflectionSymmetry::reflection_x                = "yes"
+ReflectionSymmetry::reflection_y                = "yes"
+ReflectionSymmetry::reflection_z                = "yes"
+ReflectionSymmetry::avoid_origin_x              = "no"
+ReflectionSymmetry::avoid_origin_y              = "no"
+ReflectionSymmetry::avoid_origin_z              = "no"
+
 Carpet::domain_from_coordbase 			= "yes"
 CoordBase::boundary_size_z_lower     		= 2
+CoordBase::boundary_size_z_upper     		= 2
 CoordBase::boundary_shiftout_z_lower 		= 1
 CoordBase::boundary_size_y_lower     		= 2
+CoordBase::boundary_size_y_upper     		= 2
 CoordBase::boundary_shiftout_y_lower 		= 1
 CoordBase::boundary_size_x_lower     		= 2
+CoordBase::boundary_size_x_upper     		= 2
 CoordBase::boundary_shiftout_x_lower 		= 1
 
 
@@ -68,21 +79,33 @@
 
 # Initial data 
 
-admbase::metric_type = "static conformal"
+admbase::metric_type = "physical"
 admbase::initial_data = "schwarzschild"
 
 # Evolution
 
-admbase::evolution_method = "adm_bssn" 
+admbase::evolution_method = "ML_BSSN" 
 
 MoL::ODE_Method = "ICN"
 MoL::MoL_Intermediate_Steps = 3
 
-adm_bssn::bound  = "flat"
+ML_BSSN::my_initial_boundary_condition = "extrapolate-gammas"
+ML_BSSN::ML_log_confac_bound = "flat"
+ML_BSSN::ML_metric_bound     = "flat"
+ML_BSSN::ML_Gamma_bound      = "flat"
+ML_BSSN::ML_trace_curv_bound = "flat"
+ML_BSSN::ML_curv_bound       = "flat"
+ML_BSSN::ML_lapse_bound      = "flat"
+ML_BSSN::ML_dtlapse_bound    = "flat"
+ML_BSSN::ML_shift_bound      = "flat"
+ML_BSSN::ML_dtshift_bound    = "flat"
+ML_BSSN::fdOrder             = 2
 
 # Gauge
 
-admbase::lapse_evolution_method = "1+log"
+ADMBase::lapse_evolution_method   = "ML_BSSN"
+ADMBase::dtlapse_evolution_method = "ML_BSSN"
+ML_BSSN::harmonicN = 1   # 1+log
 
 # Output
 
@@ -97,13 +120,15 @@
 carpetioascii::out2D_every = 1
 carpetioascii::out3D_every = -1
 
-carpetioascii::out1D_vars = "adm_bssn::adm_bs_gxx carpetevolutionmask::evolution_mask"
-carpetioascii::out2D_vars = "adm_bssn::adm_bs_gxx carpetevolutionmask::evolution_mask"
-carpetioascii::out3D_vars = "adm_bssn::adm_bs_gxx carpetevolutionmask::evolution_mask"
+carpetioascii::out1D_vars = "ml_bssn::gt11 carpetevolutionmask::evolution_mask"
+carpetioascii::out2D_vars = "ml_bssn::gt11 carpetevolutionmask::evolution_mask"
+carpetioascii::out3D_vars = "ml_bssn::gt11 carpetevolutionmask::evolution_mask"
 
 
-IOScalar::outScalar_every =  4
-IOScalar::outScalar_vars  = "adm_bssn::adm_bs_gxx adm_bssn::adm_bs_Axx adm_bssn::adm_bs_phi adm_bssn::adm_bs_k adm_bssn::adm_bs_Gx admbase::gxx admbase::kxx admbase::alp"
+IOScalar::outScalar_every      =  4
+IOScalar::outScalar_vars       = "ml_bssn::gt11 ml_bssn::At11 ml_bssn::phi ml_bssn::trK ml_bssn::Xt1 admbase::gxx admbase::kxx admbase::alp"
+IOScalar::outScalar_reductions = "count minimum maximum average norm1 norm2 norm_inf"
+
 
 ##################################################################
 
