Index: lib/sbin/CreateConfigurationBindings.pl
===================================================================
--- lib/sbin/CreateConfigurationBindings.pl	(revision 4862)
+++ lib/sbin/CreateConfigurationBindings.pl	(working copy)
@@ -211,9 +211,9 @@
 
   # write cctk_Capabilities.h file to bindings/include
   # this file adds the if_i_am_thorn stuff
-  $temp = '';
   foreach $thorn (sort keys %thorns)
   {
+    $temp = '';
     if ($cfg->{"\U$thorn\E REQUIRES"})
     {
       $temp .= "#include \"../Configuration/Thorns/cctki_$thorn.h\"\n";
