From 788722a99438eee22c710a6dde91686092cb38dc Mon Sep 17 00:00:00 2001
From: Roland Haas <rhaas@tapir.caltech.edu>
Date: Tue, 5 Mar 2013 09:27:18 -0800
Subject: [PATCH 03/11] GRHydro: convert more tabs to (8) spaces. No other
 changes.

---
 src/GRHydro_HLLE.F90 |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/GRHydro_HLLE.F90 b/src/GRHydro_HLLE.F90
index da262d0..4b578ee 100644
--- a/src/GRHydro_HLLE.F90
+++ b/src/GRHydro_HLLE.F90
@@ -114,13 +114,13 @@ subroutine H_viscosity(CCTK_ARGUMENTS)
         if (evolve_temper.ne.1) then
 
            call EOS_Omni_press(GRHydro_eos_handle,keytemp,GRHydro_eos_rf_prec,n,&
-	      rho(i,j,k),eps(i,j,k),xtemp,xye,xpress,keyerr,anyerr)
+              rho(i,j,k),eps(i,j,k),xtemp,xye,xpress,keyerr,anyerr)
 
-	    call EOS_Omni_DPressByDEps(GRHydro_eos_handle,keytemp,GRHydro_eos_rf_prec,n,&
-	      rho(i,j,k),eps(i,j,k),xtemp,xye,dpdeps,keyerr,anyerr)
+            call EOS_Omni_DPressByDEps(GRHydro_eos_handle,keytemp,GRHydro_eos_rf_prec,n,&
+              rho(i,j,k),eps(i,j,k),xtemp,xye,dpdeps,keyerr,anyerr)
 
            call EOS_Omni_DPressByDRho(GRHydro_eos_handle,keytemp,GRHydro_eos_rf_prec,n,&
-	      rho(i,j,k),eps(i,j,k),xtemp,xye,dpdrho,keyerr,anyerr)
+              rho(i,j,k),eps(i,j,k),xtemp,xye,dpdrho,keyerr,anyerr)
 
            cs2 = (dpdrho + xpress(1) * dpdeps / (rho(i,j,k)**2))/ &
               (1.0d0 + eps(i,j,k) + xpress(1)/rho(i,j,k))
-- 
1.7.10.4

