Index: lib/make/make.config.defn.in
===================================================================
--- lib/make/make.config.defn.in	(revision 4806)
+++ lib/make/make.config.defn.in	(working copy)
@@ -146,6 +146,7 @@
 # Dependency file generation
 export C_DEPEND             = @C_DEPEND@
 export CXX_DEPEND           = @CXX_DEPEND@
+export CL_DEPEND            = @CL_DEPEND@
 export F77_DEPEND           = @F77_DEPEND@
 export F_DEPEND             = @F_DEPEND@
 export F_DEPEND_MODULES     = @F_DEPEND_MODULES@
@@ -153,6 +154,7 @@
 
 export C_DEPEND_OUT         = @C_DEPEND_OUT@
 export CXX_DEPEND_OUT       = @CXX_DEPEND_OUT@
+export CL_DEPEND_OUT        = @CL_DEPEND_OUT@
 export F77_DEPEND_OUT       = @F77_DEPEND_OUT@
 export F_DEPEND_OUT         = @F_DEPEND_OUT@
 export F_DEPEND_MODULES_OUT = @F_DEPEND_MODULES_OUT@
@@ -167,6 +169,7 @@
 # How to postprocess object files
 export C_POSTPROCESSING   = @C_POSTPROCESSING@
 export CXX_POSTPROCESSING = @CXX_POSTPROCESSING@
+export CL_POSTPROCESSING   = @CL_POSTPROCESSING@
 export F77_POSTPROCESSING = @F77_POSTPROCESSING@
 export F_POSTPROCESSING   = @F_POSTPROCESSING@
 export F90_POSTPROCESSING = @F90_POSTPROCESSING@
Index: lib/make/force-reconfigure
===================================================================
--- lib/make/force-reconfigure	(revision 4806)
+++ lib/make/force-reconfigure	(working copy)
@@ -30,3 +30,4 @@
 30 Mar 2011: Add variables MAKE, TAR, GZIP_CMD, PATCH, GIT
 26 Apr 2011: Add CCTK_BUILTIN_EXPECT
 27 Jun 2011: Add HAVE_SCHED_H
+01 May 2012: Add OpenCL support
Index: lib/make/make.config.rules.in
===================================================================
--- lib/make/make.config.rules.in	(revision 4806)
+++ lib/make/make.config.rules.in	(working copy)
@@ -83,8 +83,14 @@
 	$(DEPENDENCY_FIXER)
 endef
 
+define CL_DEPENDENCIES
+# An OpenCL source file has no external dependencies
+	$(CL_DEPEND) $< $(CL_DEPEND_OUT)
+	$(DEPENDENCY_FIXER)
+endef
+
 define CU_DEPENDENCIES
-        $(CU_DEPEND) $< $(INCLUDE_LINE) $(EXTRA_DEFINES:%=-D%) -DCCODE $(CU_DEPEND_OUT)
+        $(CUCC_DEPEND) $< $(INCLUDE_LINE) $(EXTRA_DEFINES:%=-D%) -DCCODE $(CUCC_DEPEND_OUT)
         $(DEPENDENCY_FIXER)
 endef
 
@@ -149,6 +155,26 @@
 $(CXX_POSTPROCESSING)
 endef
 
+# Define how to do an OpenCL compilation
+define PREPROCESS_CL
+# Wrap the OpenCL source code into a string, using the thorn name and
+# file name to create a variable name, and create a C file
+{								\
+    echo 'char const *const OpenCL_source_$(THORN)_$* =';	\
+    sed -e 's/\\/\\\\/g;s/"/\\"/g;s/\(.*\)/"\1\\n"/g';		\
+    echo ';';							\
+} < $^ > $(notdir $<).c
+endef
+
+define COMPILE_CL
+# Basically a C compile, but with a different source file name
+current_wd=`$(GET_WD)` ; cd $(SCRATCH_BUILD) ; $(CC) $(CPPFLAGS) $(CFLAGS) $(CCOMPILEONLY)$(OPTIONSEP)$$current_wd$(DIRSEP)$@ $$current_wd$(DIRSEP)$(notdir $<).c $(INCLUDE_LINE) $(EXTRA_DEFINES:%=-D%) -DCCODE
+endef
+
+define POSTPROCESS_CL
+$(CL_POSTPROCESSING)
+endef
+
 # Define how to do a CUDA compilation
 define PREPROCESS_CU
 { if test $(C_LINE_DIRECTIVES) = 'yes'; then echo '#line 1 "'$<'"'; fi; cat $<; } | $(PERL) -s $(C_FILE_PROCESSOR) -line_directives=$(C_LINE_DIRECTIVES) -source_file_name=$< $(CONFIG) > $(notdir $<)
@@ -283,6 +309,20 @@
 %.cxx.d: $(SRCDIR)/%.cxx
 	$(CXX_DEPENDENCIES)
 
+# Build rules for OpenCL
+
+%.cl.o: $(SRCDIR)/%.cl
+	$(NOTIFY_PREPROCESSING)
+	$(PREPROCESS_CL)
+	$(NOTIFY_COMPILING)
+	$(COMPILE_CL)
+	$(NOTIFY_POSTPROCESSING)
+	$(POSTPROCESS_CL)
+	$(NOTIFY_DIVIDER)
+
+%.cl.d: $(SRCDIR)/%.cl
+	$(CL_DEPENDENCIES)
+
 # Build rules for CUDA
 
 %.cu.o: $(SRCDIR)/%.cu
Index: lib/make/configure.in
===================================================================
--- lib/make/configure.in	(revision 4806)
+++ lib/make/configure.in	(working copy)
@@ -1186,6 +1186,14 @@
   PERL_CONFIGURE_SCRIPT=`echo $PERL_CONFIGURE_SCRIPT | sed 's,^/cygdrive/\(.\)/,\1:/,' | sed 's,^//\(.\)/,\1:/,'`
 fi
 
+# An OpenCL source file has no external dependencies
+AC_SUBST(CL_DEPEND)
+# This command will not output a correct dependency by itself (it
+# does not output a target), but the Cactus dependency fixer will
+# overwrite the target anyway with the correct information.
+: ${CL_DEPEND='echo :'}
+AC_SUBST(CL_DEPEND_OUT)
+: ${CL_DEPEND_OUT=' > $@'}
 
 AC_SUBST(BUILD_ACTIVETHORNS)
 BUILD_ACTIVETHORNS='$(CCTK_HOME)/lib/sbin/BuildActiveThorns.pl'
@@ -1208,6 +1216,14 @@
 : ${CCOMPILEONLY='-c -o'}
 : ${FCOMPILEONLY='-c -o'}
 
+# An OpenCL source file has no external dependencies
+AC_SUBST(CL_DEPEND)
+# This command will not output a correct dependency by itself (it
+# does not output a target), but the Cactus dependency fixer will
+# overwrite the target anyway with the correct information.
+: ${CL_DEPEND='echo :'}
+AC_SUBST(CL_DEPEND_OUT)
+: ${CL_DEPEND_OUT=' > $@'}
 
 # Set the createexe flag if it's not been set by now
 AC_SUBST(CREATEEXE)
@@ -1268,6 +1284,8 @@
 : ${C_POSTPROCESSING=''}
 AC_SUBST(CXX_POSTPROCESSING)
 : ${CXX_POSTPROCESSING=''}
+AC_SUBST(CL_POSTPROCESSING)
+: ${CL_POSTPROCESSING=''}
 AC_SUBST(CU_POSTPROCESSING)
 : ${CU_POSTPROCESSING=''}
 AC_SUBST(F77_POSTPROCESSING)
