Index: configure.sh
===================================================================
--- configure.sh	(revision 91)
+++ configure.sh	(working copy)
@@ -303,7 +303,7 @@
     test_szlib=1
 fi
 if [ $test_szlib -eq 0 ]; then
-    HDF5_LIB_DIRS="$HDF5_LIB_DIRS $LIBSZ_DIR"
+    HDF5_LIB_DIRS="$HDF5_LIB_DIRS ${LIBSZ_DIR}/lib"
     if [ $is_windows -eq 0 ]; then
         HDF5_LIBS="$HDF5_LIBS sz"
     else
@@ -318,7 +318,7 @@
     test_zlib=1
 fi
 if [ $test_zlib -eq 0 ]; then
-    HDF5_LIB_DIRS="$HDF5_LIB_DIRS $LIBZ_DIR"
+    HDF5_LIB_DIRS="$HDF5_LIB_DIRS ${LIBZ_DIR}/lib"
     if [ $is_windows -eq 0 ]; then
         HDF5_LIBS="$HDF5_LIBS z"
     else
