Index: einsteintoolkit.th
===================================================================
--- einsteintoolkit.th	(revision 194)
+++ einsteintoolkit.th	(working copy)
@@ -263,6 +263,14 @@
 
 !TARGET   = $ARR
 !TYPE     = git
+!AUTH_URL = git@github.com:rhaas80/Outflow.git
+!URL      = git://github.com/rhaas80/Outflow.git
+!REPO_PATH= ../$2
+!CHECKOUT = 
+EinsteinAnalysis/Outflow
+
+!TARGET   = $ARR
+!TYPE     = git
 !AUTH_URL = git@github.com:barrywardell/EinsteinExact
 !URL      = git://github.com/barrywardell/EinsteinExact
 !REPO_PATH= $2
