Index: lib/sim-build.py
===================================================================
--- lib/sim-build.py	(revision 2162)
+++ lib/sim-build.py	(working copy)
@@ -270,7 +270,7 @@
     storedThornList = simlib.BuildPath(configBase, "ThornList")
     hasStoredThornList = os.path.exists(storedThornList)
     
-    needThornList = not hasStoredThornList or build_reconfig
+    needThornList = not hasStoredThornList
     thornlist = simlib.GetThornList(needThornList)
     info("ThornList is: %s" % thornlist)
