From 7a8784816e24ae2b64c3c3c119bb28d6cb6fa740 Mon Sep 17 00:00:00 2001
From: Barry Wardell <barry.wardell@gmail.com>
Date: Fri, 17 Jun 2011 11:21:54 +0200
Subject: [PATCH 04/11] Enable use of LoopControl.

---
 m/WeylScal4.m |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/m/WeylScal4.m b/m/WeylScal4.m
index ba5a165..31a40d9 100644
--- a/m/WeylScal4.m
+++ b/m/WeylScal4.m
@@ -316,4 +316,5 @@ CreateKrancThornTT[groups, ".", "WeylScal4",
   RealParameters -> realParameters,
   IntParameters -> intParameters,
   InheritedImplementations -> {"admbase", "methodoflines"},
-  UseJacobian -> True];
+  UseJacobian -> True,
+  UseLoopControl -> True];
-- 
1.7.7.4

