From 9cd25cc4d127f7e005a2bd054b579f594edcc286 Mon Sep 17 00:00:00 2001
From: Ian Hinder <ian.hinder@aei.mpg.de>
Date: Tue, 28 Jun 2011 16:13:00 +0200
Subject: [PATCH] Abort on missing Fortran compiler

---
 BLAS.sh |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/BLAS.sh b/BLAS.sh
index b65e66c..613828a 100644
--- a/BLAS.sh
+++ b/BLAS.sh
@@ -62,6 +62,13 @@ if [ -z "${BLAS_DIR}" -o "${BLAS_DIR}" = 'BUILD' ]; then
     DONE_FILE=${SCRATCH_BUILD}/done/${THORN}
     BLAS_DIR=${INSTALL_DIR}
     
+    if [ "${F77}" = "none" ]; then
+        echo 'BEGIN ERROR'
+        echo "Building BLAS requires a fortran compiler, but there is none configured: F77 = $F77.  Aborting."
+        echo 'END ERROR'
+        exit 1
+    fi
+
 (
     exec >&2                    # Redirect stdout to stderr
     set -x                      # Output commands
-- 
1.7.5.4

