Index: utils/Scripts/GetComponents
===================================================================
--- utils/Scripts/GetComponents	(revision 698)
+++ utils/Scripts/GetComponents	(working copy)
@@ -367,6 +367,11 @@
         # turn off warnings for this statement, we do our own checking
         {no warnings 'all'; %rec = @pairs;}
 
+        # skip ignored thorns
+        if ($rec{"TYPE"} =~ /ignore/) {
+            next;
+        }
+        
         if (!defined($rec{"CHECKOUT"})) {
             WARN_nonfatal("Nothing will be checked out from $rec{URL}\n");
             next;
