Index: src/main/ActiveThorns.c
===================================================================
--- src/main/ActiveThorns.c	(revision 5096)
+++ src/main/ActiveThorns.c	(working copy)
@@ -977,7 +977,8 @@
             new_thorn = this_thorn->activates_thorns[i];
             if (! CCTK_IsThornActive(new_thorn))
             {
-              switch (Util_StringListAdd(new_thorns, new_thorn))
+              const int ierr = Util_StringListAdd(new_thorns, new_thorn);
+              switch (ierr)
               {
               case 0:
                 /* Thorn already scheduled for activation */
@@ -988,7 +989,8 @@
                 did_add_thorns = 1;
                 break;
               default:
-                CCTK_Warn(0, __LINE__, __FILE__, "Cactus", "Internal error");
+                CCTK_VError(__LINE__, __FILE__, "Cactus", "Internal error: %d",
+                            ierr);
               }
             }
           }
Index: src/util/StringList.c
===================================================================
--- src/util/StringList.c	(revision 5096)
+++ src/util/StringList.c	(working copy)
@@ -3,12 +3,13 @@
    @date      Mon May 21 16:55:14 2001
    @author    Tom Goodale
    @desc 
-   Stuff for fixed-size lists of sorted unique strings.
+   Stuff for lists of sorted unique strings.
    @enddesc
    @version $Header$
  @@*/
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 
 #ifndef TEST_STRINGLIST
 #include "cctk.h"
@@ -30,6 +31,7 @@
 /********************************************************************
  ********************* Local Routine Prototypes *********************
  ********************************************************************/
+static int Utili_StringListExpand(uStringList *list);
 
 /********************************************************************
  ********************* Other Routine Prototypes *********************
@@ -56,7 +58,7 @@
  
    @endhistory 
    @var     size
-   @vdesc   Size of the list
+   @vdesc   Initial size of the list
    @vtype   int
    @vio     in
    @vcomment 
@@ -132,6 +134,10 @@
   {
     if (list->max_size < list->fill + 1)
     {
+      Utili_StringListExpand(list);
+    }
+    if (list->max_size < list->fill + 1)
+    {
       retval = -2;
     }
     else
@@ -153,6 +159,10 @@
       {
         if (list->max_size < list->fill + 1)
         {
+          Utili_StringListExpand(list);
+        }
+        if (list->max_size < list->fill + 1)
+        {
           retval = -2;
         }
         else
@@ -186,6 +196,10 @@
     {
       if (list->max_size < list->fill + 1)
       {
+        Utili_StringListExpand(list);
+      }
+      if (list->max_size < list->fill + 1)
+      {
         retval = -2;
       }
       else
@@ -295,6 +309,62 @@
  *********************     Local Routines   *************************
  ********************************************************************/
 
+ /*@@
+   @routine    Util_StringListExpand
+   @date       Sat Mar 15 15:27:08 PDT 2014
+   @author     Roland Haas
+   @desc
+   Doubles size of a stringlist.
+   @enddesc
+   @calls
+   @calledby
+   @history
+
+   @endhistory
+   @var     this
+   @vdesc   The stringlist object
+   @vtype   uStringList
+   @vio     inout
+   @vcomment
+
+   @endvar
+
+   @returntype int
+   @returndesc
+   1 if memory could be allocated
+   0 on failure
+   @endreturndesc
+@@*/
+static int Utili_StringListExpand(uStringList *list)
+{
+  int retval;
+
+  struct iInternalStringList *temp =
+    realloc(list->list,(2*list->max_size+1)*sizeof(struct iInternalStringList));
+  if(temp)
+  {
+    memset(temp+list->max_size+1, 0,
+           list->max_size*sizeof(struct iInternalStringList));
+    /* re-link entries based on their position in the (old) array */
+    for(int i=0; i < list->fill; i++)
+    {
+      temp[i].next = temp[i].next ? temp+(temp[i].next-list->list) : NULL;
+    }
+    list->head = list->head ? temp+(list->head-list->list) : NULL;
+    list->current = list->current ? temp+(list->current-list->list) : NULL;
+    list->list = temp;
+    list->max_size = 2*list->max_size;
+
+    retval = 1;
+  }
+  else
+  {
+    retval = 0;
+  }
+
+  return retval;
+}
+
 #ifdef TEST_STRINGLIST
 
 int main(int argc, const char *argv[])
