Index: doc/ReferenceManual.pdf
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/pdf
Index: doc/ReferenceManual/CCTKReference.tex
===================================================================
--- doc/ReferenceManual/CCTKReference.tex	(revision 4804)
+++ doc/ReferenceManual/CCTKReference.tex	(working copy)
@@ -11095,6 +11095,69 @@
 
 
 
+% ScheduleInterface.c
+\begin{FunctionDescription}{CCTK\_ScheduleQueryCurrentFunction}
+  \label{CCTK-ScheduleQueryCurrentFunction}
+   Return the cFunctionData of the function currently executing via
+   \texttt{CCTK\_CallFunction}.
+  
+  \begin{SynopsisSection}
+    \begin{Synopsis}{C}
+\begin{verbatim}
+#include "cctk.h"
+const cFunctionData *CCTK_ScheduleQueryCurrentFunction(const cGH *GH)
+\end{verbatim}
+    \end{Synopsis}
+  \end{SynopsisSection}
+  
+  \begin{ResultSection}
+    \begin{ResultNote}
+      Data of last call to \texttt{CCTK\_CallFunction}, or
+    \end{ResultNote}
+    \begin{Result}{NULL}
+      if not within a scheduled function.
+    \end{Result}
+  \end{ResultSection}
+  
+  \begin{ParameterSection}
+    \begin{Parameter}{cctkGH}
+      Pointer to a Cactus grid hierarchy.
+    \end{Parameter}
+  \end{ParameterSection}
+  
+  \begin{Discussion}
+    Returns a data structure containing the thorn and routine name of the
+    currently executing function as well as the Cactus bin name. If no
+    function is currently executing, returns \code{NULL}. This is intended to
+    be used by thorns providing callable functions to identify their caller
+    when reporting errors.
+  \end{Discussion}
+  
+  \begin{SeeAlsoSection}
+    \begin{SeeAlso}{CCTK\_CallFunction}
+      Calls a function depending upon the data passed in the the fdata
+      structure.
+    \end{SeeAlso}
+  \end{SeeAlsoSection}
+
+  \begin{ExampleSection}
+    \begin{Example}{C}
+      Output the name of the currently scheduled function:
+\begin{verbatim}
+#include <stdio.h>
+#include "cctk.h"
+const cFunctionData *fdata = CCTK_QueryScheduledFunction(cctkGH);
+printf("scheduled function: %s::%s AT %s\n",
+       fdata->thorn, fdata->routine, fdata->where);
+\end{verbatim}
+\end{Example}
+  \end{ExampleSection}
+  
+\end{FunctionDescription}
+
+
+
+
 % CommOverloadables.c
 \begin{FunctionDescription}{CCTK\_SetupGH}{Setup a new GH}
 \label{CCTK-SetupGH}
Index: src/include/cctk_Schedule.h
===================================================================
--- src/include/cctk_Schedule.h	(revision 4804)
+++ src/include/cctk_Schedule.h	(working copy)
@@ -74,6 +74,8 @@
                       cFunctionData *fdata,
                       void *data);
 
+const cFunctionData *CCTK_ScheduleQueryCurrentFunction(const cGH *GH);
+
 int CCTK_ScheduleTraverse(const char *where,
                           void *GH,
                           int (*CallFunction)(void *, cFunctionData *, void *));
Index: src/main/ScheduleInterface.c
===================================================================
--- src/main/ScheduleInterface.c	(revision 4804)
+++ src/main/ScheduleInterface.c	(working copy)
@@ -233,7 +233,9 @@
 static cTimerData *timerinfo = NULL;
 static int total_timer = -1;
 
+static const cFunctionData *current_scheduled_function = NULL;
 
+
 /********************************************************************
  *********************     External Routines   **********************
  ********************************************************************/
@@ -279,6 +281,18 @@
 
   int (*oneargfunc)(void *);
 
+  if(current_scheduled_function != NULL)
+  {
+    CCTK_VWarn(CCTK_WARN_PICKY, __LINE__, __FILE__, "Cactus",
+               "CCTK_CallFunction: recursive call, calling "
+               "'%s: %s::%s' while within '%s: %s::%s'",
+               fdata->where, fdata->thorn, fdata->routine,
+               current_scheduled_function->where,
+               current_scheduled_function->thorn,
+               current_scheduled_function->routine);
+  }
+  current_scheduled_function = fdata;
+
   switch(fdata->type)
   {
     case FunctionNoArgs:
@@ -299,6 +313,9 @@
         case LangFortran:
           fdata->FortranCaller(data, function);
           break;
+        case LangNone:
+          /* this should never happen */
+          /* fall through */
         default :
           CCTK_Warn(1,__LINE__,__FILE__,"Cactus",
                     "CCTK_CallFunction: Unknown language.");
@@ -309,10 +326,39 @@
                 "CCTK_CallFunction: Unknown function type.");
   }
 
+  current_scheduled_function = NULL;
+
   /* Return 0, meaning didn't synchronise */
   return 0;
 }
 
+ /*@@
+   @routine    CCTK_ScheduleQueryCurrentFunction
+   @date       Fri Apr 20 08:57:49 PDT 2012
+   @author     Roland Haas
+   @desc
+   Returns the cFunctionData of the function currenlty executing via
+   CCTK_CallFunction.
+   @enddesc
+   @calls
+
+   @var     GH
+   @vdesc   GH data
+   @vtype   const cGH *
+   @vio     in
+   @endvar
+
+   @returntype cFunctionData *
+   @returndesc
+   Data about the function. NULL if no function is currently executing via
+   CCTK_CallFunction.
+   @endreturndesc
+@@*/
+const cFunctionData *CCTK_ScheduleQueryCurrentFunction(const cGH *GH)
+{
+  return current_scheduled_function;
+}
+
 /*@@
    @routine    CCTKi_ScheduleFunction
    @date       Thu Sep 16 18:19:01 1999
