Index: lib/sim.py
===================================================================
--- lib/sim.py	(revision 1350)
+++ lib/sim.py	(working copy)
@@ -17,7 +17,7 @@
 
 # import all option groups.
 
-optionGroups = ['sim-sync', 'sim-build', 'sim-manage', 'sim-info', 'sim-util']
+optionGroups = ['sim-build', 'sim-manage', 'sim-info', 'sim-util']
 
 
 ############################################
Index: lib/sim-sync.py
===================================================================
--- lib/sim-sync.py	(revision 1350)
+++ lib/sim-sync.py	(working copy)
@@ -7,7 +7,6 @@
 
 import sys, os
 
-optionGroups = ['sim-sync']
 Purpose = "sync cactus sourcetree to a remote machine"
 
 ############################################
@@ -49,7 +48,7 @@
     
     return data
 
-def CompileCommand(machineName, rsyncInfo):
+def CompileCommand(machineName, rsyncInfo, paths):
   
     
     global localMachineName
@@ -59,10 +58,6 @@
     
     DefineDatabase = simsubs.DefineDatabase()
 
-    # mode of operation
-    sync_parfiles = simenv.OptionsManager.GetOption('sync-parfiles')
-    sync_sourcetree = simenv.OptionsManager.GetOption('sync-sourcetree')
-
     (rsynccmd, rsyncopts) = rsyncInfo
     
     if machineName == localMachineName:
@@ -78,11 +73,8 @@
         machineEntry = simenv.ConfigurationDatabase.GetMachine(ioMachine)
         machineName = ioMachine
     
-    mreq_keys = ['user', 'sourcebasedir', 'hostname', 'rsynccmd', 'rsyncopts', 'sshcmd', 'sshopts']
-
-    sourcebasedir = machineEntry.sourcebasedir
-    
     # make sure we have all the info we need.
+    mreq_keys = ['user', 'sourcebasedir', 'hostname', 'rsynccmd', 'rsyncopts', 'sshcmd', 'sshopts']
     simlib.VerifyKeys(machineEntry, mreq_keys)
 
     DefineDatabase.Set('USER', machineEntry.user)
@@ -97,7 +89,7 @@
     #local_sourcebasedir = DefineDatabase.SubAll(localMachineEntry.sourcebasedir)
 
     source_name = local_suffix
-    
+
     #source_name = simlib.GetDirSuffix(local_sourcebasedir)
     local_sourcedir = simlib.BuildPath(local_sourcebasedir, source_name)
     DefineDatabase.Set('SOURCEDIR', local_sourcedir)
@@ -116,42 +108,17 @@
     sshcmd = simlib.GetSSHCommand(trampoline, sshcmd)
     sshcmd = DefineDatabase.SubAll(sshcmd)
     
+    # Strip trailing slashes from any explicit paths
     rsyncfiles = []
-
+    for file in paths:
+        rsyncfiles.append(file.strip('/'))
+	
     rsyncoptions = ['--archive', '--hard-links', '--sparse', '--verbose', '--progress', '--partial',
-                    '--stats', '--compress', '--delete', '--delete-excluded']
-        
-    # sync source tree
-    if sync_sourcetree:
-        sources = simenv.ConfigurationDatabase.GetConfigOption('sync-sources')
-        #dprint("sync-sources is: %s " % sources)
-        if type(sources) == list:
-            for r in sources:
-                rp = simlib.BuildPath(local_sourcedir, r)
-                #display("checking if %s exists" % rp)
-                if os.path.exists(rp):
-                    rsyncfiles.append(r)
+                    '--stats', '--compress', '--filter \'merge simfactory/etc/filter.rules\'']
     
-    # sync par files
-    if sync_parfiles:
-        sources = simenv.ConfigurationDatabase.GetConfigOption('sync-parfiles')
-        #dprint("sync-parfiles is: %s " % sources)
-        if type(sources) == list:
-            for r in sources:
-                rp = simlib.BuildPath(local_sourcedir, r)
-                #display("checking if %s exists" % rp)
-                if os.path.exists(rp):
-                    rsyncfiles.append(r)
+    if rsyncfiles == ["."]:
+        rsyncoptions.append('--filter \'merge simfactory/etc/filter.cactus.rules\'')
 
-    # build excludes
-    sources = simenv.ConfigurationDatabase.GetConfigOption('sync-excludes')
-    if type(sources) == list:
-        for r in sources:
-            r = r.strip()
-
-            if len(r) > 0:
-                rsyncoptions.append("--exclude '%s'" % r)
-
     fullpath = "%s@%s:%s%s%s" % (machineEntry.user, machineEntry.hostname, sourcebasedir, os.sep, local_suffix)
     
     arguments = " ".join(simlib.GetArguments())
@@ -180,37 +147,29 @@
     global local_sourcebasedir
     global localMachineName
     global localMachineEntry
-    
-    machineList = simenv.OptionsManager.args
-    
-    if len(machineList) == 0:
-        fatal("Error: no machines specified")
 
-    if simenv.OptionsManager.args[0] == 'sync':
+    if len(simenv.OptionsManager.args) > 0 and simenv.OptionsManager.args[0] == 'sync':
         simenv.OptionsManager.args.pop(0)
-        
-    info("using machines: %s\n" % machineList)
+
+    if len(simenv.OptionsManager.args) == 0:
+        fatal("Error: no machine specified")
+
+    machine = simenv.OptionsManager.args.pop(0)
+ 
+    paths = simenv.OptionsManager.args
+    if len(paths) == 0:
+        paths = ["."]
+
+    info("using machine: %s\n" % machine)
     info("current working directory: %s\n" % os.getcwd())
-    
+    info("synchronising paths: %s\n" % paths)
+
     if simenv.CACTUS_PATH == None:
         fatal("cannot proceed with an unknown CACTUS_PATH")
         
     cactusDir = simenv.CACTUS_PATH
     display("Cactus Directory: %s" % cactusDir)
-
-    # mode of operation
-    sync_parfiles = simenv.OptionsManager.GetOption('sync-parfiles')
-    sync_sourcetree = simenv.OptionsManager.GetOption('sync-sourcetree')
     
-    if not(sync_parfiles or sync_sourcetree):
-        fatal("neither sync-parfiles or sync-sourcetree was specified. should never happen")
-    
-    info("Enabled Options:")
-    info("sync-parfiles: %s" % sync_parfiles)
-    info("sync-sourcetree: %s" % sync_sourcetree)
-    
-    print
-    
     # get our local_sourcebasedir
     
     local_sourcebasedir = simlib.GetLocalSourceBaseDir()
@@ -225,15 +184,10 @@
     # returns list (rsynccmd, rsyncopts)
     rsyncInfo = GetRsyncInfo(localMachineEntry)
     
-    errors = 0
-    for m in machineList:
-        cmd = CompileCommand(m, rsyncInfo)
-        ret = simlib.ExecuteCommand(cmd)
-        if ret != 0:
-            errors = errors + 1
-            display("\nError while syncing to %s." % m)
-
-        if errors == 0:
-            display("\nSync complete.")
-        else:
-            display("\nErrors occurred during synchronisation.")
+    cmd = CompileCommand(machine, rsyncInfo, paths)
+    ret = simlib.ExecuteCommand(cmd)
+    
+    if ret == 0:
+        display("\nSync complete.")
+    else:
+        display("\nErrors occurred during synchronisation.")
Index: etc/defs.ini
===================================================================
--- etc/defs.ini	(revision 1350)
+++ etc/defs.ini	(working copy)
@@ -3,38 +3,4 @@
 # any <<EOT...EOT entries are converted internally to lists, since all
 # these options do actually represent lists of files
 
-# Official Cactus, SimFactory, and GetComponents entries
-sync-sources   = <<EOT
-CONTRIBUTORS
-COPYRIGHT
-Makefile
-arrangements
-bin
-lib
-manifest
-repos
-simfactory
-src
-utils
-EOT
-
-# Official Cactus, SimFactory, and GetComponents entries
-sync-parfiles  = <<EOT
-examples
-par
-EOT
-
-sync-excludes  = <<EOT
-_darcs
-CVS
-doxygen
-.#*
-.DS_Store
-.git
-.hg
-.svn
-*~
-EOT
-# *.pyc
-
 default-configuration-name = sim
Index: etc/filter.cactus.rules
===================================================================
--- etc/filter.cactus.rules	(revision 0)
+++ etc/filter.cactus.rules	(revision 0)
@@ -0,0 +1,20 @@
+# This file contains the list of files to include/exclude.  See "man
+# rsync" for details.  The first matching pattern is used, whether it
+# is an include or an exclude.  If no pattern matches, the file is
+# included.
+
+# Official Cactus, SimFactory, and GetComponents entries
++ /CONTRIBUTORS
++ /COPYRIGHT
++ /Makefile
++ /arrangements
++ /bin
++ /examples
++ /lib
++ /manifest
++ /par
++ /repos
++ /simfactory
++ /src
++ /utils
+- /*
Index: etc/options/sim-sync.ini
===================================================================
--- etc/options/sim-sync.ini	(revision 1350)
+++ etc/options/sim-sync.ini	(working copy)
@@ -1,17 +0,0 @@
-[sync-sourcetree]
-long = --sync-sourcetree
-#short
-desc = Synchronise the source tree
-negatable = yes
-arg = no
-default = True
-#argformat
-
-[sync-parfiles]
-long = --sync-parfiles
-#short 
-desc = Synchronise the parameter files
-negatable = yes
-arg = no
-default = True
-#argformat
Index: etc/defs.local.ini.complex
===================================================================
--- etc/defs.local.ini.complex	(revision 1350)
+++ etc/defs.local.ini.complex	(working copy)
@@ -1,20 +1,5 @@
 # -*-conf-*-
 
-# redefining sync-sources, sync-parfiles, and sync-excludes in this
-# file does not overwrite them, but rather appends these additional
-# files to the already defined entries provided by defs.ini
-
-sync-sources   = <<EOT
-.gitignore
-cactus.config
-EOT
-
-#sync-parfiles  = <<EOT
-#EOT
-
-#sync-excludes  = <<EOT
-#EOT
-
 # Defaults for all machines
 [default]
 user            = YOUR_LOGIN 
Index: etc/filter.rules
===================================================================
--- etc/filter.rules	(revision 0)
+++ etc/filter.rules	(revision 0)
@@ -0,0 +1,16 @@
+# This file contains the list of files to include/exclude.  See "man
+# rsync" for details.  The first matching pattern is used, whether it
+# is an include or an exclude.  If no pattern matches, the file is
+# included.
+
+# User specified filters
+: .rsync.rules
+
+# Files to exclude
+-C
+-p _darcs
+- doxygen
+- .DS_Store
+
+# Compiled simfactory python files
+-p *.pyc
