From f10049440ba3f3c9080d1d89fccb13b008f4f359 Mon Sep 17 00:00:00 2001
From: Barry Wardell <barry.wardell@gmail.com>
Date: Fri, 17 Jun 2011 11:19:50 +0200
Subject: [PATCH 02/11] Enable multipatch support.

---
 m/WeylScal4.m |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/m/WeylScal4.m b/m/WeylScal4.m
index 8f6e1de..1f5e499 100644
--- a/m/WeylScal4.m
+++ b/m/WeylScal4.m
@@ -309,4 +309,5 @@ CreateKrancThornTT[groups, ".", "WeylScal4",
   KeywordParameters -> keywordParameters,
   RealParameters -> realParameters,
   IntParameters -> intParameters,
-  InheritedImplementations -> {"admbase", "methodoflines"}];
+  InheritedImplementations -> {"admbase", "methodoflines"},
+  UseJacobian -> True];
-- 
1.7.7.4

