From 9831eda5ee81016ce054a5fb5825289abc522baf Mon Sep 17 00:00:00 2001
From: Roland Haas <rhaas@tapir.caltech.edu>
Date: Fri, 10 May 2013 13:09:07 -0700
Subject: [PATCH 14/24] GRHydro: whitespace change, tab->8 spaces

---
 src/Con2Prim_fortran_interfaces.F90 |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/Con2Prim_fortran_interfaces.F90 b/src/Con2Prim_fortran_interfaces.F90
index aac868b..fd32753 100644
--- a/src/Con2Prim_fortran_interfaces.F90
+++ b/src/Con2Prim_fortran_interfaces.F90
@@ -6,8 +6,8 @@ module Con2Prim_fortran_interfaces
   interface
 
      subroutine Con2Prim_pt(&
-	  cctk_iteration,ii,jj,kk,&
-	  handle, &
+          cctk_iteration,ii,jj,kk,&
+          handle, &
           dens, &
           sx, sy, sz, &
           tau, &
-- 
1.7.10.4

