Index: schedule.ccl
===================================================================
--- schedule.ccl	(revision 16)
+++ schedule.ccl	(working copy)
@@ -7,7 +7,8 @@
 
 # Initialise stuff
 
-SCHEDULE TmunuBase_SetStressEnergyState AT basegrid
+#SCHEDULE TmunuBase_SetStressEnergyState AT basegrid
+SCHEDULE TmunuBase_SetStressEnergyState AT Wragh BEFORE MoL_Register
 {
   LANG: Fortran
   OPTIONS: global
