diff --git a/m/McLachlan_BSSN.m b/m/McLachlan_BSSN.m
index b44af3b..63f5694 100644
--- a/m/McLachlan_BSSN.m
+++ b/m/McLachlan_BSSN.m
@@ -37,6 +37,7 @@ derivatives =
   PDstandardNth[i_,j_] -> StandardCenteredDifferenceOperator[1,fdOrder/2,i] *
                           StandardCenteredDifferenceOperator[1,fdOrder/2,j],
   PDdissipationNth[i_] ->
+    (-1)^(fdOrder/2) *
     spacing[i]^(fdOrder+1) / 2^(fdOrder+2) *
     StandardCenteredDifferenceOperator[fdOrder+2,fdOrder/2+1,i],
   
