Index: lib/sbin/CreateScheduleBindings.pl
===================================================================
--- lib/sbin/CreateScheduleBindings.pl	(revision 4701)
+++ lib/sbin/CreateScheduleBindings.pl	(working copy)
@@ -566,6 +566,17 @@
 
   $buffer .= "0};\n";
 
+  # add check on number of timelevels in case we were using a parameter
+  for($i=0 ; $i < (scalar @$tlist) ; $i++)
+  {
+    $buffer .= "    if($$tlist[$i] > CCTK_MaxTimeLevels(\"$$mem_groups[$i]\"))\n";
+    $buffer .= "        CCTK_VWarn(0, __LINE__, __FILE__, CCTK_THORNSTRING,\n";
+    $buffer .= "                   \"Tried to schedule %d timelevels for group '%s' in schedule.ccl.\\n\"\n";
+    $buffer .= "                   \"Value must be between 1 and %d (inclusive)\",\n";
+    $buffer .= "                   $$tlist[$i], \"$$mem_groups[$i]\", CCTK_MaxTimeLevels(\"$$mem_groups[$i]\"));\n";
+    $buffer .= "\n";
+  }
+
   # Start writing out the data
   if($rhschedule_db->{"\U$thorn\E BLOCK_$block TYPE"} eq "GROUP")
   {
@@ -951,7 +962,7 @@
 
 #    print "DEBUG: validate $thorn,$implementation,$group,$timelevels\n";
 
-    if($timelevels !~ /^\d*$/)
+    if($timelevels !~ /^[[:alpha:]_][[:word:]]*$/ && $timelevels !~ /^\d*$/)
     {
       &CST_error(0,"Invalid timelevel specifier '$timelevels' in schedule.ccl of thorn '$thorn'","",__LINE__,__FILE__);
       $return_code++;
@@ -993,7 +1004,7 @@
                  ,"",__LINE__,__FILE__);
       $return_code++;
     }
-    elsif($timelevels > 0 && $timelevels  <= $allowed_timelevels)
+    elsif($timelevels =~ /^[[:alpha:]_][[:word:]]*$/ || ($timelevels > 0 && $timelevels  <= $allowed_timelevels))
     {
       next;
     }
