Index: doc/UsersGuide/Notes.tex
===================================================================
--- doc/UsersGuide/Notes.tex	(revision 4799)
+++ doc/UsersGuide/Notes.tex	(working copy)
@@ -1354,12 +1354,14 @@
         \hline
         \verb|/| & C-like division		 & \verb|sinh| & hyperbolic sine \\
         \hline
-        \verb|*| & multiplication		 & \verb|sqrt| & square root \\
+        \verb|%| & remainder of division	 & \verb|sqrt| & square root \\
+        \hline                                                                         
+        \verb|*| & multiplication		 & \verb|tan| & tangent \\
+        \hline                                                                         
+        \verb|^| & exponentiation $x^y$		 & \verb|tanh| & hyperbolic tangent \\
+        \hline                                                                         
+                                       		 & \verb|trunc| & integer part of $x$\\
         \hline
-        \verb|^| & exponentiation $x^y$		 & \verb|tan| & tangent \\
-        \hline
-                 &                               & \verb|tanh| & hyperbolic tangent \\
-        \hline
     \end{tabular}
     \caption{Supported functions inside of expressions, in increasing order of
     precedence.}
Index: doc/UsersGuide.pdf
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/pdf
Index: src/include/utili_Expression.h
===================================================================
--- src/include/utili_Expression.h	(revision 4802)
+++ src/include/utili_Expression.h	(working copy)
@@ -28,6 +28,7 @@
               OP_PLUS,
               OP_MINUS,
               OP_DIV,
+              OP_REMAINDER,
               OP_TIMES,
               OP_POWER,
               OP_NOT,
@@ -48,7 +49,8 @@
               OP_SINH,
               OP_SQRT,
               OP_TAN,
-              OP_TANH}
+              OP_TANH,
+              OP_TRUNC}
   uExpressionOpcode;
 
   /* What sort of expression types we have. */
Index: src/main/Parameters.c
===================================================================
--- src/main/Parameters.c	(revision 4802)
+++ src/main/Parameters.c	(working copy)
@@ -2193,19 +2193,24 @@
   /* try parsing as number */
   inval = strtol (value, &endptr, 0);
 
-  if(*endptr) /* if we could not parse as a number, try an expression */
+  if (*endptr) /* if we could not parse as a number, try an expression */
   {
     int type = PARAMETER_INT;
-    int ierr;
     uExpressionValue val;
     uExpression *expr;
 
     expr = Util_ExpressionParse(value);
-    assert(expr);
-    ierr = Util_ExpressionEvaluate(expr, &val, SetVarEvaluator, &type);
-    Util_ExpressionFree(expr);
+    if (expr != NULL)
+    {
+      retval = Util_ExpressionEvaluate(expr, &val, SetVarEvaluator, &type);
+      Util_ExpressionFree(expr);
+    }
+    else
+    {
+      retval = -6;
+    }
 
-    if (ierr == 0)
+    if (retval == 0)
     {
       assert(val.type == ival || val.type == rval);
 
@@ -2260,7 +2265,7 @@
   {
     CCTK_VWarn (2, __LINE__, __FILE__, "Cactus",
                 "ParameterSetInteger: Unable to set integer '%s::%s' - '%s' "
-                "is not a valid integer",
+                "does not evaluate to a valid integer",
                 param->props->thorn, param->props->name, value);
   }
 
@@ -2299,16 +2304,21 @@
   if (*endptr) /* if we cannot parse as a number, try expression */
   {
     int type = PARAMETER_REAL;
-    int ierr;
     uExpressionValue val;
     uExpression *expr;
 
     expr = Util_ExpressionParse(value);
-    assert(expr);
-    ierr = Util_ExpressionEvaluate(expr, &val, SetVarEvaluator, &type);
-    Util_ExpressionFree(expr);
+    if (expr != NULL)
+    {
+      retval = Util_ExpressionEvaluate(expr, &val, SetVarEvaluator, &type);
+      Util_ExpressionFree(expr);
+    }
+    else
+    {
+      retval = -6;
+    }
 
-    if (ierr == 0)
+    if (retval == 0)
     {
       assert(val.type == ival || val.type == rval);
 
@@ -2357,8 +2367,8 @@
   else if (retval == -6)
   {
     CCTK_VWarn (2, __LINE__, __FILE__, "Cactus",
-                "ParameterSetInteger: Unable to set integer '%s::%s' - '%s' "
-                "is not a valid floating point number",
+                "ParameterSetInteger: Unable to set real '%s::%s' - '%s' "
+                "does not evaluate to a valid floating point number",
                 param->props->thorn, param->props->name, value);
   }
 
@@ -2376,14 +2386,20 @@
   /* first try parsing as yes/no/true/false */
   retval = CCTK_SetBoolean (&inval, value);
 
-  if(retval) /* if we cannot parse as a boolean, try expression */ 
+  if (retval < 0) /* if we cannot parse as a boolean, try expression */ 
   {
     expr = Util_ExpressionParse(value);
-    assert(expr);
-    int ierr = Util_ExpressionEvaluate(expr, &val, SetVarEvaluator, (void *)&type);
-    Util_ExpressionFree(expr);
+    if(expr != NULL)
+    {
+      retval = Util_ExpressionEvaluate(expr, &val, SetVarEvaluator, &type);
+      Util_ExpressionFree(expr);
+    }
+    else
+    {
+      retval = -6;
+    }
 
-    if (!ierr)
+    if (retval == 0)
     {
       assert(val.type == ival || val.type == rval);
       if (val.type == ival)
@@ -2409,12 +2425,11 @@
 
   if (!retval)
     *(CCTK_INT *)param->data = inval != 0;
-
-  if (retval == -1)
+  else
   {
     CCTK_VWarn (2, __LINE__, __FILE__, "Cactus",
                 "ParameterSetBoolean: Unable to set boolean '%s::%s' = '%s' "
-                "not recognised",
+                "not recognised as boolean",
                 param->props->thorn, param->props->name, value);
   }
 
@@ -2707,7 +2722,9 @@
           vals[i].type = ival;
           ierr = 0;
       }
-      else if (strpbrk(vars[i], "eDdD."))
+      else if (strpbrk(vars[i], "eDdD.") ||
+               Util_StrCmpi(vars[i], "nan") == 0 ||
+               Util_StrCmpi(vars[i], "inf") == 0)
       {
         /*
          * Canonicalize the string by converting all exponent letters
Index: src/util/Expression.c
===================================================================
--- src/util/Expression.c	(revision 4802)
+++ src/util/Expression.c	(working copy)
@@ -34,7 +34,7 @@
 #include "util_String.h"
 #define strdup(a) Util_Strdup(a)
 #else
-#define CCTK_VWarn(lvl, line, file, thorn, fmt, ...)  printf("(%s) L%d at %s line %d: "fmt"\n", thorn, lvl, file, line, ## __VA_ARGS__)
+#define CCTK_VWarn(lvl, line, file, thorn, fmt, ...)  do { printf("(%s) L%d at %s line %d: "fmt"\n", thorn, lvl, file, line, ## __VA_ARGS__); if(lvl == 0) exit(1); } while(0)
 #endif
 
 /********************************************************************
@@ -99,40 +99,42 @@
   uExpressionOpcode opcode;
 } operators[] = {
   /* Binary operators. */
-  {"=",  binary, 1,OP_EQUALS},
-  {"<",  binary, 1,OP_LESS_THAN},
-  {">",  binary, 1,OP_GREATER_THAN},
-  {"<=", binary, 1,OP_LEQUALS},
-  {">=", binary, 1,OP_GEQUALS},
+  {"||", binary, 1,OP_OR},
   {"&&", binary, 2,OP_AND},
-  {"||", binary, 2,OP_OR},
-  /* Sign change. */
-  {"_",  unary, 3,OP_NEGATE}, /* unary '-' */
-  {"@",  unary, 3,OP_PASS},   /* unary '+' */
+  {"=",  binary, 3,OP_EQUALS},
+  {"<",  binary, 4,OP_LESS_THAN},
+  {">",  binary, 4,OP_GREATER_THAN},
+  {"<=", binary, 4,OP_LEQUALS},
+  {">=", binary, 4,OP_GEQUALS},
   /* Binary operators. */
-  {"+",  binary, 3,OP_PLUS},
-  {"-",  binary, 3,OP_MINUS},
-  {"/",  binary, 4,OP_DIV},
-  {"*",  binary, 4,OP_TIMES},
-  {"^",  binary, 5,OP_POWER},
+  {"+",  binary, 5,OP_PLUS},
+  {"-",  binary, 5,OP_MINUS},
+  {"/",  binary, 6,OP_DIV},
+  {"%",  binary, 6,OP_REMAINDER},
+  {"*",  binary, 6,OP_TIMES},
+  {"**", binary, 7,OP_POWER},
+  /* Sign change */
+  {"_",     unary, 8,OP_NEGATE}, /* unary '-' */
+  {"@",     unary, 8,OP_PASS},   /* unary '+' */
   /* Unary Operators - these must have the highest precedence. */
-  {"!",     unary, 6, OP_NOT},
-  {"acos",  unary, 6, OP_ACOS},
-  {"asin",  unary, 6, OP_ASIN},
-  {"atan",  unary, 6, OP_ATAN},
-  {"ceil",  unary, 6, OP_CEIL},
-  {"cos" ,  unary, 6, OP_COS},  
-  {"cosh",  unary, 6, OP_COSH},
-  {"exp",   unary, 6, OP_EXP},
-  {"fabs",  unary, 6, OP_FABS},  
-  {"floor", unary, 6, OP_FLOOR},  
-  {"log",   unary, 6, OP_LOG},  
-  {"log10", unary, 6, OP_LOG10},
-  {"sin",   unary, 6, OP_SIN},  
-  {"sinh",  unary, 6, OP_SINH},  
-  {"sqrt",  unary, 6, OP_SQRT},  
-  {"tan",   unary, 6, OP_TAN},  
-  {"tanh",  unary, 6, OP_TANH},  
+  {"!",     unary, 8, OP_NOT},
+  {"acos",  unary, 9, OP_ACOS},
+  {"asin",  unary, 9, OP_ASIN},
+  {"atan",  unary, 9, OP_ATAN},
+  {"ceil",  unary, 9, OP_CEIL},
+  {"cos" ,  unary, 9, OP_COS},  
+  {"cosh",  unary, 9, OP_COSH},
+  {"exp",   unary, 9, OP_EXP},
+  {"fabs",  unary, 9, OP_FABS},  
+  {"floor", unary, 9, OP_FLOOR},  
+  {"log",   unary, 9, OP_LOG},  
+  {"log10", unary, 9, OP_LOG10},
+  {"sin",   unary, 9, OP_SIN},  
+  {"sinh",  unary, 9, OP_SINH},  
+  {"sqrt",  unary, 9, OP_SQRT},  
+  {"tan",   unary, 9, OP_TAN},  
+  {"tanh",  unary, 9, OP_TANH},  
+  {"trunc", unary, 9, OP_TRUNC},  
   {NULL, binary, -1,OP_NONE}
 };
 
@@ -469,7 +471,7 @@
 
   /* class '@' below is "everything else", "e" can be either scientific
    * notation or the literal character 'e' */
-  const char *classes[] = {"-+", "0123456789", ".", "eEdD", "*/^=&|<>!", "(", ")", " \t", "@"};
+  const char *classes[] = {"-+", "0123456789", ".", "eEdD", "*/%=&|<>!", "(", ")", " \t\n", "@"};
   /* if we are in state s and read a charcter of class c we transition to state
    * states[s][c] */
   /* if this is -1 then this is a transition that does not happen in well
@@ -477,7 +479,7 @@
   int states[][sizeof(classes)/sizeof(classes[0])] = {
       /* -   0   .   e   *   (   )  ' '  @ */
 /* 0*/ {10,  2,  3,  9, -1,  0, -1,  0,  9}, /* beginning of subexpression */
-/* 1*/ {-1,  2,  3,  9,  1,  0, -1,  1,  9}, /* after binary operator or '!' */
+/* 1*/ { 1,  2,  3,  9,  1,  0, -1, 11,  9}, /* after first character of binary operator or '!' */
 /* 2*/ { 1,  2,  3,  4,  1, -1,  5,  5, -1}, /* after first digit of integer */
 /* 3*/ {-1,  7, -1,  4,  1, -1,  5,  5, -1}, /* after decimal point */
 /* 4*/ { 8,  8, -1, -1, -1, -1, -1, -1, -1}, /* after 'e' of exponent */
@@ -487,13 +489,14 @@
 /* 8*/ {-1,  9, -1, -1,  1, -1,  5,  5, -1}, /* after first digit of exponent */
 /* 9*/ { 1,  9, -1,  9,  1,  0,  5,  6,  9}, /* after first letter of name */
 /*10*/ {-1,  2,  3,  9, -1,  0, -1, 10,  9}, /* after unary operator */
+/*11*/ {10,  2,  3, -1,  1,  0, -1, 11,  9}, /* after space after binary op */
   };
-  /* we terminate a token whenever we transition out of a state we check here
-   * is this transition could be a termination of a token */
+  /* we possibly terminate a token when we transition out of a state. we check here
+   * if this transition is a termination of a token */
   int end_of_token[][sizeof(classes)/sizeof(classes[0])] = {
       /* -   0   .   e   *   (   )  ' '  @ */
-/* 0*/ { 1,  1,  1,  1, -1,  1, -1,  0,  1}, /* beginning of subexpression */
-/* 1*/ {-1,  1,  1,  1,  0,  1, -1,  0,  1}, /* after binary operator or '!' */
+/* 0*/ { 1,  1,  1,  1, -1,  1, -1,  0,  1}, /* after beginning of subexpression */
+/* 1*/ { 1,  1,  1,  1,  0,  1, -1,  0,  1}, /* after binary operator or '!' */
 /* 2*/ { 1,  0,  0,  0,  1,  1,  1,  1, -1}, /* after first digit of integer */
 /* 3*/ {-1,  0, -1,  0,  1, -1,  1,  1, -1}, /* after decimal point */
 /* 4*/ { 0,  0, -1, -1, -1, -1, -1, -1, -1}, /* after 'e' of exponent */
@@ -502,7 +505,8 @@
 /* 7*/ {-1,  0, -1,  0,  1, -1,  1,  1, -1}, /* after first digit of fraction */
 /* 8*/ {-1,  0, -1, -1,  1, -1,  1,  1, -1}, /* after first digit of exponent */
 /* 9*/ { 1,  0, -1,  0,  1,  1,  1,  1,  0}, /* after first letter of name */
-/*10*/ {-1,  0,  0,  0, -1,  1, -1,  1,  0}, /* after unary operator */
+/*10*/ {-1,  0,  0,  1, -1,  1, -1,  1,  1}, /* after unary operator */
+/*11*/ { 1,  1,  3, -1,  1,  1, -1,  0,  1}, /* after space after binary op */
   };
   int tokenstate, state, class, error;
 
@@ -511,7 +515,7 @@
 
   tokenstart = expression;
   state = 0;
-  while(*tokenstart)
+  while(*tokenstart) /* as long as we have tokens left */
   {
 
     /* Remove leading whitespace */ 
@@ -520,7 +524,7 @@
     tokenend = NULL;
     tokenstate = -1;
 
-    /* classify first input character */
+    /* classify first input character of token*/
     for(class = 0; class < (int)(sizeof(classes)/sizeof(classes[0]))-1; class++)
     {
       if(strchr(classes[class], *tokenstart))
@@ -534,6 +538,7 @@
     printf("Initial state: %d, token '%c', class '%c'\n", state, *tokenstart, classes[class][0]);
 #endif
 
+    /* collect remainder of token */
     error = 0;
     for(position = tokenstart+1; *position && *tokenstart; position++)
     {
@@ -544,21 +549,20 @@
           break;
       }
 
-      /* see if transition closes the previous token */
-      if(end_of_token[state][class])
-      {
-        tokenend = position-1;
-        tokenstate = state;
-      }
-
 #ifdef TEST_EXPRESSION_PARSER
       printf("read '%c', current state: %d, class '%c', will transition to %d, token = '%.*s', will %sappend token\n",
           *position, state, classes[class][0], states[state][class], (int)(position-tokenstart), tokenstart, end_of_token[state][class] ? "" : "not ");
 #endif
 
-      /* transition to new state or state 0 upon errors */
-      if(states[state][class] >= 0)
+      /* see if transition closes the current token */
+      if(end_of_token[state][class])
       {
+        tokenend = position-1;
+        tokenstate = state;
+      }
+      else if(states[state][class] >= 0)
+      {
+        /* transition to new state or state 0 upon errors */
         state = states[state][class];
       }
       else
@@ -579,7 +583,7 @@
       }
     }
 
-    /* Have we reached the end of the string ? */
+    /* Have we reached the end of the string ? If so, make sure to add last token */
     if(!tokenend)
     {
       tokenend = position;
@@ -622,8 +626,8 @@
     }
     else
     {
-      fprintf(stderr, "Unable to allocate memory for new token !\n");
-      abort();
+      CCTK_VWarn (0, __LINE__, __FILE__, "Cactus",
+                  "Unable to allocate memory for new token !\n");
     }
   }
     
@@ -826,10 +830,16 @@
   {
     if(buffer->tokens[position].type == val)
     {
+#if TEST_EXPRESSION_PARSER
+      printf("%s ",buffer->vars[buffer->tokens[position].token.varnum]);
+#endif
       stackpointer++;
     }
     else
     {
+#if TEST_EXPRESSION_PARSER
+      printf("%s ",opname(buffer->tokens[position].token.opcode));
+#endif
       /* Evaluate operation, clear operands from stack and add the result to the stack. */
       switch(buffer->tokens[position].type)
       {
@@ -876,6 +886,10 @@
     retcode = 0;
   }
 
+#if TEST_EXPRESSION_PARSER
+      printf("\n");
+#endif
+
   return retcode;
 }
 
@@ -946,6 +960,9 @@
     case OP_DIV :                                      \
       (retval) = ((val1)/(val2));                      \
       break;                                           \
+    case OP_REMAINDER :                                \
+      (retval) = (fmod(val1,val2));                    \
+      break;                                           \
     case OP_TIMES :                                    \
       (retval) = ((val1)*(val2));                      \
       break;                                           \
@@ -1112,6 +1129,9 @@
     case OP_TANH :                                     \
       (retval) = tanh(val);                            \
       break;                                           \
+    case OP_TRUNC :                                    \
+      (retval) = trunc(val);                           \
+      break;                                           \
     default :                                          \
       fprintf(stderr, "Unknown operation %d", opcode); \
       (retval) = 0;                                    \
@@ -1604,9 +1624,19 @@
 
   for(i = 0; i < nvars; i++)
   {
-    if(strchr(vars[i],'.') || strchr(vars[i],'e'))
+    if (strcmp(vars[i], "pi") == 0)
     {
       vals[i].type = rval;
+      vals[i].value.rval = M_PI;
+    }
+    else if (strcmp(vars[i], "one") == 0)
+    {
+      vals[i].type = ival;
+      vals[i].value.ival = 1;
+    }
+    else if(strchr(vars[i],'.') || strchr(vars[i],'e') || strcasecmp(vars[i], "nan") == 0 || strcasecmp(vars[i], "inf") == 0)
+    {
+      vals[i].type = rval;
       vals[i].value.rval = strtod(vars[i], &endp);
       if(*endp)
         printf("Unrecognisable number format '%s' read as %g, unread '%s'\n", vars[i], vals[i].value.rval, endp);
@@ -1649,7 +1679,7 @@
     }
     else
     {
-      printf("Value is %g\n", value.value.rval); 
+      printf("Value is %#g\n", value.value.rval); 
     }
     
     Util_ExpressionFree(buffer);
