Index: sbin/RunTestUtils.pl
===================================================================
--- sbin/RunTestUtils.pl	(revision 4872)
+++ sbin/RunTestUtils.pl	(working copy)
@@ -889,22 +889,22 @@
   }
   push (@summary, "$message\n") if ($missingtests > 0);
 
-  # Different number of processors required
-  $message = "  Tests missed for different number of processors required:\n";
-  $missingtests = 0;
-  foreach $thorn (sort split(' ',$testdata->{'THORNS'}))
-  {
-    foreach $parfile (sort split(' ',$testdata->{"$thorn TESTS"}))
-    {
-      my $nprocs = $testdata->{"$thorn $parfile NPROCS"};
-      next unless ($nprocs);
-      $message .= "\n    ".$parfile." in ". $thorn."\n";
-      $message .= "      (". $testdata->{"$thorn $parfile DESC"}.")\n";
-      $message .= "      Requires $nprocs processors\n";
-      $missingtests++;
-    }
-  }
-  push (@summary, "$message\n") if ($missingtests > 0);
+  ## Different number of processors required
+  #$message = "  Tests missed for different number of processors required:\n";
+  #$missingtests = 0;
+  #foreach $thorn (sort split(' ',$testdata->{'THORNS'}))
+  #{
+  #  foreach $parfile (sort split(' ',$testdata->{"$thorn TESTS"}))
+  #  {
+  #    my $nprocs = $testdata->{"$thorn $parfile NPROCS"};
+  #    next unless ($nprocs);
+  #    $message .= "\n    ".$parfile." in ". $thorn."\n";
+  #    $message .= "      (". $testdata->{"$thorn $parfile DESC"}.")\n";
+  #    $message .= "      Requires $nprocs processors\n";
+  #    $missingtests++;
+  #  }
+  #}
+  #push (@summary, "$message\n") if ($missingtests > 0);
 
   # Different numbers of test files
 
@@ -1238,9 +1238,13 @@
   chdir ($testdata->{"$thorn $test TESTRUNDIR"}) ;
 
   # substitute the ($nprocs, $exe, $parfile) templates in the command
+  my $nprocs_available = $config_data->{'NPROCS'};
+  my $nprocs = $testdata->{"$thorn $parfile NPROCS"};
+  $nprocs = $nprocs_available unless $nprocs;
+  my $exe = $config_data->{'EXE'};
   my $cmd = $config_data->{'COMMAND'};
-  $cmd =~ s/\$exe/$config_data->{'EXE'}/g;
-  $cmd =~ s/\$nprocs/$config_data->{'NPROCS'}/g;
+  $cmd =~ s/\$exe/$exe/g;
+  $cmd =~ s/\$nprocs/$nprocs/g;
   $cmd =~ s/\$parfile/$parfile/g;
 
   $retcode = &RunCactus($output,$test,$cmd);
@@ -1690,12 +1694,12 @@
         $testdata->{"$thorn $testbase MISSING"} .= $missing;
         $testdata->{'NUNRUNNABLE'}++;
       }
-      elsif ($nprocs_required != $nprocs_available)
-      {
-        $testdata->{"$thorn UNRUNNABLE"} .= "$testbase ";
-        $testdata->{"$thorn $testbase NPROCS"} = $nprocs_required;
-        $testdata->{'NUNRUNNABLE'}++;
-      }
+      #elsif ($nprocs_required != $nprocs_available)
+      #{
+      #  $testdata->{"$thorn UNRUNNABLE"} .= "$testbase ";
+      #  $testdata->{"$thorn $testbase NPROCS"} = $nprocs_required;
+      #  $testdata->{'NUNRUNNABLE'}++;
+      #}
       else
       {
         $testdata->{"$thorn RUNNABLE"} .= "$testbase ";
