From 4676aa07ac2f84181eaf143bfa9a2d6b59e6e8e8 Mon Sep 17 00:00:00 2001
From: Barry Wardell <barry.wardell@gmail.com>
Date: Sun, 19 Jun 2011 23:24:51 +0200
Subject: [PATCH 08/11] Enable vectorization.

---
 m/WeylScal4.m |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/m/WeylScal4.m b/m/WeylScal4.m
index d2ed4d1..816b061 100644
--- a/m/WeylScal4.m
+++ b/m/WeylScal4.m
@@ -361,4 +361,5 @@ CreateKrancThornTT[groups, ".", "WeylScal4",
   IntParameters -> intParameters,
   InheritedImplementations -> {"admbase", "methodoflines"},
   UseJacobian -> True,
-  UseLoopControl -> True];
+  UseLoopControl -> True,
+  UseVectors -> True];
-- 
1.7.7.4

