diff --git a/m/EinsteinExact.m b/m/EinsteinExact.m
index dbed0bf..0e74d25 100644
--- a/m/EinsteinExact.m
+++ b/m/EinsteinExact.m
@@ -207,7 +207,7 @@ idThorn[spacetime_, thorn_] :=
     Name -> thorn <> "_" <> when,
     Switch[when,
       "initial", Schedule -> {"in ADMBase_InitialData"},
-      "always",  Schedule -> {"at ANALYSIS"},
+      "always",  Schedule -> {"at CCTK_PRESTEP"},
       _, Throw["Unrecognised scheduling keyword"]],
 
     ConditionalOnKeyword ->
