From 62e9d0e04857f47fd2f0ee1d142c07b6a112aa5b Mon Sep 17 00:00:00 2001
From: Roland Haas <roland.haas@physics.gatech.edu>
Date: Thu, 13 Dec 2012 19:30:43 -0800
Subject: [PATCH 20/25] GRHydro: clear divB everywhere rather than just in
 interior

---
 src/GRHydro_Analysis.F90 |   11 +----------
 1 file changed, 1 insertion(+), 10 deletions(-)

diff --git a/src/GRHydro_Analysis.F90 b/src/GRHydro_Analysis.F90
index 3bed392..8417fb7 100644
--- a/src/GRHydro_Analysis.F90
+++ b/src/GRHydro_Analysis.F90
@@ -13,16 +13,7 @@ subroutine GRHydro_Analysis_Init(CCTK_ARGUMENTS)
   DECLARE_CCTK_PARAMETERS
   DECLARE_CCTK_FUNCTIONS
 
-  CCTK_INT :: i,j,k
-  !$OMP PARALLEL DO PRIVATE(i,j,k)
-  do k = GRHydro_stencil + 1 - transport_constraints, cctk_lsh(3) - GRHydro_stencil ! we need to compute Evec on all faces/edges where the fluxes are defined
-     do j = GRHydro_stencil + 1 - transport_constraints, cctk_lsh(2) - GRHydro_stencil
-        do i = GRHydro_stencil + 1 - transport_constraints, cctk_lsh(1) - GRHydro_stencil
-           divB(i,j,k) = 0.0d0
-        end do
-     end do
-  end do
-  !$OMP END PARALLEL DO
+  divB = 0.0d0
 
 end subroutine GRHydro_Analysis_Init
 
-- 
1.7.10.4

