Index: src/detect.sh
===================================================================
--- src/detect.sh	(revision 102)
+++ src/detect.sh	(working copy)
@@ -158,10 +158,13 @@
 # Check for additional libraries
 ################################################################################
 
+# Set library directory name for machine architecture:
+MACHINE=${MACHINE:=`gcc -dumpmachine`}
+
 # Set options
 # Fortran modules may be located in the lib directory
-HDF5_INC_DIRS="${HDF5_DIR}/include ${HDF5_DIR}/lib"
-HDF5_LIB_DIRS="${HDF5_DIR}/lib"
+HDF5_INC_DIRS="${HDF5_DIR}/include ${HDF5_DIR}/lib/${MACHINE} ${HDF5_DIR}/lib"
+HDF5_LIB_DIRS="${HDF5_DIR}/lib/${MACHINE} ${HDF5_DIR}/lib"
 HDF5_LIBS="${HDF5_CXX_LIBS} ${HDF5_FORTRAN_LIBS} ${HDF5_C_LIBS}"
 
 
