Index: WaveBinarySource/src/WaveBinary.c
===================================================================
--- WaveBinarySource/src/WaveBinary.c	(revision 62)
+++ WaveBinarySource/src/WaveBinary.c	(working copy)
@@ -89,9 +89,9 @@
 
   for (k=0;k<cctk_lsh[2];k++)
   {
-    for (j=0;j<=cctk_lsh[1];j++)
+    for (j=0;j<cctk_lsh[1];j++)
     {
-      for (i=0;i<=cctk_lsh[0];i++)
+      for (i=0;i<cctk_lsh[0];i++)
       {
         vindex = CCTK_GFINDEX3D(cctkGH,i,j,k);
 
