From 7a85714c3882d656f9c0186bb2d3254e75debde3 Mon Sep 17 00:00:00 2001
From: Christian Reisswig <reisswig@scriwalker.(none)>
Date: Fri, 25 May 2012 17:24:01 -0700
Subject: [PATCH 10/13] GRHydro: Make enahnced params steerable.

---
 param.ccl |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/param.ccl b/param.ccl
index 62b9f29..df6d9ae 100644
--- a/param.ccl
+++ b/param.ccl
@@ -198,12 +198,12 @@ boolean use_enhanced_ppm "Use the enhanced ppm reconstruction method proposed by
 {
 } no
 
-real enhanced_ppm_C2 "Parameter for enhancecd ppm limiter. Default from McCorquodale & Colella 2011"
+real enhanced_ppm_C2 "Parameter for enhancecd ppm limiter. Default from McCorquodale & Colella 2011" STEERABLE=ALWAYS
 {
    *:* :: "must be greater than 1. According to Colella&Sekora 2008, enhanced ppm is insensitive to C in [1.25,5]"
 } 1.25
 
-real enhanced_ppm_C3 "Parameter for enhancecd ppm limiter. Default from McCorquodale & Colella 2011"
+real enhanced_ppm_C3 "Parameter for enhancecd ppm limiter. Default from McCorquodale & Colella 2011" STEERABLE=ALWAYS
 {
    0:* :: "must be greater than 0."
 } 0.1
-- 
1.7.10

