Index: param.ccl
===================================================================
--- param.ccl	(revision 13)
+++ param.ccl	(working copy)
@@ -1,11 +1,11 @@
 # Parameter definitions for thorn TmunuBase
 
+RESTRICTED:
+
 BOOLEAN stress_energy_storage "Should the stress-energy tensor have storage?"
 {
 } no
 
-RESTRICTED:
-
 BOOLEAN stress_energy_at_RHS "Should the stress-energy tensor be calculated for the RHS evaluation?"
 {
 } no
