Index: etc/mdb/damiana.ini
===================================================================
--- etc/mdb/damiana.ini	(revision 983)
+++ etc/mdb/damiana.ini	(working copy)
@@ -10,7 +10,6 @@
 
 # Access to this machine
 hostname        = login-damiana.aei.mpg.de
-rsynccmd        = /home/eschnett/rsync-3.0.4/bin/rsync
 aliaspattern    = ^(((login-)?damiana)|(sl-\d\d))(\.aei\.mpg\.de|\.damiana\.admin)?$
 
 # Source tree management
