From d5280570a38822accd7d7e2d4509e7cdf09cc9b4 Mon Sep 17 00:00:00 2001
From: Ian Hinder <ian.hinder@aei.mpg.de>
Date: Wed, 17 Aug 2011 14:39:34 +0200
Subject: [PATCH 4/4] Make --no-verbose the default

---
 etc/options/common.ini |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/etc/options/common.ini b/etc/options/common.ini
index 6c2cd3b..b33f254 100644
--- a/etc/options/common.ini
+++ b/etc/options/common.ini
@@ -184,7 +184,7 @@ long = --verbose
 desc = Be verbose with output
 negatable = yes
 arg = no
-default = True
+default = False
 #argformat
 
 [long]
-- 
1.7.5.4

