Index: lib/pyini.py
===================================================================
--- lib/pyini.py	(revision 1815)
+++ lib/pyini.py	(working copy)
@@ -628,8 +628,8 @@
         # match [<word>]
         section_re = '\[(\S+)\]'
         
-        # match key = value up to a comment [#;]
-        key_re = '(\S+)\s*=\s*([^#;]*)[#;]?.*$'
+        # match key = value up to a comment [#]
+        key_re = '(\S+)\s*=\s*([^#]*)[#]?.*$'
         
         # match block statement key = <<EOT ([#;] comment)
         block_re = '(\S+)\s*=\s*<<(\w+).*$'
