Index: new_thorn.pl
===================================================================
--- new_thorn.pl	(revision 4697)
+++ new_thorn.pl	(working copy)
@@ -94,6 +94,12 @@
 
 close OUT;
 
+open(OUT, ">configuration.ccl") || die "Cannot create configuration.ccl";
+
+print OUT "# Configuration definitions for thorn $thorn_name\n";
+
+close OUT;
+
 open(OUT, ">README") || die "Cannot create README";
 
 print OUT "Cactus Code Thorn $thorn_name\n";
