diff -r 03821ea5555f Carpet/CarpetMask/schedule.ccl
--- a/./repos/carpet/Carpet/CarpetMask/schedule.ccl	Tue Jul 03 21:33:52 2012 -0400
+++ b/./repos/carpet/Carpet/CarpetMask/schedule.ccl	Sat Jul 07 15:14:36 2012 -0700
@@ -11,6 +11,8 @@
 {
 } "Set up the weight function"
 
+# these two need to be global loop-local since SphericalSurface_Set and
+# AHFinder are global
 SCHEDULE CarpetExcludedSetup IN SetupIMask
 {
   LANG: C

