Index: test/trigger.par
===================================================================
--- test/trigger.par	(revision 34)
+++ test/trigger.par	(working copy)
@@ -83,8 +83,8 @@
 ADMBase::initial_shift = "tov"
 
 tovsolver::TOV_Rho_Central[0] = 1.28e-3
-tovsolver::TOV_Gamma[0]       = 2.0
-tovsolver::TOV_K[0]           = 100.0
+tovsolver::TOV_Gamma          = 2.0
+tovsolver::TOV_K              = 100.0
 tovsolver::TOV_Combine_Method = "maximum"
 
 eos_omni::poly_gamma                    = 2.0
