Index: make/configure.in
===================================================================
--- make/configure.in	(revision 4893)
+++ make/configure.in	(working copy)
@@ -1260,6 +1260,12 @@
 AC_SUBST(LIBNAME_SUFFIX)
 : ${LIBNAME_SUFFIX='.a'}
 
+# How to link in all object files from an archive
+AC_SUBST(BEGIN_WHOLE_ARCHIVE)
+: ${BEGIN_WHOLE_ARCHIVE=''}
+AC_SUBST(END_WHOLE_ARCHIVE)
+: ${END_WHOLE_ARCHIVE=''}
+
 # How to link the cactus libraries
 AC_SUBST(CACTUSLIBLINKLINE)
 : ${CACTUSLIBLINKLINE='-L$(CCTK_LIBDIR) $(addprefix -l, $(ALLCACTUSLIBS))'}
Index: make/make.config.defn.in
===================================================================
--- make/make.config.defn.in	(revision 4893)
+++ make/make.config.defn.in	(working copy)
@@ -141,6 +141,9 @@
 export LIBNAME_PREFIX = @LIBNAME_PREFIX@# Make sure there is no space at end.
 export LIBNAME_SUFFIX = @LIBNAME_SUFFIX@
 
+export BEGIN_WHOLE_ARCHIVE = @BEGIN_WHOLE_ARCHIVE@
+export END_WHOLE_ARCHIVE   = @END_WHOLE_ARCHIVE@
+
 export CACTUSLIBLINKLINE = @CACTUSLIBLINKLINE@
 
 # Dependency file generation
Index: make/make.configuration
===================================================================
--- make/make.configuration	(revision 4893)
+++ make/make.configuration	(working copy)
@@ -126,18 +126,30 @@
 
 
 # Build the list of all libraries
+ifeq ($(strip $(BEGIN_WHOLE_ARCHIVE)),)
+# Repeat Cactus only if --whole-archive is not used
 ALLCACTUSLIBS_BASE = $(notdir $(CACTUSLIBS) $(THORN_LINKLIST) $(CACTUSLIBS))
+else
+ALLCACTUSLIBS_BASE = $(notdir $(THORNS) $(CACTUSLIBS))
+endif
 
 ALLCACTUSLIBS = $(ALLCACTUSLIBS_BASE:%=$(CCTK_LIBNAME_PREFIX)%)
 
+$(CCTK_LIBDIR)/LINKLIST: $(patsubst %,$(CCTK_LIBDIR)/$(LIBNAME_PREFIX)$(CCTK_LIBNAME_PREFIX)%$(LIBNAME_SUFFIX),$(notdir $(THORNS) $(CACTUSLIBS)))
+	cd $(CCTK_LIBDIR); \
+	for thorn in $(notdir $(THORNS) $(CACTUSLIBS)); do \
+		xargs -n 1 -I :OBJFILE: echo $$thorn/:OBJFILE: <$(LIBNAME_PREFIX)$(CCTK_LIBNAME_PREFIX)$$thorn$(LIBNAME_SUFFIX).objectlist; \
+	done >$@.tmp
+	mv $@.tmp $@
+
 # Build the executable
-$(EXEDIR)$(DIRSEP)$(EXE): $(CONFIG)/make.thornlist $(CONFIG)/cctki_version.h $(patsubst %,$(CCTK_LIBDIR)/$(LIBNAME_PREFIX)$(CCTK_LIBNAME_PREFIX)%$(LIBNAME_SUFFIX),$(notdir $(THORNS) $(CACTUSLIBS)))
+$(EXEDIR)$(DIRSEP)$(EXE): $(CONFIG)/make.thornlist $(CONFIG)/cctki_version.h $(patsubst %,$(CCTK_LIBDIR)/$(LIBNAME_PREFIX)$(CCTK_LIBNAME_PREFIX)%$(LIBNAME_SUFFIX),$(notdir $(THORNS) $(CACTUSLIBS))) $(CCTK_LIBDIR)/LINKLIST
 	@echo Creating $(EXE) in $(EXEDIR) from $(THORNS)
 	datetime=`/bin/date +'%Y-%m-%dT%H:%M:%S%z' 2>&1` ; \
 	if [ $$? != 0 ]; then datetime=`/bin/date` ; fi ; \
 	$(CC) $(CFLAGS) -DCCTK_COMPILE_DATETIME="$$datetime" $(CCOMPILEONLY)$(OPTIONSEP)"$(call TRANSFORM_DIRS,$(TOP)/datestamp.o)" -I"$(call TRANSFORM_DIRS,$(CCTK_HOME)/src/include)" -I"$(call TRANSFORM_DIRS,$(CONFIG))" "$(call TRANSFORM_DIRS,$(DATESTAMP))"
 	if [ ! -d $(EXEDIR) ]; then $(MKDIR) $(MKDIRFLAGS) $(EXEDIR) ; fi
-	$(LD) $(CREATEEXE)$(OPTIONSEP)"$(call TRANSFORM_DIRS,$@)" $(CPPFLAGS) $(CXXFLAGS) $(LDFLAGS) $(EXTRAFLAGS) "$(call TRANSFORM_DIRS,$(TOP)/datestamp.o)" $(CACTUSLIBLINKLINE) $(GENERAL_LIBRARIES)
+	$(LD) $(CREATEEXE)$(OPTIONSEP)"$(call TRANSFORM_DIRS,$@)" $(CPPFLAGS) $(CXXFLAGS) $(LDFLAGS) $(EXTRAFLAGS) "$(call TRANSFORM_DIRS,$(TOP)/datestamp.o)" $(BEGIN_WHOLE_ARCHIVE) $(CACTUSLIBLINKLINE) $(END_WHOLE_ARCHIVE) $(GENERAL_LIBRARIES)
 	$(NOTIFY_DIVIDER)
 
 # Version file
Index: make/make.thornlib
===================================================================
--- make/make.thornlib	(revision 4893)
+++ make/make.thornlib	(working copy)
@@ -89,7 +89,6 @@
 ##	in batches not larger than $(OBJS-words-max) files at a time.
 	$(MAKE) -f $(MAKE_DIR)/make.thornlib $(NAME).objectlist
 	xargs -n $(OBJS-words-max) $(AR) $(ARFLAGS) $@ < $(NAME).objectlist
-	$(RM) $(NAME).objectlist
 ##	Alternatively, we could create a single object file from the object
 ##	files and put it into an archive.
 #	ld -r -o $@.o $(OBJS)
