From 841481bb0f08f98324d8e1e76755ec19d3460b68 Mon Sep 17 00:00:00 2001
From: Roland Haas <rhaas@tapir.caltech.edu>
Date: Fri, 22 Mar 2013 20:42:29 -0700
Subject: [PATCH 11/12] GRHydro_InitData: fix some text only typos

---
 param.ccl |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/param.ccl b/param.ccl
index cf0138f..9ae8c56 100644
--- a/param.ccl
+++ b/param.ccl
@@ -18,7 +18,7 @@ EXTENDS KEYWORD initial_hydro ""
   "simple_wave" :: "Set initial data from Anile Miller Motta, Phys.Fluids. 26, 1450 (1983)"
   "monopole"    :: "Monopole at the center"
   "cylexp"      :: "Cylindrical Explosion"
-  "rotor"       :: "Magnetic Rotor test from DelZanna,Bucciantini, and Londrillo A&A 400, 397–413 (2003)"
+  "rotor"       :: "Magnetic Rotor test from DelZanna,Bucciantini, and Londrillo A&A 400, 397-413 (2003)"
   "advectedloop":: "Magnetic advected loop test"
   "alfvenwave"  :: "Circularly polarized Alfven wave"
   "hydro_bondi_solution" :: "Spherical single black hole Bondi solution"
@@ -275,7 +275,7 @@ CCTK_REAL cyl_r_inner "Inner Radius"
   (0:* :: "Any positive number"
 } 0.8
 
-CCTK_REAL cyl_r_outer "Inner Radius"
+CCTK_REAL cyl_r_outer "Outer Radius"
 {
   (0:* :: "Any positive number"
 } 1.0
-- 
1.7.10.4

