Index: configure.in
===================================================================
--- configure.in	(revision 4702)
+++ configure.in	(working copy)
@@ -1262,10 +1262,11 @@
 : ${SYS_INC_DIRS=' '}
 
 # How to link in the non-Cactus libraries:
-# Add -L and -l prefixes, but only for file names;
-# allow arbitrary options in between
+# Add -L, -R, and -l prefixes, but only for file names (not for options);
+# allow arbitrary options in between these;
+# sort library directories so that system directories (/opt/local/*) come last
 AC_SUBST(GENERAL_LIBRARIES)
-: ${GENERAL_LIBRARIES='$(patsubst $(LIBDIR_PREFIX)-%,-%,$(LIBDIRS:%=$(LIBDIR_PREFIX)%)) $(patsubst $(RUNDIR_PREFIX)-%,-%,$(LIBDIRS:%=$(RUNDIR_PREFIX)%)) $(patsubst $(LIBLINK_PREFIX)-%,-%,$(LIBS:%=$(LIBLINK_PREFIX)%))'}
+: ${GENERAL_LIBRARIES='$(patsubst $(LIBDIR_PREFIX)-%,-%,$(addprefix $(LIBDIR_PREFIX),$(filter-out /opt/local/%,$(LIBDIRS)) $(filter /opt/local/%,$(LIBDIRS)))) $(patsubst $(RUNDIR_PREFIX)-%,-%,$(addprefix $(RUNDIR_PREFIX),$(filter-out /opt/local/%,$(LIBDIRS)) $(filter /opt/local/%,$(LIBDIRS)))) $(patsubst $(LIBLINK_PREFIX)-%,-%,$(addprefix $(LIBLINK_PREFIX),$(LIBS)))'}
 
 
 # Do we need to make sure Perl makes a backup when editting in place ?
