From 1bd070c8563811367f097727117d6d5dcffa253e Mon Sep 17 00:00:00 2001
From: Barry Wardell <barry.wardell@gmail.com>
Date: Fri, 2 Mar 2012 09:31:49 +0000
Subject: [PATCH 2/2] Add missing casts.

---
 src/vectors-4-SSE.h |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/src/vectors-4-SSE.h b/src/vectors-4-SSE.h
index a84c9fd..46fc4da 100644
--- a/src/vectors-4-SSE.h
+++ b/src/vectors-4-SSE.h
@@ -373,7 +373,8 @@ static const union {
     CCTK_REAL4_VEC const x=x__;                                 \
     CCTK_REAL4_VEC const y=y__;                                 \
     CCTK_REAL4_VEC const z=z__;                                 \
-    CCTK_REAL4_VEC const mask = _mm_srai_epi32(x, 31);          \
+    CCTK_REAL4_VEC const mask =                                 \
+      (__m128)_mm_srai_epi32((__m128i)x, 31);                   \
     /* (z & ~mask) | (y & mask) */                              \
     _mm_or_ps(_mm_andnot_ps(mask, z), _mm_and_ps(mask, y));     \
   })
-- 
1.7.7.4

