--- HyperArch.py.orig	2014-01-25 21:30:14.904259073 -0800
+++ HyperArch.py	2014-01-25 21:29:19.376256523 -0800
@@ -902,7 +902,7 @@
         self.DropArchLock()
 
     def get_filename(self, article):
-        return '%06i.html' % (article.sequence,)
+        return '%s.html' % (article.msgid[0:article.msgid.find("@")],)
 
     def get_archives(self, article):
         """Return a list of indexes where the article should be filed.
