From 0a0a3b3adc2a546b1da6216075a89d7adef13e29 Mon Sep 17 00:00:00 2001
From: Roland Haas <roland.haas@physics.gatech.edu>
Date: Sat, 2 Jun 2012 09:42:23 -0700
Subject: [PATCH 11/11] Carpet: make parameters steerable

---
 Carpet/Carpet/param.ccl |   12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/Carpet/Carpet/param.ccl b/Carpet/Carpet/param.ccl
index e8d24f7..d901e0c 100644
--- a/Carpet/Carpet/param.ccl
+++ b/Carpet/Carpet/param.ccl
@@ -369,7 +369,7 @@ BOOLEAN sync_barriers "Insert barriers before syncs, so that the sync timer is m
 {
 } "no"
 
-BOOLEAN output_internal_data "Periodically print internal data to the screen for debugging purposes"
+BOOLEAN output_internal_data "Periodically print internal data to the screen for debugging purposes" STEERABLE=always
 {
 } "no"
 
@@ -402,7 +402,7 @@ BOOLEAN timers_verbose "Output (debug) messages when a timer is started or stopp
 {
 } "no"
 
-STRING timer_xml_clock "Which clock to use in the XML timer output file"
+STRING timer_xml_clock "Which clock to use in the XML timer output file" STEERABLE=always
 {
   ".*" :: "must be a legal clock name"
 } "gettimeofday"
@@ -411,13 +411,13 @@ BOOLEAN output_initialise_timer_tree "Output timing information in tree form to
 {
 } "no"
 
-INT output_timer_tree_every "Output timing information in tree form to standard output for Evolve every so many iterations"
+INT output_timer_tree_every "Output timing information in tree form to standard output for Evolve every so many iterations" STEERABLE=always
 {
   0   :: "don't report"
   1:* :: "report every so many iterations"
 } 0
 
-BOOLEAN output_xml_timer_tree "Output timing information in tree form as XML"
+BOOLEAN output_xml_timer_tree "Output timing information in tree form as XML" STEERABLE=always
 {
 } "no"
 
@@ -426,7 +426,7 @@ CCTK_REAL timer_tree_threshold_percentage "The percentage of the root timer belo
   0:* :: ""
 } 1.0
 
-INT timer_tree_output_precision "Number of decimal places to use in standard output for timer tree"
+INT timer_tree_output_precision "Number of decimal places to use in standard output for timer tree" STEERABLE=always
 {
   1:* :: "number of decimal places"
 } 1
@@ -590,7 +590,7 @@ BOOLEAN adaptive_stepsize "Allow adaptive timestep sizes"
 {
 } "no"
 
-BOOLEAN use_unusedpoints_mask "Turn on of off (default) storage and usage of 'unusedpoints_mask'"
+BOOLEAN use_unusedpoints_mask "Turn on of off (default) storage and usage of 'unusedpoints_mask'" STEERABLE=recover
 {
 } "no"
 
-- 
1.7.10.4

