Index: arrangements/CactusNumerical/MoL/schedule.ccl
===================================================================
--- arrangements/CactusNumerical/MoL/schedule.ccl	(revision 148)
+++ arrangements/CactusNumerical/MoL/schedule.ccl	(working copy)
@@ -675,11 +675,11 @@
 ###  PostStepModify is only scheduled during evolution.               ###
 #########################################################################
 
-schedule GROUP MoL_PostStepModify IN MoL_Step AFTER MoL_PostStep
+schedule GROUP MoL_PostStepModify IN MoL_Step AFTER MoL_Add BEFORE MoL_PostStep
 {
 } "The group for physics thorns to schedule enforcing constraints"
 
-schedule GROUP MoL_PostStepModify At PostInitial AFTER MoL_PostStep
+schedule GROUP MoL_PostStepModify At PostInitial BEFORE MoL_PostStep
 {
 } "The group for physics thorns to schedule enforcing constraints"
 
