diff --git a/Outflow/src/outflow.c b/Outflow/src/outflow.c
index ba05233..1a21e40 100644
--- a/Outflow/src/outflow.c
+++ b/Outflow/src/outflow.c
@@ -185,7 +185,7 @@ static int Outflow_write_2d_output(CCTK_ARGUMENTS, const char *varname, CCTK_INT
     fprintf(file,"# gnuplot column index:\n");
     fprintf(file,"# 1:it 2:t 3:x 4:y 5:z 6:%s 7:w_lorentz 8:eninf 9:surface_element", varname);
     for (int i = 0 ; i < num_extras ; i++) {
-      fprintf(file, " %d:%s", 9+i, CCTK_VarName(extras_ind[i]));
+      fprintf(file, " %d:%s", 10+i, CCTK_VarName(extras_ind[i]));
     }
     fprintf(file, "\n");
   }
