Index: configure.in
===================================================================
--- configure.in	(revision 4699)
+++ configure.in	(working copy)
@@ -398,8 +398,8 @@
 AC_SUBST(F90_NO_OPTIMISE_FLAGS)
 : ${F90_NO_OPTIMISE_FLAGS='-O0'}
 
-AC_SUBST(CUCC_OPTIMISE_FLAGS)
-: ${CUCC_OPTIMISE_FLAGS='-O2'}
+AC_SUBST(CUCC_NO_OPTIMISE_FLAGS)
+: ${CUCC_NO_OPTIMISE_FLAGS='-O0'}
 
 # Set the profiling flags if they've not been set by now
 # (using '-pg' as a reasonable default)
