Index: dist/hwloc-1.10.1.tar.gz
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/x-gzip

Property changes on: dist/hwloc-1.10.1.tar.gz
___________________________________________________________________
Deleted: svn:mime-type
## -1 +0,0 ##
-application/x-gzip
\ No newline at end of property
Index: dist/hwloc-1.11.12.tar.gz
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/octet-stream

Property changes on: dist/hwloc-1.11.12.tar.gz
___________________________________________________________________
Added: svn:mime-type
## -0,0 +1 ##
+application/octet-stream
\ No newline at end of property
Index: src/build.sh
===================================================================
--- src/build.sh	(revision 93)
+++ src/build.sh	(working copy)
@@ -14,7 +14,7 @@
 
 # Set locations
 THORN=hwloc
-NAME=hwloc-1.10.1
+NAME=hwloc-1.11.12
 SRCDIR="$(dirname $0)"
 BUILD_DIR=${SCRATCH_BUILD}/build/${THORN}
 if [ -z "${HWLOC_INSTALL_DIR}" ]; then
Index: src/detect.sh
===================================================================
--- src/detect.sh	(revision 93)
+++ src/detect.sh	(working copy)
@@ -140,7 +140,7 @@
     fi
 
     # Set locations
-    NAME=hwloc-1.10.1
+    NAME=hwloc-1.11.12
     SRCDIR="$(dirname $0)"
     BUILD_DIR=${SCRATCH_BUILD}/build/${THORN}
     if [ -z "${HWLOC_INSTALL_DIR}" ]; then
