Index: test/KS-tilted.par
===================================================================
--- test/KS-tilted.par	(revision 271)
+++ test/KS-tilted.par	(working copy)
@@ -10,19 +10,23 @@
 
 ActiveThorns = "AEILocalInterp LocalInterp LocalReduce"
 
-ActiveThorns = "SphericalSurface"
-
 ActiveThorns = "PUGH PUGHInterp PUGHReduce PUGHSlab"
 driver::global_nsize = 49
 driver::ghost_size   =  2
 
 ActiveThorns = "NaNChecker"
 
-ActiveThorns = "Boundary CartGrid3D CoordBase SymBase SymBase"
+ActiveThorns = "Boundary CartGrid3D CoordBase SymBase SymBase SphericalSurface CoordBase"
 CartGrid3D::type         = "byspacing"
 CartGrid3D::domain       = "full"
 CartGrid3D::dxyz         = 0.125
 CartGrid3D::avoid_origin = no
+CoordBase::boundary_size_x_upper = 2
+CoordBase::boundary_size_y_upper = 2
+CoordBase::boundary_size_z_upper = 2
+CoordBase::boundary_size_x_lower = 2
+CoordBase::boundary_size_y_lower = 2
+CoordBase::boundary_size_z_lower = 2
 
 ActiveThorns = "ADMBase ADMCoupling ADMMacros CoordGauge SpaceMask StaticConformal TmunuBase"
 SpaceMask::use_mask = yes
@@ -34,9 +38,9 @@
 Exact::exact_model              = "Kerr/Kerr-Schild"
 Exact::rotation_euler_theta     = 0.4 # in radians, approx. 23 degrees
 
-ActiveThorns = "LegoExcision"
-LegoExcision::fixed_excision = "sphere"
-LegoExcision::fixed_size     = 1.01
+ActiveThorns = "NoExcision"
+NoExcision::num_regions      = 1
+NoExcision::radius[0]        = 0.8
 
 ActiveThorns = "Time"
 Time::dtfac = 0.25
@@ -44,16 +48,22 @@
 ActiveThorns = "MoL"
 MoL::ODE_Method = "RK3"
 
-ActiveThorns = "BSSN_MoL"
-BSSN_MoL::excise       = yes
-BSSN_MoL::excisionType = "lego"
+ActiveThorns = "ML_BSSN ML_BSSN_Helper GenericFD LoopControl TmunuBase NewRad"
+ADMBase::evolution_method = ML_BSSN
+ML_BSSN::fdOrder                     = 2
+ML_BSSN::my_initial_boundary_condition = "extrapolate-gammas"
+ML_BSSN::my_rhs_boundary_condition     = "NewRad"
+# These are necessary
+ML_BSSN::ML_log_confac_bound = "none"
+ML_BSSN::ML_metric_bound     = "none"
+ML_BSSN::ML_Gamma_bound      = "none"
+ML_BSSN::ML_trace_curv_bound = "none"
+ML_BSSN::ML_curv_bound       = "none"
+ML_BSSN::ML_lapse_bound      = "none"
+ML_BSSN::ML_dtlapse_bound    = "none"
+ML_BSSN::ML_shift_bound      = "none"
+ML_BSSN::ML_dtshift_bound    = "none"
 
-ADMBase::evolution_method = ADM_BSSN
-ADM_BSSN::stencil_size               = 2
-ADM_BSSN::advection                  = "upwind2"
-ADM_BSSN::bound                      = "newrad"
-ADM_BSSN::BSSN_MaxNumConstrainedVars = 16
-
 ActiveThorns = "ADMConstraints"
 
 ActiveThorns = "AHFinderDirect"
Index: test/de_Sitter.par
===================================================================
--- test/de_Sitter.par	(revision 271)
+++ test/de_Sitter.par	(working copy)
@@ -3,14 +3,14 @@
 # $Header$
 
 # Required thorns
-ActiveThorns = "CartGrid3D CoordBase SymBase Boundary Time   PUGH PUGHReduce PUGHSlab   IOAscii IOUtil IOBasic   ADMBase ADMMacros ADMConstraints ADM_BSSN ADMCoupling StaticConformal CoordGauge SpaceMask   Exact"
+ActiveThorns = "CartGrid3D CoordBase SymBase Boundary Time   PUGH PUGHReduce PUGHSlab   IOAscii IOUtil IOBasic LocalReduce   ADMBase ADMMacros ADMConstraints ADMCoupling StaticConformal CoordGauge SpaceMask  ReflectionSymmetry Exact"
 
 # Grid
 driver::global_nx = 8
 driver::global_ny = 8
 driver::global_nz = 8
 grid::type    = "byrange"
-grid::domain  = "octant"
+grid::domain  = "full"
 
 grid::xmin = 0.0
 grid::xmax = 1.0
@@ -19,6 +19,10 @@
 grid::zmin = 0.0
 grid::zmax = 6.28
 
+ReflectionSymmetry::reflection_x = yes
+ReflectionSymmetry::reflection_y = yes
+ReflectionSymmetry::reflection_z = yes
+
 cactus::cctk_initial_time = 1
 cactus::cctk_itlast       = 3
 time::dtfac = 0.25
@@ -34,9 +38,7 @@
 ADMBase::shift_evolution_method = "exact"
 
 # Evolution
-ADMBase::evolution_method = "adm_bssn"
-ADM_BSSN::bound  = "flat"
-ADM_BSSN::method = "icn"
+ADMBase::evolution_method = "exact"
 
 # Output
 IO::out_dir = "de_Sitter"
@@ -47,9 +49,9 @@
 IOBasic::outInfo_vars  = "admconstraints::ham"
 
 IOBasic::outScalar_every =  1 
-IOBasic::outScalar_vars  = "admbase::gxx admbase::gyy admbase::gzz admconstraints::ham admbase::lapse adm_bssn::ADM_BS_phi"
+IOBasic::outScalar_vars  = "admbase::gxx admbase::gyy admbase::gzz admconstraints::ham admbase::lapse"
 
 IOASCII::out1D_every =  1
-IOASCII::out1D_vars  = "admbase::gxx admbase::gyy admbase::gzz admconstraints::ham admbase::lapse adm_bssn::ADM_BS_phi"
+IOASCII::out1D_vars  = "admbase::gxx admbase::gyy admbase::gzz admconstraints::ham admbase::lapse"
 
 
