From cf49f416b77e5bd82284f79032e7c46fbdcdf3cd Mon Sep 17 00:00:00 2001
From: Barry Wardell <barry.wardell@gmail.com>
Date: Wed, 12 Feb 2014 11:27:21 -0500
Subject: [PATCH 2/2] Suppress an error from grep if libhwloc.la does not exist

---
 configure.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.sh b/configure.sh
index f877ac6..2e2e524 100755
--- a/configure.sh
+++ b/configure.sh
@@ -199,7 +199,7 @@ else
 fi
 
 # Add libnuma manually, if necessary
-if grep -q '[-]lnuma' ${HWLOC_LIB_DIR}/lib/libhwloc.la; then
+if grep -q '[-]lnuma' ${HWLOC_LIB_DIR}/lib/libhwloc.la 2>/dev/null; then
     if ! echo '' ${HWLOC_LIBS} '' | grep -q ' numa '; then
         HWLOC_LIBS="${HWLOC_LIBS} numa"
     fi
-- 
1.8.4.1

