From c16135b2dcea43a2ccfe6d484a803ee8ffe984ba Mon Sep 17 00:00:00 2001
From: Ian Hinder <ian.hinder@aei.mpg.de>
Date: Fri, 28 Oct 2011 15:27:07 +0200
Subject: [PATCH] Prompt for Mercurial authentication information initially if
 necessary

---
 GetComponents |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/GetComponents b/GetComponents
index 2c95a9c..7f0c302 100755
--- a/GetComponents
+++ b/GetComponents
@@ -842,7 +842,8 @@ sub process_users {
             and (  $component->{TYPE} eq 'cvs'
                 or $component->{TYPE} eq 'svn'
                 or $component->{TYPE} eq 'darcs'
-                or $component->{TYPE} eq 'git' )
+                or $component->{TYPE} eq 'git'
+                or $component->{TYPE} eq 'hg' )
           )
         {
 
-- 
1.7.5.4

