From 1e1a33c46a948066b6fd98fd8a792ddb464ce273 Mon Sep 17 00:00:00 2001
From: Roland Haas <rhaas@tapir.caltech.edu>
Date: Tue, 9 Apr 2013 20:12:23 -0700
Subject: [PATCH 06/12] GRHydro: correct parameter description text

---
 param.ccl |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/param.ccl b/param.ccl
index b56d061..8f0b60d 100644
--- a/param.ccl
+++ b/param.ccl
@@ -83,7 +83,7 @@ USES REAL poly_k
 
 restricted:
 
-KEYWORD use_evolution_mask "Set this to 'true' to skip validity tests in regions where CarpetEvolutionMask::evolution_mask vanishes." STEERABLE=ALWAYS
+KEYWORD use_evolution_mask "Set this to 'always' to skip validity tests in regions where CarpetEvolutionMask::evolution_mask vanishes." STEERABLE=ALWAYS
 {
   "always" :: "use the mask"
   "auto"   :: "check if CarpetEvolutionMask is active, then use the mask"
-- 
1.7.10.4

