Index: src/make.configuration.defn
===================================================================
--- src/make.configuration.defn	(revision 65)
+++ src/make.configuration.defn	(working copy)
@@ -1,4 +1,4 @@
 # make.configuration.defn file for thorn hwloc
 
 # Define the hwloc utilities
-ALL_UTILS += hwloc-assembler hwloc-assembler-remote hwloc-bind hwloc-calc hwloc-distances hwloc-distrib hwloc-info hwloc-ls hwloc-ps lstopo lstopo-no-graphics
+ALL_UTILS += hwloc-assembler hwloc-assembler-remote hwloc-bind hwloc-calc hwloc-distances hwloc-distrib hwloc-info hwloc-ls hwloc-ps lstopo
