diff --git a/Boundary/src/ScalarBoundary.c b/Boundary/src/ScalarBoundary.c
index f7b90cb..53fabaa 100644
--- a/Boundary/src/ScalarBoundary.c
+++ b/Boundary/src/ScalarBoundary.c
@@ -147,7 +147,7 @@ CCTK_INT BndScalar(const cGH *GH, CCTK_INT num_vars, CCTK_INT *vars,
     /* fill it with values, either from table or the boundary_width
        parameter */
     if (widths[i] < 0) {
-      err = Util_TableGetIntArray(tables[i], gdim, width_alldirs,
+      err = Util_TableGetIntArray(tables[i], 2 * gdim, width_alldirs,
                                   "BOUNDARY_WIDTH");
       if (err < 0) {
         CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
