Index: src/piraha/Call.cc
===================================================================
--- src/piraha/Call.cc	(revision 4997)
+++ src/piraha/Call.cc	(working copy)
@@ -208,7 +208,6 @@
 	return o;
 }
 
-std::map<std::string,std::map<std::string,smart_ptr<Value> > > values;
 typedef std::map<std::string,std::map<std::string,smart_ptr<Value> > >::iterator th_iter;
 typedef std::map<std::string,smart_ptr<Value> >::iterator nm_iter;
 
@@ -216,21 +215,47 @@
 
 // If the value was already defined in this parameter file, look
 // it up in the map. Otherwise, get it from Cactus.
-smart_ptr<Value> find_val(std::string thorn,std::string name) {
-	smart_ptr<Value> ret;
-	th_iter th = values.find(thorn);
-	if(th != values.end()) {
-		nm_iter nm = th->second.find(name);
-		if(nm != th->second.end()) {
-			ret = nm->second;
-			return ret;
-		}
-	}
-	const cParamData *data = CCTK_ParameterData(name.c_str(),thorn.c_str());
-	if(data != NULL) {
-		std::string expr = data->defval;
-		ret = eval_expr(expr);
-	}
+smart_ptr<Value> find_val(smart_ptr<Group> gr,std::string thorn,std::string name) {
+	smart_ptr<Value> ret = new Value(gr);
+    int type;
+    const void *result;
+    result = CCTK_ParameterGet(name.c_str(),thorn.c_str(),&type);
+    if(result == NULL) {
+		std::ostringstream msg;
+		msg << "Undefined or inaccessible variable: " << thorn << "::" << name << std::endl;
+		std::string par = get_parfile();
+		CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
+    }
+    
+    switch(type) {
+        case PARAMETER_REAL:
+            ret->type = PIR_REAL;
+            ret->ddata = *(const CCTK_REAL*)result;
+            break;
+        case PARAMETER_INT:
+            ret->type = PIR_INT;
+            ret->idata = *(const CCTK_INT*)result;
+            break;
+        case PARAMETER_BOOLEAN:
+            ret->type = PIR_BOOL;
+            ret->idata = *(const CCTK_INT*)result;
+            break;
+        case PARAMETER_STRING:
+        case PARAMETER_SENTENCE:
+        case PARAMETER_KEYWORD:
+            {
+                ret->type = PIR_STRING;
+                const char *s = *(const char **)result;
+                ret->sdata += s;
+            }
+            break;
+        default:
+		    std::ostringstream msg;
+		    msg << "Unexpected type result from ParameterGet=" << type << std::endl;
+		    std::string par = get_parfile();
+		    CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
+    }
+    std::cout << "Piraha: GET: " << thorn << "::" << name << "=" << ret << std::endl;
 	return ret;
 }
 
@@ -425,11 +450,11 @@
 				o << name << "[" << index->idata << "]";
 				o << std::flush;
 				std::string keyi = o.str();
-				ret = find_val(thorn,keyi);
+				ret = find_val(gr,thorn,keyi);
 				return ret;
 			}
 		} else {
-			ret = find_val(thorn,name);
+			ret = find_val(gr,thorn,name);
 			return ret;
 		}
 		std::ostringstream msg;
@@ -573,9 +598,11 @@
 					ret->sdata = v1->sdata + v2->sdata;
 				} else {
 					std::ostringstream msg;
+                    ret->sdata = v1->sdata + addop + v2->sdata;
 					msg << "Unknown add operator: " << addop << std::endl;
+                    msg << "Interpreting as literal string with value '" << ret->sdata << "'" << std::endl;
 					std::string par = get_parfile();
-					CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
+					CCTK_Warn(1,gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
 				}
 			} else if(v1->intOrDouble() && v2->intOrDouble()) {
 				ret->type = PIR_REAL;
@@ -643,6 +670,14 @@
 					std::string par = get_parfile();
 					CCTK_Error(gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
 				}
+			} else if(v1->type == PIR_STRING && v2->type == PIR_STRING) {
+				std::ostringstream msg;
+                ret->type = PIR_STRING;
+                ret->sdata = v1->sdata + mulop + v2->sdata;
+				msg << "Unknown operation: " << v1->type << " " << mulop << v2->type << std::endl;
+                msg << "Interpreting as literal string with value '" << ret->sdata << "'" << std::endl;
+				std::string par = get_parfile();
+				CCTK_Warn(1,gr->line(),par.c_str(),current_thorn.c_str(),msg.str().c_str());
 			} else {
 				std::ostringstream msg;
 				msg << "Unknown operation: " << v1->type << " " << mulop << v2->type << std::endl;
@@ -722,7 +757,7 @@
 			"# Note that / occurs in some par files. It is my\n"
 			"# feeling that this should require quote marks.\n"
 
-			"name = [@a-zA-Z_][@/a-zA-Z0-9_-]*\n"
+			"name = [a-zA-Z][a-zA-Z0-9_]*\n"
 			"dname = [0-9][a-zA-Z_]{2,}\n"
 			"inquot = ({var}|\\\\.|[^\\\\\"])*\n"
 			"fname = \\.?/[-\\./0-9a-zA-Z_]+\n"
@@ -813,7 +848,6 @@
     				smart_ptr<Value> smv = meval(aexpr);
 					val = smv->copy();
     				assert(smv.valid());
-    				values[thorn][key] = smv;
     				smv->integerize();
     				if(data != NULL) {
         				if(data->type == PARAMETER_REAL)
@@ -834,7 +868,6 @@
     					keyi << key << '[' << i << ']';
     					smart_ptr<Value> smv = meval(aexpr);
     					val = smv->copy();
-        				values[thorn][keyi.str()] = smv;
         				if(data != NULL) {
             				if(data->type == PARAMETER_REAL)
             					smv->integerize();
