Index: bluewaters-cray.ini
===================================================================
--- bluewaters-cray.ini	(revision 2300)
+++ bluewaters-cray.ini	(working copy)
@@ -14,7 +14,7 @@
 # Access to this machine
 hostname        = h2ologin.ncsa.illinois.edu
 rsynccmd        = /u/sciteam/eschnett/rsync-3.0.9/bin/rsync
-envsetup        = source /etc/profile && module unload PrgEnv-cray PrgEnv-gnu PrgEnv-intel PrgEnv-pathscale PrgEnv-pgi && module load PrgEnv-cray && module swap cce cce/8.1.5 && module load fftw/3.3.0.1 && module load hdf5/1.8.8 && module load papi/5.0.1
+envsetup        = source /etc/profile && module unload PrgEnv-cray PrgEnv-gnu PrgEnv-intel PrgEnv-pathscale PrgEnv-pgi && module load PrgEnv-cray && module swap cce cce/8.1.5 && module load fftw/3.3.0.1 && module load hdf5/1.8.8 && module load papi/5.0.1 && module load atp
 aliaspattern    = $^
 
 # Source tree management
Index: bluewaters-pgi.ini
===================================================================
--- bluewaters-pgi.ini	(revision 2300)
+++ bluewaters-pgi.ini	(working copy)
@@ -14,7 +14,7 @@
 # Access to this machine
 hostname        = h2ologin.ncsa.illinois.edu
 rsynccmd        = /u/sciteam/eschnett/rsync-3.0.9/bin/rsync
-envsetup        = source /etc/profile && module unload PrgEnv-cray PrgEnv-gnu PrgEnv-intel PrgEnv-pathscale PrgEnv-pgi && module load PrgEnv-pgi && module swap pgi pgi/13.10.0 && module load acml/5.3.1 && module load cray-hdf5/1.8.11 && module load cray-petsc/3.4.3.0 && module load cudatoolkit/5.5.20-1.0402.7700.8.1 && module load fftw/3.3.0.4 && module load gsl/1.15.1 && module load papi/5.2.0 && module load pmi
+envsetup        = source /etc/profile && module unload PrgEnv-cray PrgEnv-gnu PrgEnv-intel PrgEnv-pathscale PrgEnv-pgi && module load PrgEnv-pgi && module swap pgi pgi/13.10.0 && module load acml/5.3.1 && module load cray-hdf5/1.8.11 && module load cray-petsc/3.4.3.0 && module load cudatoolkit/5.5.20-1.0402.7700.8.1 && module load fftw/3.3.0.4 && module load gsl/1.15.1 && module load papi/5.2.0 && module load pmi && module load atp
 aliaspattern    = $^
 
 # Source tree management
Index: bluewaters.ini
===================================================================
--- bluewaters.ini	(revision 2300)
+++ bluewaters.ini	(working copy)
@@ -15,7 +15,7 @@
 hostname        = h2ologin.ncsa.illinois.edu
 rsynccmd        = /u/sciteam/eschnett/rsync-3.0.9/bin/rsync
 # module load perftools/6.0.1
-envsetup        = source /etc/profile && module unload PrgEnv-cray PrgEnv-gnu PrgEnv-intel PrgEnv-pathscale PrgEnv-pgi && module load PrgEnv-gnu && module load cudatoolkit/5.0.35 && module load acml/5.3.1 && module load boost/1.53.0 && module load cray-hdf5/1.8.11 && module load cray-petsc/3.4.3.0 && module load cudatoolkit/5.5.20-1.0402.7700.8.1 && module load fftw/3.3.0.4 && module load gsl/1.15.1 && module load papi/5.2.0 && module load pmi
+envsetup        = source /etc/profile && module unload PrgEnv-cray PrgEnv-gnu PrgEnv-intel PrgEnv-pathscale PrgEnv-pgi && module load PrgEnv-gnu && module load cudatoolkit/5.0.35 && module load acml/5.3.1 && module load boost/1.53.0 && module load cray-hdf5/1.8.11 && module load cray-petsc/3.4.3.0 && module load cudatoolkit/5.5.20-1.0402.7700.8.1 && module load fftw/3.3.0.4 && module load gsl/1.15.1 && module load papi/5.2.0 && module load pmi && module load atp
 aliaspattern    = ^h2ologin[1-4](\.ncsa\.illinois\.edu)?$
 
 # Source tree management
