Index: src/Comm.c
===================================================================
--- src/Comm.c	(revision 527)
+++ src/Comm.c	(working copy)
@@ -1062,7 +1062,7 @@
   /* return if no storage assigned */
   if (! GA->storage)
   {
-    CCTK_VWarn(2, __LINE__, __FILE__, CCTK_THORNSTRING,
+    CCTK_VWarn(CCTK_WARN_DEBUG, __LINE__, __FILE__, CCTK_THORNSTRING,
                "Trying to synchronize variable '%s' with no storage", GA->name);
     return (0);
   }
Index: src/PostReceiveGA.c
===================================================================
--- src/PostReceiveGA.c	(revision 527)
+++ src/PostReceiveGA.c	(working copy)
@@ -56,7 +56,7 @@
   /* return if no storage assigned */
   if (! GA->storage)
   {
-    CCTK_VWarn(2, __LINE__, __FILE__, CCTK_THORNSTRING,
+    CCTK_VWarn(CCTK_WARN_DEBUG, __LINE__, __FILE__, CCTK_THORNSTRING,
                "Trying to synchronize variable '%s' with no storage", GA->name);
     return;
   }
