Index: lib/simrestart.py
===================================================================
--- lib/simrestart.py	(revision 1616)
+++ lib/simrestart.py	(working copy)
@@ -704,7 +704,7 @@
             'memory': memory,
             'cpufreq': cpufreq,
             'pbsSimulationName': pbsSimulationName,
-            'walltime': Walltime.Walltime
+            'walltime': walltt.Walltime
         }
         
         if chainedJobId == None:
