Index: dist/openmpi-1.10.1.tar.gz
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/octet-stream

Property changes on: dist/openmpi-1.10.1.tar.gz
___________________________________________________________________
Deleted: svn:mime-type
## -1 +0,0 ##
-application/octet-stream
\ No newline at end of property
Index: dist/openmpi-1.10.7.tar.gz
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/octet-stream

Property changes on: dist/openmpi-1.10.7.tar.gz
___________________________________________________________________
Added: svn:mime-type
## -0,0 +1 ##
+application/octet-stream
\ No newline at end of property
Index: src/build.pl
===================================================================
--- src/build.pl	(revision 87)
+++ src/build.pl	(working copy)
@@ -7,7 +7,7 @@
 
 # Set locations
 my $THORN = "MPI";
-my $NAME = "openmpi-1.10.1";
+my $NAME = "openmpi-1.10.7";
 my $INSTALL_DIR = undef;
 my $BUILD_DIR = undef;
 my $SRCDIR = $0;
Index: src/detect.pl
===================================================================
--- src/detect.pl	(revision 87)
+++ src/detect.pl	(working copy)
@@ -129,7 +129,6 @@
     }
 
     # Set locations
-    my $NAME = "openmpi-1.10.1";
     my $INSTALL_DIR = undef;
     my $BUILD_DIR = undef;
     my $SRCDIR = $0;
